<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Annotation-agnostic differential expression analysis of RNA-seq data at
    base-pair resolution via the DER Finder approach • derfinder</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Annotation-agnostic differential expression analysis of RNA-seq data at
    base-pair resolution via the DER Finder approach">
<meta property="og:description" content="This package provides functions for annotation-agnostic
    differential expression analysis of RNA-seq data. Two implementations of
    the DER Finder approach are included in this package: (1) single base-level
    F-statistics and (2) DER identification at the expressed regions-level.
    The DER Finder approach can also be used to identify differentially bounded
    ChIP-seq peaks.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="derfinder">derfinder<a class="anchor" aria-label="anchor" href="#derfinder"></a>
</h1></div>
<!-- badges: start -->

<p>Annotation-agnostic differential expression analysis of RNA-seq data at base-pair resolution via the DER Finder approach. This package contains two different implementations of this approach. The first one is the single base-level F-statistics implementation and the second one is via identifying expressed regions. For more information about <code>derfinder</code> check the vignettes <a href="http://www.bioconductor.org/packages/derfinder" class="external-link">here</a>.</p>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<p>For more information about <code>derfinder</code> check the vignettes <a href="http://bioconductor.org/packages/derfinder" class="external-link">through Bioconductor</a> or at the <a href="http://lcolladotor.github.io/derfinder" class="external-link">documentation website</a>.</p>
</div>
<div class="section level2">
<h2 id="further-documentation">Further documentation<a class="anchor" aria-label="anchor" href="#further-documentation"></a>
</h2>
<p>You can generate HTML reports from the results using <strong>regionReport</strong> available <a href="https://github.com/lcolladotor/regionReport" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="installation-instructions">Installation instructions<a class="anchor" aria-label="anchor" href="#installation-instructions"></a>
</h2>
<p>Get the latest stable <code>R</code> release from <a href="http://cran.r-project.org/" class="external-link">CRAN</a>. Then install <code>derfinder</code> using from <a href="http://bioconductor.org/" class="external-link">Bioconductor</a> the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"derfinder"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Below is the citation output from using <code>citation('derfinder')</code> in R. Please run this yourself to check for any updates on how to cite <strong>derfinder</strong>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"derfinder"</span><span class="op">)</span>, bibtex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Collado-Torres L, Nellore A, Frazee AC, Wilks C, Love MI, Langmead B,</span></span>
<span><span class="co">#&gt; Irizarry RA, Leek JT, Jaffe AE (2017). "Flexible expressed region</span></span>
<span><span class="co">#&gt; analysis for RNA-seq with derfinder." _Nucl. Acids Res._. doi:</span></span>
<span><span class="co">#&gt; 10.1093/nar/gkw852 (URL: https://doi.org/10.1093/nar/gkw852), &lt;URL:</span></span>
<span><span class="co">#&gt; http://nar.oxfordjournals.org/content/early/2016/09/29/nar.gkw852&gt;.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; A BibTeX entry for LaTeX users is</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   @Article{,</span></span>
<span><span class="co">#&gt;     title = {Flexible expressed region analysis for RNA-seq with derfinder},</span></span>
<span><span class="co">#&gt;     author = {Leonardo Collado-Torres and Abhinav Nellore and Alyssa C. Frazee and Christopher Wilks and Michael I. Love and Ben Langmead and Rafael A. Irizarry and Jeffrey T. Leek and Andrew E. Jaffe},</span></span>
<span><span class="co">#&gt;     year = {2017},</span></span>
<span><span class="co">#&gt;     journal = {Nucl. Acids Res.},</span></span>
<span><span class="co">#&gt;     doi = {10.1093/nar/gkw852},</span></span>
<span><span class="co">#&gt;     url = {http://nar.oxfordjournals.org/content/early/2016/09/29/nar.gkw852},</span></span>
<span><span class="co">#&gt;   }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Frazee AC, Sabunciyan S, Hansen KD, Irizarry RA, Leek JT (2014).</span></span>
<span><span class="co">#&gt; "Differential expression analysis of RNA-seq data at single-base</span></span>
<span><span class="co">#&gt; resolution." _Biostatistics_, *15 (3)*, 413-426. doi:</span></span>
<span><span class="co">#&gt; 10.1093/biostatistics/kxt053 (URL:</span></span>
<span><span class="co">#&gt; https://doi.org/10.1093/biostatistics/kxt053), &lt;URL:</span></span>
<span><span class="co">#&gt; http://biostatistics.oxfordjournals.org/content/15/3/413.long&gt;.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; A BibTeX entry for LaTeX users is</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   @Article{,</span></span>
<span><span class="co">#&gt;     title = {Differential expression analysis of RNA-seq data at single-base resolution},</span></span>
<span><span class="co">#&gt;     author = {Alyssa C. Frazee and Sarven Sabunciyan and Kasper D. Hansen and Rafael A. Irizarry and Jeffrey T. Leek},</span></span>
<span><span class="co">#&gt;     year = {2014},</span></span>
<span><span class="co">#&gt;     journal = {Biostatistics},</span></span>
<span><span class="co">#&gt;     volume = {15 (3)},</span></span>
<span><span class="co">#&gt;     pages = {413-426},</span></span>
<span><span class="co">#&gt;     doi = {10.1093/biostatistics/kxt053},</span></span>
<span><span class="co">#&gt;     url = {http://biostatistics.oxfordjournals.org/content/15/3/413.long},</span></span>
<span><span class="co">#&gt;   }</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Collado-Torres L, Jaffe AE, Leek JT (2017). _derfinder:</span></span>
<span><span class="co">#&gt; Annotation-agnostic differential expression analysis of RNA-seq data at</span></span>
<span><span class="co">#&gt; base-pair resolution via the DER Finder approach_. doi:</span></span>
<span><span class="co">#&gt; 10.18129/B9.bioc.derfinder (URL:</span></span>
<span><span class="co">#&gt; https://doi.org/10.18129/B9.bioc.derfinder),</span></span>
<span><span class="co">#&gt; https://github.com/lcolladotor/derfinder - R package version 1.25.3,</span></span>
<span><span class="co">#&gt; &lt;URL: http://www.bioconductor.org/packages/derfinder&gt;.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; A BibTeX entry for LaTeX users is</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   @Manual{,</span></span>
<span><span class="co">#&gt;     title = {derfinder: Annotation-agnostic differential expression analysis of RNA-seq data at base-pair resolution via the DER Finder approach},</span></span>
<span><span class="co">#&gt;     author = {Leonardo Collado-Torres and Andrew E. Jaffe and Jeffrey T. Leek},</span></span>
<span><span class="co">#&gt;     year = {2017},</span></span>
<span><span class="co">#&gt;     url = {http://www.bioconductor.org/packages/derfinder},</span></span>
<span><span class="co">#&gt;     note = {https://github.com/lcolladotor/derfinder - R package version 1.25.3},</span></span>
<span><span class="co">#&gt;     doi = {10.18129/B9.bioc.derfinder},</span></span>
<span><span class="co">#&gt;   }</span></span></code></pre></div>
<p>Please note that the <code>derfinder</code> was only made possible thanks to many other R and bioinformatics software authors, which are cited either in the vignettes and/or the paper(s) describing this package.</p>
</div>
<div class="section level2">
<h2 id="der-finder-versions">DER Finder versions<a class="anchor" aria-label="anchor" href="#der-finder-versions"></a>
</h2>
<ul>
<li>The original implementation of the DER Finder approach as published in Frazee et al, Biostatistics 2014 is available via GitHub at <a href="https://github.com/leekgroup/derfinder" class="external-link">derfinder</a>.</li>
<li>The version implementing the single base-level approach via calculating F-stastics as described in the pre-print Collado-Torres et al, Nucleic Acids Research 2017 is available via Bioconductor at <a href="http://bioconductor.org/packages/derfinder" class="external-link">derfinder</a>. The same package has the functions required for the expressed regions-level approach.</li>
</ul>
</div>
<div class="section level2">
<h2 id="code-of-conduct">Code of Conduct<a class="anchor" aria-label="anchor" href="#code-of-conduct"></a>
</h2>
<p>Please note that the derfinder project is released with a <a href="https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html" class="external-link">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
</div>
<div class="section level2">
<h2 id="development-tools">Development tools<a class="anchor" aria-label="anchor" href="#development-tools"></a>
</h2>
<ul>
<li>Continuous code testing is possible thanks to <a href="https://www.tidyverse.org/blog/2020/04/usethis-1-6-0/" class="external-link">GitHub actions</a> through <em><a href="https://CRAN.R-project.org/package=usethis" class="external-link">usethis</a></em>, <em><a href="https://CRAN.R-project.org/package=remotes" class="external-link">remotes</a></em>, <em><a href="https://github.com/r-hub/sysreqs" class="external-link">sysreqs</a></em> and <em><a href="https://CRAN.R-project.org/package=rcmdcheck" class="external-link">rcmdcheck</a></em> customized to use <a href="https://www.bioconductor.org/help/docker/" class="external-link">Bioconductor’s docker containers</a> and <em><a href="https://bioconductor.org/packages/3.13/BiocCheck" class="external-link">BiocCheck</a></em>.</li>
<li>Code coverage assessment is possible thanks to <a href="https://codecov.io/gh" class="external-link">codecov</a> and <em><a href="https://CRAN.R-project.org/package=covr" class="external-link">covr</a></em>.</li>
<li>The <a href="http://lcolladotor.github.io/derfinder" class="external-link">documentation website</a> is automatically updated thanks to <em><a href="https://CRAN.R-project.org/package=pkgdown" class="external-link">pkgdown</a></em>.</li>
<li>The code is styled automatically thanks to <em><a href="https://CRAN.R-project.org/package=styler" class="external-link">styler</a></em>.</li>
<li>The documentation is formatted thanks to <em><a href="https://CRAN.R-project.org/package=devtools" class="external-link">devtools</a></em> and <em><a href="https://CRAN.R-project.org/package=roxygen2" class="external-link">roxygen2</a></em>.</li>
</ul>
<p>For more details, check the <code>dev</code> directory.</p>
<p>This package was developed using <em><a href="https://bioconductor.org/packages/3.13/biocthis" class="external-link">biocthis</a></em>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/derfinder" class="external-link">View on Bioconductor</a></li>
<li><a href="https://github.com/lcolladotor/derfinder/" class="external-link">Browse source code</a></li>
<li><a href="https://support.bioconductor.org/t/derfinder/" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0" class="external-link">Artistic-2.0</a></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
<li><a href="SUPPORT.html">Getting help</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing derfinder</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Leonardo Collado-Torres <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-2140-308X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Andrew E. Jaffe <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-6886-1454" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jeffrey T. Leek <br><small class="roles"> Author, thesis advisor </small> <a href="https://orcid.org/0000-0002-2873-2671" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#stable" class="external-link"><img src="https://img.shields.io/badge/lifecycle-stable-brightgreen.svg" alt="Lifecycle: stable"></a></li>
<li><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/derfinder" class="external-link"><img src="http://www.bioconductor.org/shields/build/release/bioc/derfinder.svg" alt="BioC status"></a></li>
<li><a href="https://bioconductor.org/checkResults/devel/bioc-LATEST/derfinder" class="external-link"><img src="http://www.bioconductor.org/shields/build/devel/bioc/derfinder.svg" alt="BioC dev status"></a></li>
<li><a href="https://codecov.io/gh/lcolladotor/derfinder?branch=master" class="external-link"><img src="https://codecov.io/gh/lcolladotor/derfinder/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://github.com/lcolladotor/derfinder/actions" class="external-link"><img src="https://github.com/lcolladotor/derfinder/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a></li>
<li><a href="https://support.bioconductor.org/t/derfinder/" class="external-link"><img src="http://www.bioconductor.org/shields/posts/derfinder.svg" alt="Support site activity, last 6 months: tagged questions/avg. answers per question/avg. comments per question/accepted answers, or 0 if no tagged posts."></a></li>
<li><a href="https://github.com/lcolladotor/derfinder/issues" class="external-link"><img src="https://img.shields.io/github/issues/lcolladotor/derfinder" alt="GitHub issues"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
