<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Run the derfinder analysis on a chromosome — analyzeChr • derfinder</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Run the derfinder analysis on a chromosome — analyzeChr"><meta property="og:description" content="This is a major wrapper for running several key functions from this package.
It is meant to be used after loadCoverage has been used for a
specific chromosome. The steps run include makeModels,
preprocessCoverage, calculateStats, calculatePvalues
and annotating with annotateTranscripts and
matchGenes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run the derfinder analysis on a chromosome</h1>
    <small class="dont-index">Source: <a href="https://github.com/lcolladotor/derfinder/blob/HEAD/R/analyzeChr.R" class="external-link"><code>R/analyzeChr.R</code></a></small>
    <div class="hidden name"><code>analyzeChr.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is a major wrapper for running several key functions from this package.
It is meant to be used after <a href="loadCoverage.html">loadCoverage</a> has been used for a
specific chromosome. The steps run include <a href="makeModels.html">makeModels</a>,
<a href="preprocessCoverage.html">preprocessCoverage</a>, <a href="calculateStats.html">calculateStats</a>, <a href="calculatePvalues.html">calculatePvalues</a>
and annotating with <a href="https://rdrr.io/pkg/bumphunter/man/annotateTranscripts.html" class="external-link">annotateTranscripts</a> and
<a href="https://rdrr.io/pkg/bumphunter/man/matchGenes.html" class="external-link">matchGenes</a>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">analyzeChr</span><span class="op">(</span></span>
<span>  <span class="va">chr</span>,</span>
<span>  <span class="va">coverageInfo</span>,</span>
<span>  <span class="va">models</span>,</span>
<span>  cutoffPre <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  cutoffFstat <span class="op">=</span> <span class="fl">1e-08</span>,</span>
<span>  cutoffType <span class="op">=</span> <span class="st">"theoretical"</span>,</span>
<span>  nPermute <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seeds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">nPermute</span><span class="op">)</span>,</span>
<span>  <span class="va">groupInfo</span>,</span>
<span>  txdb <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  writeOutput <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  runAnnotation <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  lowMemDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">chr</span>, <span class="st">"chunksDir"</span><span class="op">)</span>,</span>
<span>  smooth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  smoothFunction <span class="op">=</span> <span class="fu">bumphunter</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/bumphunter/man/locfitByCluster.html" class="external-link">locfitByCluster</a></span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>chr</dt>
<dd><p>Used for naming the output files when <code>writeOutput=TRUE</code> and
the resulting <code>GRanges</code> object.</p></dd>


<dt>coverageInfo</dt>
<dd><p>A list containing a DataFrame --<code>$coverage</code>-- with
the coverage data and a logical Rle --<code>$position</code>-- with the positions
that passed the cutoff. This object is generated using <a href="loadCoverage.html">loadCoverage</a>.
You should have specified a cutoff value for <a href="loadCoverage.html">loadCoverage</a> unless
that you are using <code>colsubset</code> which will force a filtering step
with <a href="filterData.html">filterData</a> when running <a href="preprocessCoverage.html">preprocessCoverage</a>.</p></dd>


<dt>models</dt>
<dd><p>The output from <a href="makeModels.html">makeModels</a>.</p></dd>


<dt>cutoffPre</dt>
<dd><p>This argument is passed to <a href="preprocessCoverage.html">preprocessCoverage</a>
(<code>cutoff</code>).</p></dd>


<dt>cutoffFstat</dt>
<dd><p>This is used to determine the cutoff argument of
<a href="calculatePvalues.html">calculatePvalues</a> and it's behaviour is determined by
<code>cutoffType</code>.</p></dd>


<dt>cutoffType</dt>
<dd><p>If set to <code>empirical</code>, the <code>cutoffFstat</code>
(example: 0.99) quantile is used via <a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a>. If set to
<code>theoretical</code>, the theoretical <code>cutoffFstats</code> (example: 1e-08) is
calculated via <a href="https://rdrr.io/r/stats/Fdist.html" class="external-link">qf</a>. If set to <code>manual</code>, <code>cutoffFstats</code> is
passed to <a href="calculatePvalues.html">calculatePvalues</a> without any other calculation.</p></dd>


<dt>nPermute</dt>
<dd><p>The number of permutations. Note that for a full chromosome,
a small amount (10) of permutations is sufficient. If set to 0, no
permutations are performed and thus no null regions are used, however, the
<code>$regions</code> component is created.</p></dd>


<dt>seeds</dt>
<dd><p>An integer vector of length <code>nPermute</code> specifying the
seeds to be used for each permutation. If <code>NULL</code> no seeds are used.</p></dd>


<dt>groupInfo</dt>
<dd><p>A factor specifying the group membership of each sample
that can later be used with the plotting functions in the
<code>derfinderPlot</code> package.</p></dd>


<dt>txdb</dt>
<dd><p>This argument is passed to
<a href="https://rdrr.io/pkg/bumphunter/man/annotateTranscripts.html" class="external-link">annotateTranscripts</a>. If <code>NULL</code>,
TxDb.Hsapiens.UCSC.hg19.knownGene
is used.</p></dd>


<dt>writeOutput</dt>
<dd><p>If <code>TRUE</code>, output Rdata files are created at each
step inside a directory with the chromosome name (example: 'chr21' if
<code>chrnum='21'</code>). One Rdata file is created for each component described
in the return section.</p></dd>


<dt>runAnnotation</dt>
<dd><p>If <code>TRUE</code> <a href="https://rdrr.io/pkg/bumphunter/man/annotateTranscripts.html" class="external-link">annotateTranscripts</a>
and <a href="https://rdrr.io/pkg/bumphunter/man/matchGenes.html" class="external-link">matchGenes</a> are run. Otherwise these steps are skipped.</p></dd>


<dt>lowMemDir</dt>
<dd><p>If specified, each chunk is saved into a separate Rdata
file under <code>lowMemDir</code> and later loaded in
<a href="https://rdrr.io/pkg/derfinderHelper/man/fstats.apply.html" class="external-link">fstats.apply</a> when
running <a href="calculateStats.html">calculateStats</a> and <a href="calculatePvalues.html">calculatePvalues</a>. Using this option
helps reduce the memory load as each fork in <a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html" class="external-link">bplapply</a>
loads only the data needed for the chunk processing. The downside is a bit
longer computation time due to input/output.</p></dd>


<dt>smooth</dt>
<dd><p>Whether to smooth the F-statistics (<code>fstats</code>) or not. This
is by default <code>FALSE</code>. For RNA-seq data we recommend using <code>FALSE</code>.</p></dd>


<dt>weights</dt>
<dd><p>Weights used by the smoother as described in
<a href="https://rdrr.io/pkg/bumphunter/man/smoother.html" class="external-link">smoother</a>.</p></dd>


<dt>smoothFunction</dt>
<dd><p>A function to be used for smoothing the F-statistics.
Two functions are provided by the <code>bumphunter</code> package:
<a href="https://rdrr.io/pkg/bumphunter/man/loessByCluster.html" class="external-link">loessByCluster</a> and <a href="https://rdrr.io/pkg/bumphunter/man/runmedByCluster.html" class="external-link">runmedByCluster</a>. If
you are using your own custom function, it has to return a named list with
an element called <code>$fitted</code> that contains the smoothed F-statistics and
an element claled <code>$smoothed</code> that is a logical vector indicating
whether the F-statistics were smoothed or not. If they are not smoothed, the
original values will be used.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to other methods and/or advanced arguments.
Advanced arguments:</p><dl><dt>verbose </dt>
<dd><p>If <code>TRUE</code> basic status updates will be printed along
the way. Default <code>TRUE</code>.</p></dd>

<dt>scalefac </dt>
<dd><p>This argument is passed to <a href="preprocessCoverage.html">preprocessCoverage</a>.</p></dd>

<dt>chunksize </dt>
<dd><p>This argument is passed to <a href="preprocessCoverage.html">preprocessCoverage</a>.</p></dd>

<dt>returnOutput </dt>
<dd><p>If <code>TRUE</code>, it returns a list with the results
from each step. Otherwise, it returns <code>NULL</code>. Default: the opposite of
<code>writeOutput</code>.</p></dd>


</dl><p>Passed to <a href="extendedMapSeqlevels.html">extendedMapSeqlevels</a>, <a href="preprocessCoverage.html">preprocessCoverage</a>,
<a href="calculateStats.html">calculateStats</a>, <a href="calculatePvalues.html">calculatePvalues</a>,
<a href="https://rdrr.io/pkg/bumphunter/man/annotateTranscripts.html" class="external-link">annotateTranscripts</a>, <a href="https://rdrr.io/pkg/bumphunter/man/matchGenes.html" class="external-link">matchGenes</a>,
and <a href="define_cluster.html">define_cluster</a>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>If <code>returnOutput=TRUE</code>, a list with six components:</p><dl><dt>timeinfo </dt>
<dd><p>The wallclock timing information for each step.</p></dd>

<dt>optionsStats </dt>
<dd><p>The main options used when running this function.</p></dd>

<dt>coveragePrep </dt>
<dd><p>The output from <a href="preprocessCoverage.html">preprocessCoverage</a>.</p></dd>

<dt>fstats</dt>
<dd><p>The output from <a href="calculateStats.html">calculateStats</a>.</p></dd>

<dt>regions</dt>
<dd><p>The output from <a href="calculatePvalues.html">calculatePvalues</a>.</p></dd>

<dt>annotation</dt>
<dd><p>The output from <a href="https://rdrr.io/pkg/bumphunter/man/matchGenes.html" class="external-link">matchGenes</a>.</p></dd>


</dl><p>These are the same components that are written to Rdata files if
<code>writeOutput=TRUE</code>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>If you are working with data from an organism different from 'Homo sapiens'
specify so by setting the global 'species' and 'chrsStyle' options. For
example:
<code>options(species = 'arabidopsis_thaliana')</code>
<code>options(chrsStyle = 'NCBI')</code></p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="makeModels.html">makeModels</a>, <a href="preprocessCoverage.html">preprocessCoverage</a>,
<a href="calculateStats.html">calculateStats</a>, <a href="calculatePvalues.html">calculatePvalues</a>,
<a href="https://rdrr.io/pkg/bumphunter/man/annotateTranscripts.html" class="external-link">annotateTranscripts</a>, <a href="https://rdrr.io/pkg/bumphunter/man/matchGenes.html" class="external-link">matchGenes</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Leonardo Collado-Torres</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Collapse the coverage information</span></span></span>
<span class="r-in"><span><span class="va">collapsedFull</span> <span class="op">&lt;-</span> <span class="fu"><a href="collapseFullCoverage.html">collapseFullCoverage</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">genomeData</span><span class="op">$</span><span class="va">coverage</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:46.274007 collapseFullCoverage: Sorting fullCov</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:46.279638 collapseFullCoverage: Collapsing chromosomes information by sample</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Calculate library size adjustments</span></span></span>
<span class="r-in"><span><span class="va">sampleDepths</span> <span class="op">&lt;-</span> <span class="fu"><a href="sampleDepth.html">sampleDepth</a></span><span class="op">(</span><span class="va">collapsedFull</span>,</span></span>
<span class="r-in"><span>    probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span>, nonzero <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:46.283256 sampleDepth: Calculating sample quantiles</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:46.294342 sampleDepth: Calculating sample adjustments</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Build the models</span></span></span>
<span class="r-in"><span><span class="va">groupInfo</span> <span class="op">&lt;-</span> <span class="va">genomeInfo</span><span class="op">$</span><span class="va">pop</span></span></span>
<span class="r-in"><span><span class="va">adjustvars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">genomeInfo</span><span class="op">$</span><span class="va">gender</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">models</span> <span class="op">&lt;-</span> <span class="fu"><a href="makeModels.html">makeModels</a></span><span class="op">(</span><span class="va">sampleDepths</span>, testvars <span class="op">=</span> <span class="va">groupInfo</span>, adjustvars <span class="op">=</span> <span class="va">adjustvars</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Analyze the chromosome</span></span></span>
<span class="r-in"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">analyzeChr</span><span class="op">(</span></span></span>
<span class="r-in"><span>    chr <span class="op">=</span> <span class="st">"21"</span>, coverageInfo <span class="op">=</span> <span class="va">genomeData</span>, models <span class="op">=</span> <span class="va">models</span>,</span></span>
<span class="r-in"><span>    cutoffFstat <span class="op">=</span> <span class="fl">1</span>, cutoffType <span class="op">=</span> <span class="st">"manual"</span>, groupInfo <span class="op">=</span> <span class="va">groupInfo</span>, mc.cores <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>    writeOutput <span class="op">=</span> <span class="cn">FALSE</span>, returnOutput <span class="op">=</span> <span class="cn">TRUE</span>, method <span class="op">=</span> <span class="st">"regular"</span>,</span></span>
<span class="r-in"><span>    runAnnotation <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> extendedMapSeqlevels: sequence names mapped from NCBI to UCSC for species homo_sapiens</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:47.180357 analyzeChr: Pre-processing the coverage data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:47.596579 analyzeChr: Calculating statistics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:47.604873 calculateStats: calculating the F-statistics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:47.67367 analyzeChr: Calculating pvalues</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:47.673971 analyzeChr: Using the following manual cutoff for the F-statistics 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:47.67516 calculatePvalues: identifying data segments</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:47.685086 findRegions: segmenting information</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:47.692626 findRegions: identifying candidate regions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:47.775036 findRegions: identifying region clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:47.947191 calculatePvalues: calculating F-statistics for permutation 1 and seed 20230508</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:48.006135 findRegions: segmenting information</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:48.010265 findRegions: identifying candidate regions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:48.102272 calculatePvalues: calculating the p-values</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:48.14843 calculatePvalues: skipping q-value calculation.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:48.18245 analyzeChr: Annotating regions</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "timeinfo"     "optionsStats" "coveragePrep" "fstats"       "regions"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [6] "annotation"  </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

