<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Assign genomic states to regions — annotateRegions • derfinder</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Assign genomic states to regions — annotateRegions"><meta property="og:description" content="This function takes the regions found in calculatePvalues and assigns
them genomic states contructed with makeGenomicState. The main
workhorse functions are countOverlaps and
findOverlaps."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Assign genomic states to regions</h1>
    <small class="dont-index">Source: <a href="https://github.com/lcolladotor/derfinder/blob/HEAD/R/annotateRegions.R" class="external-link"><code>R/annotateRegions.R</code></a></small>
    <div class="hidden name"><code>annotateRegions.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes the regions found in <a href="calculatePvalues.html">calculatePvalues</a> and assigns
them genomic states contructed with <a href="makeGenomicState.html">makeGenomicState</a>. The main
workhorse functions are <a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">countOverlaps</a> and
<a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">findOverlaps</a>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">annotateRegions</span><span class="op">(</span><span class="va">regions</span>, <span class="va">genomicState</span>, annotate <span class="op">=</span> <span class="cn">TRUE</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>regions</dt>
<dd><p>The <code>$regions</code> output from <a href="calculatePvalues.html">calculatePvalues</a>.</p></dd>


<dt>genomicState</dt>
<dd><p>A GRanges object created with <a href="makeGenomicState.html">makeGenomicState</a>.
It can be either the <code>genomicState$fullGenome</code> or
<code>genomicState$codingGenome</code> component.</p></dd>


<dt>annotate</dt>
<dd><p>If <code>TRUE</code> then the regions are annotated by the genomic
state. Otherwise, only the overlaps between the regions and the genomic
states are computed.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to other methods and/or advanced arguments.
Advanced arguments:</p><dl><dt>verbose </dt>
<dd><p>If <code>TRUE</code> basic status updates will be printed along
the way.</p></dd>

<dt>ignore.strand </dt>
<dd><p>Passed on to
<a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">findOverlaps-methods</a> and
<a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">countOverlaps</a>. Default: <code>TRUE</code>.</p></dd>


</dl><p>Passed to <a href="extendedMapSeqlevels.html">extendedMapSeqlevels</a>, <a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">countOverlaps</a>
and <a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">findOverlaps-methods</a>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list with elements <code>countTable</code> and <code>annotationList</code></p>


<p>(only if <code>annotate=TRUE</code>).</p><dl><dt>countTable </dt>
<dd><p>This is a data.frame with the number of overlaps from the
regions vs the genomic states with one type per column. For example, if
<code>fullOrCoding='full'</code> then the columns are <code>exon</code>,
<code>intergenic</code> and <code>intron</code>.</p></dd>

<dt>annotationList </dt>
<dd><p>This is a <code>GRangesList</code> with the genomic states
that overlapped with the regions. The
names of this <code>GRangesList</code> correspond to the region index in
<code>regions</code>.</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>You might want to specify arguments such as <code>minoverlap</code> to control
how the overlaps are determined. See <a href="https://rdrr.io/pkg/GenomicRanges/man/findOverlaps-methods.html" class="external-link">findOverlaps</a>
for further details.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="makeGenomicState.html">makeGenomicState</a>, <a href="calculatePvalues.html">calculatePvalues</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Andrew Jaffe, Leonardo Collado-Torres</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Annotate regions, first two only</span></span></span>
<span class="r-in"><span><span class="va">annotatedRegions</span> <span class="op">&lt;-</span> <span class="fu">annotateRegions</span><span class="op">(</span></span></span>
<span class="r-in"><span>    regions <span class="op">=</span> <span class="va">genomeRegions</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>    genomicState <span class="op">=</span> <span class="va">genomicState</span><span class="op">$</span><span class="va">fullGenome</span>, minoverlap <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:11:33.311126 annotateRegions: counting</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:11:33.387289 annotateRegions: annotating</span>
<span class="r-in"><span><span class="va">annotatedRegions</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $countTable</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   exon intergenic intron</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    1          0      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    1          0      0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $annotationList</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRangesList object of length 2:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`1`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 1 range and 4 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        seqnames            ranges strand |   theRegion         tx_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;IntegerList&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2238    chr21 47409522-47409566      + |        exon         72950</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                tx_name          gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        &lt;CharacterList&gt; &lt;IntegerList&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2238      uc002zhu.1            55</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from hg19 genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $`2`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 1 range and 4 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        seqnames            ranges strand |   theRegion         tx_id</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;IntegerList&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2250    chr21 47411924-47411986      + |        exon         72950</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                tx_name          gene</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        &lt;CharacterList&gt; &lt;IntegerList&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2250      uc002zhu.1            55</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from hg19 genome</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

