<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate p-values and identify regions — calculatePvalues • derfinder</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate p-values and identify regions — calculatePvalues"><meta property="og:description" content="First, this function finds the regions of interest according to specified
cutoffs. Then it permutes the samples and re-calculates the F-statistics.
The area of the statistics from these segments are then used to calculate
p-values for the original regions."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate p-values and identify regions</h1>
    <small class="dont-index">Source: <a href="https://github.com/lcolladotor/derfinder/blob/HEAD/R/calculatePvalues.R" class="external-link"><code>R/calculatePvalues.R</code></a></small>
    <div class="hidden name"><code>calculatePvalues.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>First, this function finds the regions of interest according to specified
cutoffs. Then it permutes the samples and re-calculates the F-statistics.
The area of the statistics from these segments are then used to calculate
p-values for the original regions.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calculatePvalues</span><span class="op">(</span></span>
<span>  <span class="va">coveragePrep</span>,</span>
<span>  <span class="va">models</span>,</span>
<span>  <span class="va">fstats</span>,</span>
<span>  nPermute <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  seeds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.Date</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">nPermute</span><span class="op">)</span>,</span>
<span>  <span class="va">chr</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">fstats</span>, <span class="fl">0.99</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  significantCut <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>  lowMemDir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  smooth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  smoothFunction <span class="op">=</span> <span class="fu">bumphunter</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/bumphunter/man/locfitByCluster.html" class="external-link">locfitByCluster</a></span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>coveragePrep</dt>
<dd><p>A list with <code>$coverageProcessed</code>,
<code>$mclapplyIndex</code>, and <code>$position</code> normally generated using
<a href="preprocessCoverage.html">preprocessCoverage</a>.</p></dd>


<dt>models</dt>
<dd><p>A list with <code>$mod</code> and <code>$mod0</code> normally generated
using <a href="makeModels.html">makeModels</a>.</p></dd>


<dt>fstats</dt>
<dd><p>A numerical Rle with the F-statistics normally generated using
<a href="calculateStats.html">calculateStats</a>.</p></dd>


<dt>nPermute</dt>
<dd><p>The number of permutations. Note that for a full chromosome,
a small amount (10) of permutations is sufficient. If set to 0, no
permutations are performed and thus no null regions are used, however, the
<code>$regions</code> component is created.</p></dd>


<dt>seeds</dt>
<dd><p>An integer vector of length <code>nPermute</code> specifying the
seeds to be used for each permutation. If <code>NULL</code> no seeds are used.</p></dd>


<dt>chr</dt>
<dd><p>A single element character vector specifying the chromosome name.
This argument is passed to <a href="findRegions.html">findRegions</a>.</p></dd>


<dt>cutoff</dt>
<dd><p>F-statistic cutoff to use to determine segments.</p></dd>


<dt>significantCut</dt>
<dd><p>A vector of length two specifiying the cutoffs used to
determine significance. The first element is used to determine significance
for the P-values, while the second element is used for the Q-values (FDR
adjusted P-values).</p></dd>


<dt>lowMemDir</dt>
<dd><p>The directory where the processed chunks are saved when
using <a href="preprocessCoverage.html">preprocessCoverage</a> with a specified <code>lowMemDir</code>.</p></dd>


<dt>smooth</dt>
<dd><p>Whether to smooth the F-statistics (<code>fstats</code>) or not. This
is by default <code>FALSE</code>. For RNA-seq data we recommend using <code>FALSE</code>.</p></dd>


<dt>weights</dt>
<dd><p>Weights used by the smoother as described in
<a href="https://rdrr.io/pkg/bumphunter/man/smoother.html" class="external-link">smoother</a>.</p></dd>


<dt>smoothFunction</dt>
<dd><p>A function to be used for smoothing the F-statistics.
Two functions are provided by the <code>bumphunter</code> package:
<a href="https://rdrr.io/pkg/bumphunter/man/loessByCluster.html" class="external-link">loessByCluster</a> and <a href="https://rdrr.io/pkg/bumphunter/man/runmedByCluster.html" class="external-link">runmedByCluster</a>. If
you are using your own custom function, it has to return a named list with
an element called <code>$fitted</code> that contains the smoothed F-statistics and
an element claled <code>$smoothed</code> that is a logical vector indicating
whether the F-statistics were smoothed or not. If they are not smoothed, the
original values will be used.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to other methods and/or advanced arguments.
Advanced arguments:</p><dl><dt>verbose </dt>
<dd><p>If <code>TRUE</code> basic status updates will be printed along
the way.</p></dd>

<dt>scalefac </dt>
<dd><p>This argument is passed to
<a href="https://rdrr.io/pkg/derfinderHelper/man/fstats.apply.html" class="external-link">fstats.apply</a> and should be the same as the one used
in <a href="preprocessCoverage.html">preprocessCoverage</a>. Default: 32.</p></dd>

<dt>method </dt>
<dd><p>Has to be either 'Matrix' (default), 'Rle' or 'regular'. See
details in <a href="https://rdrr.io/pkg/derfinderHelper/man/fstats.apply.html" class="external-link">fstats.apply</a>.</p></dd>

<dt>adjustF </dt>
<dd><p>A single value to adjust that is added in the denominator
of the F-stat calculation. Useful when the Residual Sum of Squares of the
alternative model is very small. Default: 0.</p></dd>

<dt>writeOutput </dt>
<dd><p>If <code>TRUE</code> then the regions are saved before
calculating q-values, and then overwritten once the q-values are written.
This argument was introduced to save the results from the permutations (can
take some time) to investigate the problem described at
https://support.bioconductor.org/p/62026/</p></dd>

<dt>maxRegionGap </dt>
<dd><p>Passed to internal functions of <a href="findRegions.html">findRegions</a>.
Default: 0.</p></dd>


</dl><p>Passed to <a href="findRegions.html">findRegions</a>, <code>smoothFunction</code> and
<a href="define_cluster.html">define_cluster</a>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list with four components:</p><dl><dt>regions </dt>
<dd><p>is a GRanges with metadata columns given by
<a href="findRegions.html">findRegions</a> with the additional metadata column <code>pvalues</code>:
p-value of the region calculated via permutations of the samples;
<code>qvalues</code>: the qvalues calculated using <a href="https://rdrr.io/pkg/qvalue/man/qvalue.html" class="external-link">qvalue</a>;
<code>significant</code>: whether the p-value is less than 0.05 (by default);
<code>significantQval</code>: whether the q-value is less than 0.10 (by default).
It also includes the mean coverage of the region (mean from the mean
coverage at each base calculated in <a href="preprocessCoverage.html">preprocessCoverage</a>). Furthermore,
if <code>groupInfo</code> was not <code>NULL</code> in <a href="preprocessCoverage.html">preprocessCoverage</a>, then
the group mean coverage is calculated as well as the log 2 fold change
(using group 1 as the reference).</p></dd>

<dt>nullStats</dt>
<dd><p>is a numeric Rle with the mean of the null statistics by
segment.</p></dd>

<dt>nullWidths</dt>
<dd><p>is a numeric Rle with the length of each of the segments
in the null distribution. The area can be obtained by multiplying the
absolute <code>nullstats</code> by the corresponding lengths.</p></dd>

<dt>nullPermutation</dt>
<dd><p>is a Rle with the permutation number from which the
null region originated from.</p></dd>


</dl></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="findRegions.html">findRegions</a>, <a href="https://rdrr.io/pkg/derfinderHelper/man/fstats.apply.html" class="external-link">fstats.apply</a>,
<a href="https://rdrr.io/pkg/qvalue/man/qvalue.html" class="external-link">qvalue</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Leonardo Collado-Torres</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Collapse the coverage information</span></span></span>
<span class="r-in"><span><span class="va">collapsedFull</span> <span class="op">&lt;-</span> <span class="fu"><a href="collapseFullCoverage.html">collapseFullCoverage</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">genomeData</span><span class="op">$</span><span class="va">coverage</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:49.730899 collapseFullCoverage: Sorting fullCov</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:49.733398 collapseFullCoverage: Collapsing chromosomes information by sample</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Calculate library size adjustments</span></span></span>
<span class="r-in"><span><span class="va">sampleDepths</span> <span class="op">&lt;-</span> <span class="fu"><a href="sampleDepth.html">sampleDepth</a></span><span class="op">(</span><span class="va">collapsedFull</span>, probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:49.735516 sampleDepth: Calculating sample quantiles</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:49.741328 sampleDepth: Calculating sample adjustments</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Build the models</span></span></span>
<span class="r-in"><span><span class="va">group</span> <span class="op">&lt;-</span> <span class="va">genomeInfo</span><span class="op">$</span><span class="va">pop</span></span></span>
<span class="r-in"><span><span class="va">adjustvars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">genomeInfo</span><span class="op">$</span><span class="va">gender</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">models</span> <span class="op">&lt;-</span> <span class="fu"><a href="makeModels.html">makeModels</a></span><span class="op">(</span><span class="va">sampleDepths</span>, testvars <span class="op">=</span> <span class="va">group</span>, adjustvars <span class="op">=</span> <span class="va">adjustvars</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Preprocess the data</span></span></span>
<span class="r-in"><span><span class="co">## Automatic chunksize used to then compare 1 vs 4 cores in the 'do not run'</span></span></span>
<span class="r-in"><span><span class="co">## section</span></span></span>
<span class="r-in"><span><span class="va">prep</span> <span class="op">&lt;-</span> <span class="fu"><a href="preprocessCoverage.html">preprocessCoverage</a></span><span class="op">(</span><span class="va">genomeData</span>,</span></span>
<span class="r-in"><span>    groupInfo <span class="op">=</span> <span class="va">group</span>, cutoff <span class="op">=</span> <span class="fl">0</span>,</span></span>
<span class="r-in"><span>    scalefac <span class="op">=</span> <span class="fl">32</span>, chunksize <span class="op">=</span> <span class="cn">NULL</span>, colsubset <span class="op">=</span> <span class="cn">NULL</span>, mc.cores <span class="op">=</span> <span class="fl">4</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get the F statistics</span></span></span>
<span class="r-in"><span><span class="va">fstats</span> <span class="op">&lt;-</span> <span class="va">genomeFstats</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## We recommend determining the cutoff to use based on the F-distribution</span></span></span>
<span class="r-in"><span><span class="co">## although you could also based it on the observed F-statistics.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## In this example we use a low cutoff used for illustrative purposes</span></span></span>
<span class="r-in"><span><span class="va">cutoff</span> <span class="op">&lt;-</span> <span class="fl">1</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Calculate the p-values and define the regions of interest.</span></span></span>
<span class="r-in"><span><span class="va">regsWithP</span> <span class="op">&lt;-</span> <span class="fu">calculatePvalues</span><span class="op">(</span><span class="va">prep</span>, <span class="va">models</span>, <span class="va">fstats</span>,</span></span>
<span class="r-in"><span>    nPermute <span class="op">=</span> <span class="fl">1</span>, seeds <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>    chr <span class="op">=</span> <span class="st">"chr21"</span>, cutoff <span class="op">=</span> <span class="va">cutoff</span>, mc.cores <span class="op">=</span> <span class="fl">1</span>, method <span class="op">=</span> <span class="st">"regular"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:50.051702 calculatePvalues: identifying data segments</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:50.054925 findRegions: segmenting information</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:50.058097 findRegions: identifying candidate regions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:50.095897 findRegions: identifying region clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:50.179223 calculatePvalues: calculating F-statistics for permutation 1 and seed 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:50.420165 findRegions: segmenting information</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:50.423615 findRegions: identifying candidate regions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:50.473597 calculatePvalues: calculating the p-values</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:05:50.497118 calculatePvalues: skipping q-value calculation.</span>
<span class="r-in"><span><span class="va">regsWithP</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $regions</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 33 ranges and 14 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      seqnames            ranges strand |     value      area indexStart</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;  &lt;integer&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up    chr21 47409522-47409560      * |  15.05364   587.092        128</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up    chr21 47411924-47411986      * |   7.09670   447.092        618</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up    chr21 47417352-47417397      * |   8.77845   403.808       1292</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up    chr21 47412662-47412724      * |   5.67122   357.287        843</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up    chr21 47408998-47409026      * |  10.68525   309.872         69</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..      ...               ...    ... .       ...       ...        ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up    chr21 47417666-47417672      * |   1.21552   8.50863       1390</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up    chr21 47409685-47409688      * |   1.21585   4.86342        192</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up    chr21 47417335-47417337      * |   1.59162   4.77487       1275</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up    chr21          47418068      * |   3.99766   3.99766       1434</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up    chr21 47410311-47410312      * |   1.14363   2.28725        246</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       indexEnd cluster clusterL meanCoverage   meanCEU   meanYRI</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;integer&gt;   &lt;Rle&gt;    &lt;Rle&gt;    &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up       166       3      167     0.415219  0.612943 0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up       680       6      389     0.584741  0.837491 0.0539683</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up      1337      10      338     0.364656  0.533126 0.0108696</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up       905       7       63     0.501280  0.739985 0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up        97       2      232     0.195773  0.288998 0.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..       ...     ...      ...          ...       ...       ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up      1396      10      338    0.2258065  0.333333       0.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up       195       3      167    0.4758065  0.654762       0.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up      1277      10      338    0.1290323  0.190476       0.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up      1434      11        1    0.0322581  0.047619       0.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up       247       4      159    0.7258065  1.071429       0.0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      log2FoldChangeYRIvsCEU   pvalues significant   qvalues significantQval</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   &lt;numeric&gt; &lt;numeric&gt;    &lt;factor&gt; &lt;logical&gt;       &lt;logical&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up                   -Inf 0.0175439       TRUE       &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up               -3.95589 0.0175439       TRUE       &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up               -5.61611 0.0526316       FALSE      &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up                   -Inf 0.1578947       FALSE      &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up                   -Inf 0.1929825       FALSE      &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..                    ...       ...         ...       ...             ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up                   -Inf  0.859649       FALSE      &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up               -2.71097  0.894737       FALSE      &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up                   -Inf  0.894737       FALSE      &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up                   -Inf  0.894737       FALSE      &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   up                   -Inf  0.929825       FALSE      &lt;NA&gt;            &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $nullStats</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> numeric-Rle of length 56 with 54 runs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Lengths:       1       1       1       1 ...       1       1       1       1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Values : 2.25473 3.63681 4.12769 9.73262 ... 1.16174 8.10124 5.20438 1.16174</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $nullWidths</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> integer-Rle of length 56 with 48 runs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Lengths:  1  1  1  1  2  2  1  1  2  1  1 ...  1  1  1  1  2  1  1  1  1  1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Values :  9 20 54 45 27 54 38 21 36  3  8 ... 63 45 36 38 36  3  1 47 39  1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $nullPermutation</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> integer-Rle of length 56 with 1 run</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Lengths: 56</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Values :  1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co">## Calculate again, but with 10 permutations instead of just 1</span></span></span>
<span class="r-in"><span><span class="va">regsWithP</span> <span class="op">&lt;-</span> <span class="fu">calculatePvalues</span><span class="op">(</span><span class="va">prep</span>, <span class="va">models</span>, <span class="va">fstats</span>,</span></span>
<span class="r-in"><span>    nPermute <span class="op">=</span> <span class="fl">10</span>, seeds <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span></span>
<span class="r-in"><span>    chr <span class="op">=</span> <span class="st">"chr21"</span>, cutoff <span class="op">=</span> <span class="va">cutoff</span>, mc.cores <span class="op">=</span> <span class="fl">2</span>, method <span class="op">=</span> <span class="st">"regular"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Check that they are the same as the previously calculated regions</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://testthat.r-lib.org" class="external-link">testthat</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://testthat.r-lib.org/reference/expect_that.html" class="external-link">expect_that</a></span><span class="op">(</span><span class="va">regsWithP</span>, <span class="fu"><a href="https://testthat.r-lib.org/reference/oldskool.html" class="external-link">equals</a></span><span class="op">(</span><span class="va">genomeRegions</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Histogram of the theoretical p-values by region</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Fdist.html" class="external-link">pf</a></span><span class="op">(</span><span class="va">regsWithP</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">value</span>, <span class="va">df1</span> <span class="op">-</span> <span class="va">df0</span>, <span class="va">n</span> <span class="op">-</span> <span class="va">df1</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"Distribution</span></span></span>
<span class="r-in"><span><span class="st">    original p-values by region"</span>, freq <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Histogram of the permutted p-values by region</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">regsWithP</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">pvalues</span>, main <span class="op">=</span> <span class="st">"Distribution permutted p-values by</span></span></span>
<span class="r-in"><span><span class="st">    region"</span>, freq <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## MA style plot</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    mean <span class="op">=</span> <span class="va">regsWithP</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">meanCoverage</span>,</span></span>
<span class="r-in"><span>    log2FoldChange <span class="op">=</span> <span class="va">regsWithP</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">log2FoldChangeYRIvsCEU</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">ma</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">mean</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">log2FoldChange</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Fold Change (log2)"</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Mean coverage (log2)"</span><span class="op">)</span> <span class="op">+</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"MA style plot"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Annotate the results</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/rafalab/bumphunter" class="external-link">"bumphunter"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bumphunter/man/annotateTranscripts.html" class="external-link">annotateTranscripts</a></span><span class="op">(</span><span class="fu">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/TxDb.Hsapiens.UCSC.hg19.knownGene/man/package.html" class="external-link">TxDb.Hsapiens.UCSC.hg19.knownGene</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bumphunter/man/matchGenes.html" class="external-link">matchGenes</a></span><span class="op">(</span><span class="va">regsWithP</span><span class="op">$</span><span class="va">regions</span>, <span class="va">genes</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">annotation</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

