<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate F-statistics at base pair resolution from a loaded BAM files — calculateStats • derfinder</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate F-statistics at base pair resolution from a loaded BAM files — calculateStats"><meta property="og:description" content="After defining the models of interest (see makeModels) and
pre-processing the data (see preprocessCoverage), use
calculateStats to calculate the F-statistics at base-pair resolution."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate F-statistics at base pair resolution from a loaded BAM files</h1>
    <small class="dont-index">Source: <a href="https://github.com/lcolladotor/derfinder/blob/HEAD/R/calculateStats.R" class="external-link"><code>R/calculateStats.R</code></a></small>
    <div class="hidden name"><code>calculateStats.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>After defining the models of interest (see <a href="makeModels.html">makeModels</a>) and
pre-processing the data (see <a href="preprocessCoverage.html">preprocessCoverage</a>), use
calculateStats to calculate the F-statistics at base-pair resolution.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">calculateStats</span><span class="op">(</span><span class="va">coveragePrep</span>, <span class="va">models</span>, lowMemDir <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>coveragePrep</dt>
<dd><p>A list with <code>$coverageProcessed</code>,
<code>$mclapplyIndex</code>, and <code>$position</code> normally generated using
<a href="preprocessCoverage.html">preprocessCoverage</a>.</p></dd>


<dt>models</dt>
<dd><p>A list with <code>$mod</code> and <code>$mod0</code> normally generated
using <a href="makeModels.html">makeModels</a>.</p></dd>


<dt>lowMemDir</dt>
<dd><p>The directory where the processed chunks are saved when
using <a href="preprocessCoverage.html">preprocessCoverage</a> with a specified <code>lowMemDir</code>.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to other methods and/or advanced arguments.
Advanced arguments:</p><dl><dt>verbose </dt>
<dd><p>If <code>TRUE</code> basic status updates will be printed along
the way.</p></dd>

<dt>scalefac </dt>
<dd><p>This argument is passed to
<a href="https://rdrr.io/pkg/derfinderHelper/man/fstats.apply.html" class="external-link">fstats.apply</a> and should be the same as the one used
in <a href="preprocessCoverage.html">preprocessCoverage</a>. Default: 32.</p></dd>

<dt>method </dt>
<dd><p>Has to be either 'Matrix' (default), 'Rle' or 'regular'. See
details in <a href="https://rdrr.io/pkg/derfinderHelper/man/fstats.apply.html" class="external-link">fstats.apply</a>.</p></dd>

<dt>adjustF </dt>
<dd><p>A single value to adjust that is added in the denominator
of the F-stat calculation. Useful when the Residual Sum of Squares of the
alternative model is very small. Default: 0.</p></dd>


</dl><p>Passed to <a href="define_cluster.html">define_cluster</a>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A numeric Rle with the F-statistics per base pair that passed the
cutoff.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="makeModels.html">makeModels</a>, <a href="preprocessCoverage.html">preprocessCoverage</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Leonardo Collado-Torres</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Collapse the coverage information</span></span></span>
<span class="r-in"><span><span class="va">collapsedFull</span> <span class="op">&lt;-</span> <span class="fu"><a href="collapseFullCoverage.html">collapseFullCoverage</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">genomeData</span><span class="op">$</span><span class="va">coverage</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 06:01:09.436288 collapseFullCoverage: Sorting fullCov</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 06:01:09.438749 collapseFullCoverage: Collapsing chromosomes information by sample</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Calculate library size adjustments</span></span></span>
<span class="r-in"><span><span class="va">sampleDepths</span> <span class="op">&lt;-</span> <span class="fu"><a href="sampleDepth.html">sampleDepth</a></span><span class="op">(</span><span class="va">collapsedFull</span>, probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 06:01:09.440814 sampleDepth: Calculating sample quantiles</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 06:01:09.44656 sampleDepth: Calculating sample adjustments</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Build the models</span></span></span>
<span class="r-in"><span><span class="va">group</span> <span class="op">&lt;-</span> <span class="va">genomeInfo</span><span class="op">$</span><span class="va">pop</span></span></span>
<span class="r-in"><span><span class="va">adjustvars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">genomeInfo</span><span class="op">$</span><span class="va">gender</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">models</span> <span class="op">&lt;-</span> <span class="fu"><a href="makeModels.html">makeModels</a></span><span class="op">(</span><span class="va">sampleDepths</span>, testvars <span class="op">=</span> <span class="va">group</span>, adjustvars <span class="op">=</span> <span class="va">adjustvars</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Preprocess the data</span></span></span>
<span class="r-in"><span><span class="va">prep</span> <span class="op">&lt;-</span> <span class="fu"><a href="preprocessCoverage.html">preprocessCoverage</a></span><span class="op">(</span><span class="va">genomeData</span>,</span></span>
<span class="r-in"><span>    cutoff <span class="op">=</span> <span class="fl">0</span>, scalefac <span class="op">=</span> <span class="fl">32</span>,</span></span>
<span class="r-in"><span>    chunksize <span class="op">=</span> <span class="fl">1e3</span>, colsubset <span class="op">=</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Run the function</span></span></span>
<span class="r-in"><span><span class="va">fstats</span> <span class="op">&lt;-</span> <span class="fu">calculateStats</span><span class="op">(</span><span class="va">prep</span>, <span class="va">models</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>, method <span class="op">=</span> <span class="st">"regular"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 06:01:09.708332 calculateStats: calculating the F-statistics</span>
<span class="r-in"><span><span class="va">fstats</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> numeric-Rle of length 1434 with 361 runs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Lengths:           1           5           3 ...           1           1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Values : 5.10948e+00 3.85229e+00 5.12448e-01 ... 0.032237427 3.997655116</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co">## Compare vs pre-packaged F-statistics</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://testthat.r-lib.org" class="external-link">"testthat"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://testthat.r-lib.org/reference/expect_that.html" class="external-link">expect_that</a></span><span class="op">(</span><span class="va">fstats</span>, <span class="fu"><a href="https://testthat.r-lib.org/reference/oldskool.html" class="external-link">is_equivalent_to</a></span><span class="op">(</span><span class="va">genomeFstats</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

