<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Extract coverage information for exons — coverageToExon • derfinder</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Extract coverage information for exons — coverageToExon" />
<meta property="og:description" content="This function extracts the coverage information calculated by
fullCoverage for a set of exons determined by
makeGenomicState. The underlying code is similar to
getRegionCoverage with additional tweaks for calculating RPKM values." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.21.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/lcolladotor/derfinder/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Extract coverage information for exons</h1>
    <small class="dont-index">Source: <a href='https://github.com/lcolladotor/derfinder/blob/master/R/coverageToExon.R'><code>R/coverageToExon.R</code></a></small>
    <div class="hidden name"><code>coverageToExon.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function extracts the coverage information calculated by
<a href='fullCoverage.html'>fullCoverage</a> for a set of exons determined by
<a href='makeGenomicState.html'>makeGenomicState</a>. The underlying code is similar to
<a href='getRegionCoverage.html'>getRegionCoverage</a> with additional tweaks for calculating RPKM values.</p>
    </div>

    <pre class="usage"><span class='fu'>coverageToExon</span>(
  <span class='kw'>fullCov</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>genomicState</span>,
  <span class='kw'>L</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>returnType</span> <span class='kw'>=</span> <span class='st'>"raw"</span>,
  <span class='kw'>files</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>fullCov</th>
      <td><p>A list where each element is the result from
<a href='loadCoverage.html'>loadCoverage</a> used with <code>returnCoverage = TRUE</code>. Can be generated
using <a href='fullCoverage.html'>fullCoverage</a>. Alternatively, specify <code>files</code> to extract
the coverage information from the regions of interest. This can be
helpful if you do not wish to store <code>fullCov</code> for memory reasons.</p></td>
    </tr>
    <tr>
      <th>genomicState</th>
      <td><p>A GRanges object created with <a href='makeGenomicState.html'>makeGenomicState</a>.
It can be either the <code>genomicState$fullGenome</code> or
<code>genomicState$codingGenome</code> component.</p></td>
    </tr>
    <tr>
      <th>L</th>
      <td><p>The width of the reads used. Either a vector of length 1 or length
equal to the number of samples.</p></td>
    </tr>
    <tr>
      <th>returnType</th>
      <td><p>If <code>raw</code>, then the raw coverage information per exon
is returned. If <code>rpkm</code>, RPKM values are calculated for each exon.</p></td>
    </tr>
    <tr>
      <th>files</th>
      <td><p>A character vector with the full path to the sample BAM files
(or BigWig files).
The names are used for the column names of the DataFrame. Check
<a href='rawFiles.html'>rawFiles</a> for constructing <code>files</code>. <code>files</code> can also be a
<code>BamFileList</code> object created with <a href='https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html'>BamFileList</a> or a
<code>BigWigFileList</code> object created with <a href='https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html'>BigWigFileList</a>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Arguments passed to other methods and/or advanced arguments.
Advanced arguments:</p><dl'>
<dt>verbose </dt><dd><p>If <code>TRUE</code> basic status updates will be printed along
the way.</p></dd>
<dt>BPPARAM.strandStep </dt><dd><p>A BPPARAM object to use for the strand step. If
not specified, then <code>strandCores</code> specifies the number of cores to use
for the strand step. The actual number of cores used is the minimum of
<code>strandCores</code>, <code>mc.cores</code> and the number of strands in the data.</p></dd>
<dt>BPPARAM.chrStep </dt><dd><p>A BPPRAM object to use for the chr step. If not
specified, then <code>mc.cores</code> specifies the number of cores to use for
the chr step. The actual number of cores used is the minimum of
<code>mc.cores</code> and the number of samples.</p></dd>

</dl><p>Passed to <a href='extendedMapSeqlevels.html'>extendedMapSeqlevels</a> and <a href='define_cluster.html'>define_cluster</a>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A matrix (nrow = number of exons in <code>genomicState</code>
corresponding to the chromosomes in <code>fullCov</code>, ncol = number of
samples) with the number of reads (or RPKM) per exon. The row names
correspond to the row indexes of <code>genomicState$fullGenome</code>  (if
<code>fullOrCoding='full'</code>) or <code>genomicState$codingGenome</code> (if
<code>fullOrCoding='coding'</code>).</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Parallelization is used twice.
First, it is used by strand. Second, for processing the exons by
chromosome. So there is no gain in using <code>mc.cores</code> greater than the
maximum of the number of strands and number of chromosomes.</p>
<p>If <code>fullCov</code> is <code>NULL</code> and <code>files</code> is specified, this function
will attempt to read the coverage from the files. Note that if you used
'totalMapped' and 'targetSize' before, you will have to specify them again
to get the same results.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='fullCoverage.html'>fullCoverage</a>, <a href='getRegionCoverage.html'>getRegionCoverage</a></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Obtain fullCov object</span>
<span class='no'>fullCov</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='st'>"21"</span> <span class='kw'>=</span> <span class='no'>genomeDataRaw</span>$<span class='no'>coverage</span>)

<span class='co'>## Use only the first two exons</span>
<span class='no'>smallGenomicState</span> <span class='kw'>&lt;-</span> <span class='no'>genomicState</span>
<span class='no'>smallGenomicState</span>$<span class='no'>fullGenome</span> <span class='kw'>&lt;-</span> <span class='no'>smallGenomicState</span>$<span class='no'>fullGenome</span>[
    <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span>(<span class='no'>smallGenomicState</span>$<span class='no'>fullGenome</span>$<span class='no'>theRegion</span> <span class='kw'>==</span> <span class='st'>"exon"</span>)[<span class='fl'>1</span>:<span class='fl'>2</span>]
]

<span class='co'>## Finally, get the coverage information for each exon</span>
<span class='no'>exonCov</span> <span class='kw'>&lt;-</span> <span class='fu'>coverageToExon</span>(
    <span class='kw'>fullCov</span> <span class='kw'>=</span> <span class='no'>fullCov</span>,
    <span class='kw'>genomicState</span> <span class='kw'>=</span> <span class='no'>smallGenomicState</span>$<span class='no'>fullGenome</span>, <span class='kw'>L</span> <span class='kw'>=</span> <span class='fl'>36</span>
)</div><div class='output co'>#&gt; <span class='message'>extendedMapSeqlevels: sequence names mapped from NCBI to UCSC for species homo_sapiens</span></div><div class='output co'>#&gt; class: SerialParam
#&gt;   bpisup: FALSE; bpnworkers: 1; bptasks: 0; bpjobname: BPJOB
#&gt;   bplog: FALSE; bpthreshold: INFO; bpstopOnError: TRUE
#&gt;   bpRNGseed: ; bptimeout: 2592000; bpprogressbar: FALSE
#&gt;   bpexportglobals: TRUE
#&gt;   bplogdir: NA
#&gt;   bpresultdir: NA
#&gt; class: SerialParam
#&gt;   bpisup: FALSE; bpnworkers: 1; bptasks: 0; bpjobname: BPJOB
#&gt;   bplog: FALSE; bpthreshold: INFO; bpstopOnError: TRUE
#&gt;   bpRNGseed: ; bptimeout: 2592000; bpprogressbar: FALSE
#&gt;   bpexportglobals: TRUE
#&gt;   bplogdir: NA
#&gt;   bpresultdir: NA</div><div class='output co'>#&gt; <span class='message'>2020-04-18 06:56:10 coverageToExon: processing chromosome chr21</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


