<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Filter the positions of interest — filterData • derfinder</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Filter the positions of interest — filterData"><meta property="og:description" content="For a group of samples this function reads the coverage information for a
specific chromosome directly from the BAM files. It then merges them into a
DataFrame and removes the bases that do not pass the cutoff. This is a
helper function for loadCoverage and preprocessCoverage."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Filter the positions of interest</h1>
    <small class="dont-index">Source: <a href="https://github.com/lcolladotor/derfinder/blob/HEAD/R/filterData.R" class="external-link"><code>R/filterData.R</code></a></small>
    <div class="hidden name"><code>filterData.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>For a group of samples this function reads the coverage information for a
specific chromosome directly from the BAM files. It then merges them into a
DataFrame and removes the bases that do not pass the cutoff. This is a
helper function for <a href="loadCoverage.html">loadCoverage</a> and <a href="preprocessCoverage.html">preprocessCoverage</a>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">filterData</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  index <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter <span class="op">=</span> <span class="st">"one"</span>,</span>
<span>  totalMapped <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  targetSize <span class="op">=</span> <span class="fl">8e+07</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>data</dt>
<dd><p>Either a list of Rle objects or a DataFrame with the coverage
information.</p></dd>


<dt>cutoff</dt>
<dd><p>The base-pair level cutoff to use. It's behavior is controlled
by <code>filter</code>.</p></dd>


<dt>index</dt>
<dd><p>A logical Rle with the positions of the chromosome that passed
the cutoff. If <code>NULL</code> it is assumed that this is the first time using
filterData and thus no previous index exists.</p></dd>


<dt>filter</dt>
<dd><p>Has to be either <code>'one'</code> (default) or <code>'mean'</code>. In
the first case, at least one sample has to have coverage above <code>cutoff</code>.
In the second case, the mean coverage has to be greater than <code>cutoff</code>.</p></dd>


<dt>totalMapped</dt>
<dd><p>A vector with the total number of reads mapped for each
sample. The vector should be in the same order as the samples in <code>data</code>.
Providing this data adjusts the coverage to reads in <code>targetSize</code>
library prior to filtering. See <a href="getTotalMapped.html">getTotalMapped</a> for
calculating this vector.</p></dd>


<dt>targetSize</dt>
<dd><p>The target library size to adjust the coverage to. Used
only when <code>totalMapped</code> is specified. By default, it adjusts to
libraries with 80 million reads.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to other methods and/or advanced arguments.
Advanced arguments:</p><dl><dt>verbose </dt>
<dd><p>If <code>TRUE</code> basic status updates will be printed along
the way.</p></dd>

<dt>returnMean </dt>
<dd><p>If <code>TRUE</code> the mean coverage is included in the
result. <code>FALSE</code> by default.</p></dd>

<dt>returnCoverage </dt>
<dd><p>If <code>TRUE</code>, the coverage DataFrame is returned.
<code>TRUE</code> by default.</p></dd>


</dl></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list with up to three components.</p>


<dl><dt>coverage </dt>
<dd><p>is a DataFrame object where each column represents a
sample. The number of rows depends on the number of base pairs that passed
the cutoff and the information stored is the coverage at that given base.
Included only when <code>returnCoverage = TRUE</code>.</p></dd>

<dt>position </dt>
<dd><p>is a logical Rle with the positions of the chromosome
that passed the cutoff.</p></dd>

<dt>meanCoverage </dt>
<dd><p>is a numeric Rle with the mean coverage at each base.
Included only when <code>returnMean = TRUE</code>.</p></dd>

<dt>colnames </dt>
<dd><p>Specifies the column names to be used for the results
DataFrame. If <code>NULL</code>, names from <code>data</code> are used.</p></dd>

<dt>smoothMean </dt>
<dd><p>Whether to smooth the mean. Used only when
<code>filter = 'mean'</code>. This option is used internally by
<a href="regionMatrix.html">regionMatrix</a>.</p></dd>


</dl><p>Passed to the internal function <code>.smootherFstats</code>, see
<a href="findRegions.html">findRegions</a>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>If <code>cutoff</code> is <code>NULL</code> then the data is grouped into
DataFrame without applying any cutoffs. This can be useful if you want to
use <a href="loadCoverage.html">loadCoverage</a> to build the coverage DataFrame without applying any
cutoffs for other downstream purposes like plotting the coverage values of a
given region. You can always specify the <code>colsubset</code> argument in
<a href="preprocessCoverage.html">preprocessCoverage</a> to filter the data before calculating the F
statistics.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="loadCoverage.html">loadCoverage</a>, <a href="preprocessCoverage.html">preprocessCoverage</a>,
<a href="getTotalMapped.html">getTotalMapped</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Leonardo Collado-Torres</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Construct some toy data</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/IRanges" class="external-link">"IRanges"</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: BiocGenerics</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘BiocGenerics’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     IQR, mad, sd, var, xtabs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     as.data.frame, basename, cbind, colnames, dirname, do.call,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     tapply, union, unique, unsplit, which.max, which.min</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: S4Vectors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: stats4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘S4Vectors’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:utils’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     findMatches</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     I, expand.grid, unname</span>
<span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1e4</span>, max <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1e4</span>, max <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Rle-class.html" class="external-link">Rle</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1e4</span>, max <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">DF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">z</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Filter the data</span></span></span>
<span class="r-in"><span><span class="va">filt1</span> <span class="op">&lt;-</span> <span class="fu">filterData</span><span class="op">(</span><span class="va">DF</span>, <span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:11:43.483445 filterData: originally there were 10000 rows, now there are 8300 rows. Meaning that 17 percent was filtered.</span>
<span class="r-in"><span><span class="va">filt1</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $coverage</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 8300 rows and 3 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          x     y     z</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;Rle&gt; &lt;Rle&gt; &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        1     3     7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        7     3    10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        4    10     8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        8     6     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        6     1     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ...    ...   ...   ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8296     9     4     2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8297     4     8    10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8298     9     2     6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8299     8     0     9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8300     6     1     9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> logical-Rle of length 10000 with 2850 runs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Lengths:    13     1     1     1     1 ...     1     1     1     1     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Values :  TRUE FALSE  TRUE FALSE  TRUE ... FALSE  TRUE FALSE  TRUE FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Filter again but only using the first two samples</span></span></span>
<span class="r-in"><span><span class="va">filt2</span> <span class="op">&lt;-</span> <span class="fu">filterData</span><span class="op">(</span><span class="va">filt1</span><span class="op">$</span><span class="va">coverage</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="fl">5</span>, index <span class="op">=</span> <span class="va">filt1</span><span class="op">$</span><span class="va">position</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:11:43.527625 filterData: originally there were 8300 rows, now there are 6924 rows. Meaning that 16.58 percent was filtered.</span>
<span class="r-in"><span><span class="va">filt2</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $coverage</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> DataFrame with 6924 rows and 2 columns</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          x     y</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      &lt;Rle&gt; &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1        7     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        4    10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3        8     6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        6     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        8     9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ...    ...   ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6920     9     4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6921     4     8</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6922     9     2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6923     8     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6924     6     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $position</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> logical-Rle of length 10000 with 4265 runs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Lengths:     1    12     1     1     1 ...     1     1     1     1     1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Values : FALSE  TRUE FALSE  TRUE FALSE ... FALSE  TRUE FALSE  TRUE FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

