<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference • derfinder</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="analyzeChr.html">analyzeChr()</a></code> </p>
        </td>
        <td><p>Run the derfinder analysis on a chromosome</p></td>
      </tr><tr><td>
          <p><code><a href="annotateRegions.html">annotateRegions()</a></code> </p>
        </td>
        <td><p>Assign genomic states to regions</p></td>
      </tr><tr><td>
          <p><code><a href="calculatePvalues.html">calculatePvalues()</a></code> </p>
        </td>
        <td><p>Calculate p-values and identify regions</p></td>
      </tr><tr><td>
          <p><code><a href="calculateStats.html">calculateStats()</a></code> </p>
        </td>
        <td><p>Calculate F-statistics at base pair resolution from a loaded BAM files</p></td>
      </tr><tr><td>
          <p><code><a href="coerceGR.html">coerceGR()</a></code> </p>
        </td>
        <td><p>Coerce the coverage to a GRanges object for a given sample</p></td>
      </tr><tr><td>
          <p><code><a href="collapseFullCoverage.html">collapseFullCoverage()</a></code> </p>
        </td>
        <td><p>Collapse full coverage information for efficient quantile computations</p></td>
      </tr><tr><td>
          <p><code><a href="coverageToExon.html">coverageToExon()</a></code> </p>
        </td>
        <td><p>Extract coverage information for exons</p></td>
      </tr><tr><td>
          <p><code><a href="createBw.html">createBw()</a></code> </p>
        </td>
        <td><p>Export coverage to BigWig files</p></td>
      </tr><tr><td>
          <p><code><a href="createBwSample.html">createBwSample()</a></code> </p>
        </td>
        <td><p>Create a BigWig file with the coverage information for a given sample</p></td>
      </tr><tr><td>
          <p><code><a href="define_cluster.html">define_cluster()</a></code> </p>
        </td>
        <td><p>Define a cluster to use.</p></td>
      </tr><tr><td>
          <p><code><a href="derfinder-deprecated.html">advancedArg()</a></code> </p>
        </td>
        <td><p>Deprecated functions in package ‘derfinder’</p></td>
      </tr><tr><td>
          <p><code><a href="extendedMapSeqlevels.html">extendedMapSeqlevels()</a></code> </p>
        </td>
        <td><p>Change naming style for a set of sequence names</p></td>
      </tr><tr><td>
          <p><code><a href="filterData.html">filterData()</a></code> </p>
        </td>
        <td><p>Filter the positions of interest</p></td>
      </tr><tr><td>
          <p><code><a href="findRegions.html">findRegions()</a></code> </p>
        </td>
        <td><p>Find non-zero regions in a Rle</p></td>
      </tr><tr><td>
          <p><code><a href="fullCoverage.html">fullCoverage()</a></code> </p>
        </td>
        <td><p>Load the unfiltered coverage information from a group of BAM files and a
list of chromosomes</p></td>
      </tr><tr><td>
          <p><code><a href="genomeData.html">genomeData</a></code> </p>
        </td>
        <td><p>Genome samples processed data</p></td>
      </tr><tr><td>
          <p><code><a href="genomeDataRaw.html">genomeDataRaw</a></code> </p>
        </td>
        <td><p>Genome samples processed data</p></td>
      </tr><tr><td>
          <p><code><a href="genomeFstats.html">genomeFstats</a></code> </p>
        </td>
        <td><p>F-statistics for the example data</p></td>
      </tr><tr><td>
          <p><code><a href="genomeInfo.html">genomeInfo</a></code> </p>
        </td>
        <td><p>Genome samples information</p></td>
      </tr><tr><td>
          <p><code><a href="genomeRegions.html">genomeRegions</a></code> </p>
        </td>
        <td><p>Candidate DERs for example data</p></td>
      </tr><tr><td>
          <p><code><a href="genomicState.html">genomicState</a></code> </p>
        </td>
        <td><p>Genomic State for Hsapiens.UCSC.hg19.knownGene</p></td>
      </tr><tr><td>
          <p><code><a href="getRegionCoverage.html">getRegionCoverage()</a></code> </p>
        </td>
        <td><p>Extract coverage information for a set of regions</p></td>
      </tr><tr><td>
          <p><code><a href="getTotalMapped.html">getTotalMapped()</a></code> </p>
        </td>
        <td><p>Calculate the total number of mapped reads</p></td>
      </tr><tr><td>
          <p><code><a href="loadCoverage.html">loadCoverage()</a></code> </p>
        </td>
        <td><p>Load the coverage information from a group of BAM files</p></td>
      </tr><tr><td>
          <p><code><a href="makeGenomicState.html">makeGenomicState()</a></code> </p>
        </td>
        <td><p>Obtain the genomic state per region from annotation</p></td>
      </tr><tr><td>
          <p><code><a href="makeModels.html">makeModels()</a></code> </p>
        </td>
        <td><p>Build model matrices for differential expression</p></td>
      </tr><tr><td>
          <p><code><a href="mergeResults.html">mergeResults()</a></code> </p>
        </td>
        <td><p>Merge results from different chromosomes</p></td>
      </tr><tr><td>
          <p><code><a href="preprocessCoverage.html">preprocessCoverage()</a></code> </p>
        </td>
        <td><p>Transform and split the data</p></td>
      </tr><tr><td>
          <p><code><a href="railMatrix.html">railMatrix()</a></code> </p>
        </td>
        <td><p>Identify regions data by a coverage filter and get a count matrix from
BigWig files</p></td>
      </tr><tr><td>
          <p><code><a href="rawFiles.html">rawFiles()</a></code> </p>
        </td>
        <td><p>Construct full paths to a group of raw input files</p></td>
      </tr><tr><td>
          <p><code><a href="regionMatrix.html">regionMatrix()</a></code> </p>
        </td>
        <td><p>Identify regions data by a coverage filter and get a count matrix</p></td>
      </tr><tr><td>
          <p><code><a href="sampleDepth.html">sampleDepth()</a></code> </p>
        </td>
        <td><p>Calculate adjustments for library size</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

