<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Load the coverage information from a group of BAM files — loadCoverage • derfinder</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Load the coverage information from a group of BAM files — loadCoverage"><meta property="og:description" content="For a group of samples this function reads the coverage information for a
specific chromosome directly from the BAM files. It then merges them into a
DataFrame and removes the bases that do not pass the cutoff."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load the coverage information from a group of BAM files</h1>
    <small class="dont-index">Source: <a href="https://github.com/lcolladotor/derfinder/blob/HEAD/R/loadCoverage.R" class="external-link"><code>R/loadCoverage.R</code></a></small>
    <div class="hidden name"><code>loadCoverage.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>For a group of samples this function reads the coverage information for a
specific chromosome directly from the BAM files. It then merges them into a
DataFrame and removes the bases that do not pass the cutoff.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">loadCoverage</span><span class="op">(</span></span>
<span>  <span class="va">files</span>,</span>
<span>  <span class="va">chr</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter <span class="op">=</span> <span class="st">"one"</span>,</span>
<span>  chrlen <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bai <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>files</dt>
<dd><p>A character vector with the full path to the sample BAM files
(or BigWig files).
The names are used for the column names of the DataFrame. Check
<a href="rawFiles.html">rawFiles</a> for constructing <code>files</code>. <code>files</code> can also be a
<a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFileList</a>, <a href="https://rdrr.io/pkg/Rsamtools/man/BamFile-class.html" class="external-link">BamFile</a>,
<a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html" class="external-link">BigWigFileList</a>, or <a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html" class="external-link">BigWigFile</a> object.</p></dd>


<dt>chr</dt>
<dd><p>Chromosome to read. Should be in the format matching the one used
in the raw data.</p></dd>


<dt>cutoff</dt>
<dd><p>This argument is passed to <a href="filterData.html">filterData</a>.</p></dd>


<dt>filter</dt>
<dd><p>Has to be either <code>'one'</code> (default) or <code>'mean'</code>. In
the first case, at least one sample has to have coverage above <code>cutoff</code>.
In the second case, the mean coverage has to be greater than <code>cutoff</code>.</p></dd>


<dt>chrlen</dt>
<dd><p>The chromosome length in base pairs. If it's <code>NULL</code>, the
chromosome length is extracted from the BAM files.</p></dd>


<dt>output</dt>
<dd><p>If <code>NULL</code> then no output is saved in disk. If <code>auto</code>
then an automatic name is constructed using UCSC names (chrXCovInfo.Rdata
for example). If another character is specified, then that name is used for
the output file.</p></dd>


<dt>bai</dt>
<dd><p>The full path to the BAM index files. If <code>NULL</code> it is
assumed that the BAM index files are in the same location as the BAM files
and that they have the .bai extension. Ignored if <code>files</code> is a
<code>BamFileList</code> object or if <code>inputType=='BigWig'</code>.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to other methods and/or advanced arguments.
Advanced arguments:</p><dl><dt>verbose </dt>
<dd><p>If <code>TRUE</code> basic status updates will be printed along
the way.</p></dd>

<dt>inputType </dt>
<dd><p>Has to be either <code>bam</code> or <code>BigWig</code>. It specifies
the format of the raw data files. By default it's set to <code>bam</code> before
trying to guess it from the file names.</p></dd>

<dt>tilewidth </dt>
<dd><p>When specified, <a href="https://rdrr.io/pkg/GenomicRanges/man/tileGenome.html" class="external-link">tileGenome</a> is used
to break up the chromosome into chunks. We don't recommend this for BAM
files as the coverage in the borders of the chunks might be slightly off.</p></dd>

<dt>which </dt>
<dd><p><code>NULL</code> by default. When a <code>GRanges</code> is specified,
this specific region of the genome is loaded instead of the full chromosome.</p></dd>

<dt>fileStyle </dt>
<dd><p>The naming style of the chromosomes in the input files.
If the global option <code>chrsStyle</code> is not set, the naming style won't be
changed. This option is useful when you want to use a specific naming style
but the raw files use another style.</p></dd>

<dt>protectWhich </dt>
<dd><p>When not <code>NULL</code> and <code>which</code> is specified,
this argument specifies by how much the ranges in <code>which</code> will be
expanded. This helps get the same base level coverage you would get from
reading the coverage for a full chromosome from BAM files. Otherwise some
reads might be excluded and thus the base level coverage will be lower.
<code>NULL</code> by default.</p></dd>

<dt>drop.D </dt>
<dd><p>Whether to drop the bases with 'D' in the CIGAR strings
or to include them. Only used with BAM files. <code>FALSE</code> by default.</p></dd>

<dt>sampleNames </dt>
<dd><p>Column names to be used the samples. By default it's
<code>names(files)</code>.</p></dd>


</dl><p>Passed to <a href="extendedMapSeqlevels.html">extendedMapSeqlevels</a>, <a href="define_cluster.html">define_cluster</a>,
<a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html" class="external-link">scanBamFlag</a> and <a href="filterData.html">filterData</a>.
Note that <a href="filterData.html">filterData</a> is used internally
by loadCoverage and has the important arguments <code>totalMapped</code>
and <code>targetSize</code> which can be used to normalize the coverage by
library size. See <a href="getTotalMapped.html">getTotalMapped</a> for calculating <code>totalMapped</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list with two components.</p><dl><dt>coverage </dt>
<dd><p>is a DataFrame object where each column represents a
sample. The number of rows depends on the number of base pairs that passed
the cutoff and the information stored is the coverage at that given base.</p></dd>

<dt>position </dt>
<dd><p>is a logical Rle with the positions of the chromosome
that passed the cutoff.</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>The <code>...</code> argument can be used to control which alignments to consider
when reading from BAM files. See <a href="https://rdrr.io/pkg/Rsamtools/man/ScanBamParam-class.html" class="external-link">scanBamFlag</a>.</p>
<p>Parallelization for loading the data in chunks is used only used when
<code>tilewidth</code> is specified. You may use up to one core per tile.</p>
<p>If you set the advanced argument <code>drop.D = TRUE</code>, bases with CIGAR
string "D" (deletion from reference) will be excluded from the base-level
coverage calculation.</p>
<p>If you are working with data from an organism different from 'Homo sapiens'
specify so by setting the global 'species' and 'chrsStyle' options. For
example:
<code>options(species = 'arabidopsis_thaliana')</code>
<code>options(chrsStyle = 'NCBI')</code></p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="fullCoverage.html">fullCoverage</a>, <a href="getTotalMapped.html">getTotalMapped</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Leonardo Collado-Torres, Andrew Jaffe</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">datadir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"genomeData"</span>, package <span class="op">=</span> <span class="st">"derfinder"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="rawFiles.html">rawFiles</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    datadir <span class="op">=</span> <span class="va">datadir</span>, samplepatt <span class="op">=</span> <span class="st">"*accepted_hits.bam$"</span>,</span></span>
<span class="r-in"><span>    fileterm <span class="op">=</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## Shorten the column names</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_accepted_hits.bam"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Read and filter the data, only for 2 files</span></span></span>
<span class="r-in"><span><span class="va">dataSmall</span> <span class="op">&lt;-</span> <span class="fu">loadCoverage</span><span class="op">(</span>files <span class="op">=</span> <span class="va">files</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, chr <span class="op">=</span> <span class="st">"21"</span>, cutoff <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:15:47.30742 loadCoverage: finding chromosome lengths</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:15:47.324577 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009101_accepted_hits.bam</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:15:47.392495 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009102_accepted_hits.bam</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:15:47.457829 loadCoverage: applying the cutoff to the merged data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:15:47.485599 filterData: originally there were 48129895 rows, now there are 454 rows. Meaning that 100 percent was filtered.</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co">## Export to BigWig files</span></span></span>
<span class="r-in"><span><span class="fu"><a href="createBw.html">createBw</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"chr21"</span> <span class="op">=</span> <span class="va">dataSmall</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Load data from BigWig files</span></span></span>
<span class="r-in"><span><span class="va">dataSmall.bw</span> <span class="op">&lt;-</span> <span class="fu">loadCoverage</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    ERR009101 <span class="op">=</span> <span class="st">"ERR009101.bw"</span>, ERR009102 <span class="op">=</span></span></span>
<span class="r-in"><span>        <span class="st">"ERR009102.bw"</span></span></span>
<span class="r-in"><span><span class="op">)</span>, chr <span class="op">=</span> <span class="st">"chr21"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Compare them</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="va">x</span> <span class="op">-</span> <span class="va">y</span></span></span>
<span class="r-in"><span><span class="op">}</span>, <span class="va">dataSmall</span><span class="op">$</span><span class="va">coverage</span>, <span class="va">dataSmall.bw</span><span class="op">$</span><span class="va">coverage</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Note that the only difference is the type of Rle (integer vs numeric) used</span></span></span>
<span class="r-in"><span><span class="co">## to store the data.</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

