<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Obtain the genomic state per region from annotation — makeGenomicState • derfinder</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Obtain the genomic state per region from annotation — makeGenomicState"><meta property="og:description" content="This function summarizes the annotation contained in a
TxDb at each given base of the genome based
on annotated transcripts. It groups contiguous base pairs classified as the
same type into regions."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Obtain the genomic state per region from annotation</h1>
    <small class="dont-index">Source: <a href="https://github.com/lcolladotor/derfinder/blob/HEAD/R/makeGenomicState.R" class="external-link"><code>R/makeGenomicState.R</code></a></small>
    <div class="hidden name"><code>makeGenomicState.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function summarizes the annotation contained in a
<a href="https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html" class="external-link">TxDb</a> at each given base of the genome based
on annotated transcripts. It groups contiguous base pairs classified as the
same type into regions.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">makeGenomicState</span><span class="op">(</span><span class="va">txdb</span>, chrs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">22</span><span class="op">)</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>txdb</dt>
<dd><p>A <a href="https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html" class="external-link">TxDb</a> object with chromosome lengths
(check <code>seqlengths(txdb)</code>). If you are using a
<a href="https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html" class="external-link">TxDb</a> object created from a GFF/GTF file, you will
find this <a href="https://support.bioconductor.org/p/93235/" class="external-link">https://support.bioconductor.org/p/93235/</a> useful.</p></dd>


<dt>chrs</dt>
<dd><p>The names of the chromosomes to use as denoted in the
<code>txdb</code> object. Check <a href="https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html" class="external-link">isActiveSeq</a>.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to <a href="extendedMapSeqlevels.html">extendedMapSeqlevels</a>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A <code>GRangesList</code> object with two elements: <code>fullGenome</code> and
<code>codingGenome</code>. Both have metadata information for the type of region
(theRegion), transcript IDs (tx_id), transcript name (tx_name), and gene ID
(gene_id). <code>fullGenome</code> classifies each region as either being exon,
intron or intergenic. <code>codingGenome</code> classfies the regions as being
promoter, exon, intro, 5UTR, 3UTR or intergenic.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html" class="external-link">TxDb</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Andrew Jaffe, Leonardo Collado-Torres</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Load the example data base from the GenomicFeatures vignette</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">"GenomicFeatures"</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: AnnotationDbi</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: Biobase</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Welcome to Bioconductor</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Vignettes contain introductory material; view with</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="r-in"><span><span class="va">samplefile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"hg19_knownGene_sample.sqlite"</span>,</span></span>
<span class="r-in"><span>    package <span class="op">=</span> <span class="st">"GenomicFeatures"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="va">samplefile</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Generate genomic state object, only for chr6</span></span></span>
<span class="r-in"><span><span class="va">sampleGenomicState</span> <span class="op">&lt;-</span> <span class="fu">makeGenomicState</span><span class="op">(</span><span class="va">txdb</span>, chrs <span class="op">=</span> <span class="st">"chr6"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:1 mapping between keys and columns</span>
<span class="r-in"><span><span class="co">#</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co">## Create the GenomicState object for Hsapiens.UCSC.hg19.knownGene</span></span></span>
<span class="r-in"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/TxDb.Hsapiens.UCSC.hg19.knownGene/man/package.html" class="external-link">TxDb.Hsapiens.UCSC.hg19.knownGene</a></span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Creating this GenomicState object takes around 8 min for all chrs and</span></span></span>
<span class="r-in"><span><span class="co">## around 30 secs for chr21</span></span></span>
<span class="r-in"><span><span class="va">GenomicState.Hsapiens.UCSC.hg19.knownGene.chr21</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>    <span class="fu">makeGenomicState</span><span class="op">(</span>txdb <span class="op">=</span> <span class="va">txdb</span>, chrs <span class="op">=</span> <span class="st">"chr21"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## For convinience, this object is already included in derfinder</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://testthat.r-lib.org" class="external-link">"testthat"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://testthat.r-lib.org/reference/expect_that.html" class="external-link">expect_that</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">GenomicState.Hsapiens.UCSC.hg19.knownGene.chr21</span>,</span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://testthat.r-lib.org/reference/oldskool.html" class="external-link">is_equivalent_to</a></span><span class="op">(</span><span class="va">genomicState</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Hsapiens ENSEMBL GRCh37</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicFeatures" class="external-link">"GenomicFeatures"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">## Can take several minutes and speed will depend on your internet speed</span></span></span>
<span class="r-in"><span><span class="va">xx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbPackage.html" class="external-link">makeTxDbPackageFromBiomart</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    version <span class="op">=</span> <span class="st">"0.99"</span>, maintainer <span class="op">=</span> <span class="st">"Your Name"</span>,</span></span>
<span class="r-in"><span>    author <span class="op">=</span> <span class="st">"Your Name"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">loadDb</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"TxDb.Hsapiens.BioMart.ensembl.GRCh37.p11"</span>, <span class="st">"inst"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"extdata"</span>, <span class="st">"TxDb.Hsapiens.BioMart.ensembl.GRCh37.p11.sqlite"</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Creating this GenomicState object takes around 13 min</span></span></span>
<span class="r-in"><span><span class="va">GenomicState.Hsapiens.ensembl.GRCh37.p11</span> <span class="op">&lt;-</span> <span class="fu">makeGenomicState</span><span class="op">(</span></span></span>
<span class="r-in"><span>    txdb <span class="op">=</span> <span class="va">txdb</span>,</span></span>
<span class="r-in"><span>    chrs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Save for later use</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">GenomicState.Hsapiens.ensembl.GRCh37.p11</span>,</span></span>
<span class="r-in"><span>    file <span class="op">=</span> <span class="st">"GenomicState.Hsapiens.ensembl.GRCh37.p11.Rdata"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

