<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Identify regions data by a coverage filter and get a count matrix from
BigWig files — railMatrix • derfinder</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Identify regions data by a coverage filter and get a count matrix from
BigWig files — railMatrix"><meta property="og:description" content="Rail (available at http://rail.bio) generates coverage BigWig files. These
files are faster to load in R and to process. Rail creates an un-adjusted
coverage BigWig file per sample and an adjusted summary coverage BigWig file
by chromosome (median or mean). railMatrix reads in the mean (or
median) coverage BigWig file and applies a threshold cutoff to identify
expressed regions (ERs). Then it goes back to the sample coverage BigWig
files and extracts the base level coverage for each sample. Finally it
summarizes this information in a matrix with one row per ERs and one column
per sample. This function is similar to regionMatrix but is faster
thanks to the advantages presented by BigWig files."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Identify regions data by a coverage filter and get a count matrix from
BigWig files</h1>
    <small class="dont-index">Source: <a href="https://github.com/lcolladotor/derfinder/blob/HEAD/R/railMatrix.R" class="external-link"><code>R/railMatrix.R</code></a></small>
    <div class="hidden name"><code>railMatrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Rail (available at http://rail.bio) generates coverage BigWig files. These
files are faster to load in R and to process. Rail creates an un-adjusted
coverage BigWig file per sample and an adjusted summary coverage BigWig file
by chromosome (median or mean). railMatrix reads in the mean (or
median) coverage BigWig file and applies a threshold cutoff to identify
expressed regions (ERs). Then it goes back to the sample coverage BigWig
files and extracts the base level coverage for each sample. Finally it
summarizes this information in a matrix with one row per ERs and one column
per sample. This function is similar to <a href="regionMatrix.html">regionMatrix</a> but is faster
thanks to the advantages presented by BigWig files.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">railMatrix</span><span class="op">(</span></span>
<span>  <span class="va">chrs</span>,</span>
<span>  <span class="va">summaryFiles</span>,</span>
<span>  <span class="va">sampleFiles</span>,</span>
<span>  <span class="va">L</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxClusterGap <span class="op">=</span> <span class="fl">300L</span>,</span>
<span>  totalMapped <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  targetSize <span class="op">=</span> <span class="fl">4e+07</span>,</span>
<span>  file.cores <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  chrlens <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>chrs</dt>
<dd><p>A character vector with the names of the chromosomes.</p></dd>


<dt>summaryFiles</dt>
<dd><p>A character vector (or BigWigFileList) with the paths to
the summary BigWig files created by Rail. Either mean or median files. These
are library size adjusted by default in Rail. The order of the files in this
vector must match the order in <code>chrs</code>.</p></dd>


<dt>sampleFiles</dt>
<dd><p>A character vector with the paths to the BigWig files
by sample. These files are unadjusted for library size.</p></dd>


<dt>L</dt>
<dd><p>The width of the reads used. Either a vector of length 1 or length
equal to the number of samples.</p></dd>


<dt>cutoff</dt>
<dd><p>The base-pair level cutoff to use. It's behavior is controlled
by <code>filter</code>.</p></dd>


<dt>maxClusterGap</dt>
<dd><p>This determines the maximum gap between candidate ERs.</p></dd>


<dt>totalMapped</dt>
<dd><p>A vector with the total number of reads mapped for each
sample. The vector should be in the same order as the samples in <code>data</code>.
Providing this data adjusts the coverage to reads in <code>targetSize</code>
library prior to filtering. See <a href="getTotalMapped.html">getTotalMapped</a> for
calculating this vector.</p></dd>


<dt>targetSize</dt>
<dd><p>The target library size to adjust the coverage to. Used
only when <code>totalMapped</code> is specified. By default, it adjusts to
libraries with 40 million reads, which matches the default used in Rail.</p></dd>


<dt>file.cores</dt>
<dd><p>Number of cores used for loading the BigWig files per chr.</p></dd>


<dt>chrlens</dt>
<dd><p>The chromosome lengths in base pairs. If it's <code>NULL</code>,
the chromosome length is extracted from the BAM files. Otherwise, it should
have the same length as <code>chrs</code>.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to other methods and/or advanced arguments.
Advanced arguments:</p><dl><dt>verbose </dt>
<dd><p>If <code>TRUE</code> basic status updates will be printed along
the way. Default: <code>TRUE</code>.</p></dd>

<dt>verbose.load </dt>
<dd><p>If <code>TRUE</code> basic status updates will be printed
along the way when loading data. Default: <code>TRUE</code>.</p></dd>

<dt>BPPARAM.railChr </dt>
<dd><p>A BPPARAM object to use for the chr step. Set to
<a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a> when <code>file.cores = 1</code> and
<a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a> otherwise.</p></dd>

<dt>chunksize </dt>
<dd><p>Chunksize to use. Default: 1000.</p></dd>


</dl><p>Passed to <a href="filterData.html">filterData</a>, <a href="findRegions.html">findRegions</a> and <a href="define_cluster.html">define_cluster</a>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list with one entry per chromosome. Then per chromosome, a list
with two components.</p><dl><dt>regions </dt>
<dd><p>A set of regions based on the coverage filter cutoff as
returned by <a href="findRegions.html">findRegions</a>.</p></dd>

<dt>coverageMatrix </dt>
<dd><p>A matrix with the mean coverage by sample for each
candidate region.</p></dd>


</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Given a set of un-filtered coverage data (see <a href="fullCoverage.html">fullCoverage</a>), create
candidate regions by applying a cutoff on the coverage values,
and obtain a count matrix where the number of rows corresponds to the number
of candidate regions and the number of columns corresponds to the number of
samples. The values are the mean coverage for a given sample for a given
region.</p>
<p>This function uses several other <a href="derfinder-package.html">derfinder-package</a>
functions. Inspect the code if interested.</p>
<p>You should use at most one core per chromosome.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Leonardo Collado-Torres</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## BigWig files are not supported in Windows</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="va">.Platform</span><span class="op">$</span><span class="va">OS.type</span> <span class="op">!=</span> <span class="st">"windows"</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="co">## Get data</span></span></span>
<span class="r-in"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/leekgroup/derfinderData" class="external-link">"derfinderData"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co">## Identify sample files</span></span></span>
<span class="r-in"><span>    <span class="va">sampleFiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="rawFiles.html">rawFiles</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"AMY"</span>,</span></span>
<span class="r-in"><span>        package <span class="op">=</span></span></span>
<span class="r-in"><span>            <span class="st">"derfinderData"</span></span></span>
<span class="r-in"><span>    <span class="op">)</span>, samplepatt <span class="op">=</span> <span class="st">"bw"</span>, fileterm <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sampleFiles</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".bw"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sampleFiles</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co">## Create the mean bigwig file. This file is normally created by Rail</span></span></span>
<span class="r-in"><span>    <span class="co">## but in this example we'll create it manually.</span></span></span>
<span class="r-in"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">"GenomicRanges"</a></span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">fullCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="fullCoverage.html">fullCoverage</a></span><span class="op">(</span>files <span class="op">=</span> <span class="va">sampleFiles</span>, chrs <span class="op">=</span> <span class="st">"chr21"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">meanCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">fullCov</span><span class="op">$</span><span class="va">chr21</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">fullCov</span><span class="op">$</span><span class="va">chr21</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="createBw.html">createBw</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"chr21"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span><span class="st">"meanChr21"</span> <span class="op">=</span> <span class="va">meanCov</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>        keepGR <span class="op">=</span></span></span>
<span class="r-in"><span>            <span class="cn">FALSE</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="va">summaryFile</span> <span class="op">&lt;-</span> <span class="st">"meanChr21.bw"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co">## Get the regions</span></span></span>
<span class="r-in"><span>    <span class="va">regionMat</span> <span class="op">&lt;-</span> <span class="fu">railMatrix</span><span class="op">(</span></span></span>
<span class="r-in"><span>        chrs <span class="op">=</span> <span class="st">"chr21"</span>, summaryFiles <span class="op">=</span> <span class="va">summaryFile</span>,</span></span>
<span class="r-in"><span>        sampleFiles <span class="op">=</span> <span class="va">sampleFiles</span>, L <span class="op">=</span> <span class="fl">76</span>, cutoff <span class="op">=</span> <span class="fl">5.1</span>,</span></span>
<span class="r-in"><span>        maxClusterGap <span class="op">=</span> <span class="fl">3000L</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>    <span class="co">## Explore results</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">coverageMatrix</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:10.260162 fullCoverage: processing chromosome chr21</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:10.265168 loadCoverage: finding chromosome lengths</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:10.274939 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB113.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:10.435157 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB123.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:10.622388 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB126.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:10.739908 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB130.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:10.876353 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB135.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:10.993499 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB136.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:11.13026 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB145.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:11.267248 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB153.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:11.416524 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB159.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:11.532903 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB178.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:11.654058 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB92.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:11.797876 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB97.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:11.947208 loadCoverage: applying the cutoff to the merged data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:11.973882 filterData: originally there were 48129895 rows, now there are 48129895 rows. Meaning that 0 percent was filtered.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:13.011339 coerceGR: coercing sample meanChr21</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:13.110892 createBwSample: exporting bw for sample meanChr21</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:14.042934 loadCoverage: finding chromosome lengths</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:14.050446 loadCoverage: loading BigWig file meanChr21.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:14.416429 loadCoverage: applying the cutoff to the merged data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:16.465184 filterData: originally there were 48129895 rows, now there are 48129895 rows. Meaning that 0 percent was filtered.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:16.671673 filterData: originally there were 48129895 rows, now there are 15927 rows. Meaning that 99.97 percent was filtered.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:16.67505 findRegions: identifying potential segments</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:16.678528 findRegions: segmenting information</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:16.678853 .getSegmentsRle: segmenting with cutoff(s) 5.1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:16.685313 findRegions: identifying candidate regions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:16.724271 findRegions: identifying region clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:16.811908 railMatrix: processing regions 1 to 204</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:16.817193 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB113.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:16.906503 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB123.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:16.993726 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB126.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:17.077456 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB130.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:17.16439 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB135.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:17.25015 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB136.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:17.337518 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB145.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:17.426829 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB153.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:17.516243 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB159.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:17.605176 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB178.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:17.695337 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB92.bw</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 03:06:17.824402 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB97.bw</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 204  12</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

