<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Identify regions data by a coverage filter and get a count matrix — regionMatrix • derfinder</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Identify regions data by a coverage filter and get a count matrix — regionMatrix" />
<meta property="og:description" content="Given a set of un-filtered coverage data (see fullCoverage), create
candidate regions by applying a cutoff on the coverage values,
and obtain a count matrix where the number of rows corresponds to the number
of candidate regions and the number of columns corresponds to the number of
samples. The values are the mean coverage for a given sample for a given
region." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.21.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/lcolladotor/derfinder/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Identify regions data by a coverage filter and get a count matrix</h1>
    <small class="dont-index">Source: <a href='https://github.com/lcolladotor/derfinder/blob/master/R/regionMatrix.R'><code>R/regionMatrix.R</code></a></small>
    <div class="hidden name"><code>regionMatrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given a set of un-filtered coverage data (see <a href='fullCoverage.html'>fullCoverage</a>), create
candidate regions by applying a cutoff on the coverage values,
and obtain a count matrix where the number of rows corresponds to the number
of candidate regions and the number of columns corresponds to the number of
samples. The values are the mean coverage for a given sample for a given
region.</p>
    </div>

    <pre class="usage"><span class='fu'>regionMatrix</span>(
  <span class='no'>fullCov</span>,
  <span class='kw'>cutoff</span> <span class='kw'>=</span> <span class='fl'>5</span>,
  <span class='no'>L</span>,
  <span class='kw'>totalMapped</span> <span class='kw'>=</span> <span class='fl'>8e+07</span>,
  <span class='kw'>targetSize</span> <span class='kw'>=</span> <span class='fl'>8e+07</span>,
  <span class='kw'>runFilter</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>returnBP</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>fullCov</th>
      <td><p>A list where each element is the result from
<a href='loadCoverage.html'>loadCoverage</a> used with <code>returnCoverage = TRUE</code>. Can be generated
using <a href='fullCoverage.html'>fullCoverage</a>. If <code>runFilter = FALSE</code>, then
<code>returnMean = TRUE</code> must have been used.</p></td>
    </tr>
    <tr>
      <th>cutoff</th>
      <td><p>The base-pair level cutoff to use. It's behavior is controlled
by <code>filter</code>.</p></td>
    </tr>
    <tr>
      <th>L</th>
      <td><p>The width of the reads used. Either a vector of length 1 or length
equal to the number of samples.</p></td>
    </tr>
    <tr>
      <th>totalMapped</th>
      <td><p>A vector with the total number of reads mapped for each
sample. The vector should be in the same order as the samples in
<code>fullCov</code>. Providing this argument adjusts the coverage to reads in
<code>targetSize</code> library prior to filtering. See <a href='getTotalMapped.html'>getTotalMapped</a> for
calculating this vector.</p></td>
    </tr>
    <tr>
      <th>targetSize</th>
      <td><p>The target library size to adjust the coverage to. Used
only when <code>totalMapped</code> is specified. By default, it adjusts to
libraries with 80 million reads.</p></td>
    </tr>
    <tr>
      <th>runFilter</th>
      <td><p>This controls whether to run <a href='filterData.html'>filterData</a> or not. If
set to <code>FALSE</code> then <code>returnMean = TRUE</code> must have been used to
create each element of <code>fullCov</code> and the data must have been
normalized (<code>totalMapped</code> equal to <code>targetSize</code>).</p></td>
    </tr>
    <tr>
      <th>returnBP</th>
      <td><p>If <code>TRUE</code>, returns <code>$bpCoverage</code> explained below.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Arguments passed to other methods and/or advanced arguments.
Advanced arguments:</p><dl'>
<dt>verbose </dt><dd><p>If <code>TRUE</code> basic status updates will be printed along
the way.</p></dd>
<dt>chrsStyle </dt><dd><p>Default: <code>UCSC</code>. Passed to
<a href='extendedMapSeqlevels.html'>extendedMapSeqlevels</a> via <a href='getRegionCoverage.html'>getRegionCoverage</a>.</p></dd>
<dt>species </dt><dd><p>Default: <code>homo_sapiens</code>. Passed to
<a href='extendedMapSeqlevels.html'>extendedMapSeqlevels</a> via <a href='getRegionCoverage.html'>getRegionCoverage</a>.</p></dd>
<dt>currentStyle </dt><dd><p>Default: <code>NULL</code>. Passed to
<a href='extendedMapSeqlevels.html'>extendedMapSeqlevels</a> via <a href='getRegionCoverage.html'>getRegionCoverage</a>.</p></dd>

</dl><p>Passed to <a href='filterData.html'>filterData</a>, <a href='findRegions.html'>findRegions</a> and <a href='define_cluster.html'>define_cluster</a>.</p>
<p>Note that <a href='filterData.html'>filterData</a> is used internally
by <a href='loadCoverage.html'>loadCoverage</a> (and hence <code>fullCoverage</code>) and has the important
arguments <code>totalMapped</code> and <code>targetSize</code> which can be used to
normalize the coverage by library size. If you already used these arguments
when creating the <code>fullCov</code> object, then don't specify them a second
time in regionMatrix. If you have not used these arguments, we
recommend using them to normalize the mean coverage.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with one entry per chromosome. Then per chromosome, a list
with three components.</p><dl'>
<dt>regions </dt><dd><p>A set of regions based on the coverage filter cutoff as
returned by <a href='findRegions.html'>findRegions</a>.</p></dd>
<dt>bpCoverage </dt><dd><p>A list with one element per region. Each element is a matrix with numbers of rows equal to the number of base pairs in the region and number of columns equal to the number of samples. It contains the base-level coverage information for the regions. Only returned when <code>returnBP = TRUE</code>.</p></dd>
<dt>coverageMatrix </dt><dd><p>A matrix with the mean coverage by sample for each
candidate region.</p></dd>

</dl>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function uses several other <a href='derfinder-package.html'>derfinder-package</a>
functions. Inspect the code if interested.</p>
<p>You should use at most one core per chromosome.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Create some toy data</span>
<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='st'>"IRanges"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>Rle</span>(<span class='fu'><a href='https://rdrr.io/r/base/Round.html'>round</a></span>(<span class='fu'><a href='https://rdrr.io/r/stats/Uniform.html'>runif</a></span>(<span class='fl'>1e4</span>, <span class='kw'>max</span> <span class='kw'>=</span> <span class='fl'>10</span>)))
<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'>Rle</span>(<span class='fu'><a href='https://rdrr.io/r/base/Round.html'>round</a></span>(<span class='fu'><a href='https://rdrr.io/r/stats/Uniform.html'>runif</a></span>(<span class='fl'>1e4</span>, <span class='kw'>max</span> <span class='kw'>=</span> <span class='fl'>10</span>)))
<span class='no'>z</span> <span class='kw'>&lt;-</span> <span class='fu'>Rle</span>(<span class='fu'><a href='https://rdrr.io/r/base/Round.html'>round</a></span>(<span class='fu'><a href='https://rdrr.io/r/stats/Uniform.html'>runif</a></span>(<span class='fl'>1e4</span>, <span class='kw'>max</span> <span class='kw'>=</span> <span class='fl'>10</span>)))
<span class='no'>fullCov</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='st'>"chr21"</span> <span class='kw'>=</span> <span class='fu'>DataFrame</span>(<span class='no'>x</span>, <span class='no'>y</span>, <span class='no'>z</span>))

<span class='co'>## Calculate a proxy of library size</span>
<span class='no'>libSize</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span>(<span class='no'>fullCov</span>$<span class='no'>chr21</span>, <span class='no'>sum</span>)

<span class='co'>## Run region matrix normalizing the coverage</span>
<span class='no'>regionMat</span> <span class='kw'>&lt;-</span> <span class='fu'>regionMatrix</span>(
    <span class='kw'>fullCov</span> <span class='kw'>=</span> <span class='no'>fullCov</span>, <span class='kw'>maxRegionGap</span> <span class='kw'>=</span> <span class='fl'>10L</span>,
    <span class='kw'>maxClusterGap</span> <span class='kw'>=</span> <span class='fl'>300L</span>, <span class='kw'>L</span> <span class='kw'>=</span> <span class='fl'>36</span>, <span class='kw'>totalMapped</span> <span class='kw'>=</span> <span class='no'>libSize</span>, <span class='kw'>targetSize</span> <span class='kw'>=</span> <span class='fl'>4e4</span>
)</div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 regionMatrix: processing chr21</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 filterData: normalizing coverage</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 filterData: done normalizing coverage</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 filterData: originally there were 10000 rows, now there are 2586 rows. Meaning that 74.14 percent was filtered.</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 findRegions: identifying potential segments</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 findRegions: segmenting information</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 findRegions: identifying candidate regions</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 findRegions: identifying region clusters</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 getRegionCoverage: processing chr21</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 getRegionCoverage: done processing chr21</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 regionMatrix: calculating coverageMatrix</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 regionMatrix: adjusting coverageMatrix for 'L'</span></div><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='co'>## You can alternatively use filterData() on fullCov to reduce the required</span>
<span class='co'>## memory before using regionMatrix(). This can be useful when mc.cores &gt; 1</span>
<span class='no'>filteredCov</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span>(<span class='no'>fullCov</span>, <span class='no'>filterData</span>,
    <span class='kw'>returnMean</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>filter</span> <span class='kw'>=</span> <span class='st'>"mean"</span>,
    <span class='kw'>cutoff</span> <span class='kw'>=</span> <span class='fl'>5</span>, <span class='kw'>totalMapped</span> <span class='kw'>=</span> <span class='no'>libSize</span>, <span class='kw'>targetSize</span> <span class='kw'>=</span> <span class='fl'>4e4</span>
)
<span class='no'>regionMat2</span> <span class='kw'>&lt;-</span> <span class='fu'>regionMatrix</span>(<span class='no'>filteredCov</span>,
    <span class='kw'>maxRegionGap</span> <span class='kw'>=</span> <span class='fl'>10L</span>,
    <span class='kw'>maxClusterGap</span> <span class='kw'>=</span> <span class='fl'>300L</span>, <span class='kw'>L</span> <span class='kw'>=</span> <span class='fl'>36</span>, <span class='kw'>runFilter</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)
}

<span class='co'>## regionMatrix() can work with multiple chrs as shown below.</span>
<span class='no'>fullCov2</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='st'>"chr21"</span> <span class='kw'>=</span> <span class='fu'>DataFrame</span>(<span class='no'>x</span>, <span class='no'>y</span>, <span class='no'>z</span>), <span class='st'>"chr22"</span> <span class='kw'>=</span> <span class='fu'>DataFrame</span>(<span class='no'>x</span>, <span class='no'>y</span>, <span class='no'>z</span>))
<span class='no'>regionMat2</span> <span class='kw'>&lt;-</span> <span class='fu'>regionMatrix</span>(
    <span class='kw'>fullCov</span> <span class='kw'>=</span> <span class='no'>fullCov2</span>, <span class='kw'>maxRegionGap</span> <span class='kw'>=</span> <span class='fl'>10L</span>,
    <span class='kw'>maxClusterGap</span> <span class='kw'>=</span> <span class='fl'>300L</span>, <span class='kw'>L</span> <span class='kw'>=</span> <span class='fl'>36</span>, <span class='kw'>totalMapped</span> <span class='kw'>=</span> <span class='no'>libSize</span>, <span class='kw'>targetSize</span> <span class='kw'>=</span> <span class='fl'>4e4</span>
)</div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 regionMatrix: processing chr21</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 filterData: normalizing coverage</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 filterData: done normalizing coverage</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 filterData: originally there were 10000 rows, now there are 2586 rows. Meaning that 74.14 percent was filtered.</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 findRegions: identifying potential segments</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 findRegions: segmenting information</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 findRegions: identifying candidate regions</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:29 findRegions: identifying region clusters</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 getRegionCoverage: processing chr21</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 getRegionCoverage: done processing chr21</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 regionMatrix: calculating coverageMatrix</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 regionMatrix: adjusting coverageMatrix for 'L'</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 regionMatrix: processing chr22</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 filterData: normalizing coverage</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 filterData: done normalizing coverage</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 filterData: originally there were 10000 rows, now there are 2586 rows. Meaning that 74.14 percent was filtered.</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 findRegions: identifying potential segments</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 findRegions: segmenting information</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 findRegions: identifying candidate regions</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 findRegions: identifying region clusters</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 getRegionCoverage: processing chr22</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 getRegionCoverage: done processing chr22</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 regionMatrix: calculating coverageMatrix</span></div><div class='output co'>#&gt; <span class='message'>2020-04-19 20:55:30 regionMatrix: adjusting coverageMatrix for 'L'</span></div><div class='input'>
<span class='co'>## Combine results from multiple chromosomes</span>
<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='st'>"GenomicRanges"</span>)

<span class='co'>## First extract the data</span>
<span class='no'>regs</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/unlist.html'>unlist</a></span>(<span class='fu'><a href='https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html'>GRangesList</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span>(<span class='no'>regionMat2</span>, <span class='st'>"[["</span>, <span class='st'>"regions"</span>)))
<span class='no'>covMat</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/do.call.html'>do.call</a></span>(<span class='no'>rbind</span>, <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span>(<span class='no'>regionMat2</span>, <span class='st'>"[["</span>, <span class='st'>"coverageMatrix"</span>))
<span class='no'>covBp</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/do.call.html'>do.call</a></span>(<span class='no'>c</span>, <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span>(<span class='no'>regionMat2</span>, <span class='st'>"[["</span>, <span class='st'>"bpCoverage"</span>))
<span class='co'>## Force the names to match</span>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span>(<span class='no'>regs</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>covMat</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span>(<span class='no'>covBp</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq_len</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>regs</span>))
<span class='co'>## Combine into a list (not really needed)</span>
<span class='no'>mergedRegionMat</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(
    <span class='st'>"regions"</span> <span class='kw'>=</span> <span class='no'>regs</span>, <span class='st'>"coverageMatrix"</span> <span class='kw'>=</span> <span class='no'>covMat</span>,
    <span class='st'>"bpCoverage"</span> <span class='kw'>=</span> <span class='no'>covBp</span>
)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


