<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate adjustments for library size — sampleDepth • derfinder</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate adjustments for library size — sampleDepth"><meta property="og:description" content="For a given data set calculate the per-sample coverage adjustments. Hector
Corrada's group proposed calculating the sum of the coverage for genes below
a given sample quantile. In this function, we calculate the sample quantiles
of interest by sample, and then the sum of the coverage for bases below or
equal to quantiles of interest. The resulting values are transformed log2(x

scalefac)
 to avoid very large numbers that could potentially affect the
stability of the F-statistics calculation. The sample coverage adjustments
are then used in makeModels for constructing the null and alternative
models."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate adjustments for library size</h1>
    <small class="dont-index">Source: <a href="https://github.com/lcolladotor/derfinder/blob/HEAD/R/sampleDepth.R" class="external-link"><code>R/sampleDepth.R</code></a></small>
    <div class="hidden name"><code>sampleDepth.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>For a given data set calculate the per-sample coverage adjustments. Hector
Corrada's group proposed calculating the sum of the coverage for genes below
a given sample quantile. In this function, we calculate the sample quantiles
of interest by sample, and then the sum of the coverage for bases below or
equal to quantiles of interest. The resulting values are transformed log2(x
</p><ul><li><p>scalefac)</p></li>
</ul> to avoid very large numbers that could potentially affect the
stability of the F-statistics calculation. The sample coverage adjustments
are then used in <a href="makeModels.html">makeModels</a> for constructing the null and alternative
models.
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sampleDepth</span><span class="op">(</span><span class="va">collapsedFull</span>, probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span>, scalefac <span class="op">=</span> <span class="fl">32</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>collapsedFull</dt>
<dd><p>The full coverage data collapsed by sample as produced
by <a href="collapseFullCoverage.html">collapseFullCoverage</a>.</p></dd>


<dt>probs</dt>
<dd><p>Number(s) between 0 and 1 representing the quantile(s) of
interest. For example, 0.5 is the median.</p></dd>


<dt>scalefac</dt>
<dd><p>Number added to the sample coverage adjustments before the
log2 transformation.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed to other methods and/or advanced arguments.
Advanced arguments:</p><dl><dt>verbose </dt>
<dd><p>If <code>TRUE</code> basic status updates will be printed along
the way.</p></dd>

<dt>nonzero </dt>
<dd><p>If <code>TRUE</code> only the nonzero counts are used to
calculate the library size adjustment. Default: <code>TRUE</code>.</p></dd>

<dt>center </dt>
<dd><p>If <code>TRUE</code> the sample coverage adjustements are
centered. In some cases, this could be helpful for interpretation purposes.
Default: <code>FALSE</code>.</p></dd>


</dl></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A matrix (vector of <code>length(probs) == 1</code>) with the library size depth
adjustments per sample to be used in <a href="makeModels.html">makeModels</a>. The number of rows
corresponds to the number of quantiles used for the sample adjustments.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Paulson, J. N., Stine, O. C., Bravo, H. C. &amp; Pop, M. Differential abundance
analysis for microbial marker-gene surveys. Nat. Methods (2013).
doi:10.1038/nmeth.2658</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><a href="collapseFullCoverage.html">collapseFullCoverage</a>, <a href="makeModels.html">makeModels</a></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Leonardo Collado-Torres</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">## Collapse the coverage information</span></span></span>
<span class="r-in"><span><span class="va">collapsedFull</span> <span class="op">&lt;-</span> <span class="fu"><a href="collapseFullCoverage.html">collapseFullCoverage</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">genomeData</span><span class="op">$</span><span class="va">coverage</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    verbose <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:12:02.513079 collapseFullCoverage: Sorting fullCov</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:12:02.515498 collapseFullCoverage: Collapsing chromosomes information by sample</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Calculate library size adjustments</span></span></span>
<span class="r-in"><span><span class="va">sampleDepths</span> <span class="op">&lt;-</span> <span class="fu">sampleDepth</span><span class="op">(</span><span class="va">collapsedFull</span>, probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:12:02.517499 sampleDepth: Calculating sample quantiles</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2023-05-07 05:12:02.523817 sampleDepth: Calculating sample adjustments</span>
<span class="r-in"><span><span class="va">sampleDepths</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      ERR009101 ERR009102 ERR009105 ERR009107 ERR009108 ERR009112 ERR009115</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  50%  8.366322  7.954196  9.379378  8.596190  8.523562  8.179909  6.129283</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 100%  8.614710  7.977280 10.939579  9.194757  8.894818  8.321928  6.129283</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      ERR009116 ERR009131 ERR009138 ERR009144 ERR009145 ERR009148 ERR009151</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  50%  8.233620  7.266787  8.400879  8.044394   6.72792  7.366322  9.134426</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 100%  8.611025  7.491853  8.748193  8.888743   6.72792  8.475733  9.845490</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      ERR009152 ERR009153 ERR009159 ERR009161 ERR009163 ERR009164 ERR009167</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  50%  8.044394  8.262095  8.991522  7.000000   6.72792  7.285402  9.214319</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 100%  9.103288  8.741467  9.957102  8.475733   6.72792  7.754888 11.016112</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      SRR031812 SRR031835 SRR031867 SRR031868 SRR031900 SRR031904 SRR031914</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  50%  6.066089  6.285402         5         5         5  6.285402         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 100%  6.066089  6.285402         5         5         5  6.285402         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      SRR031936 SRR031958 SRR031960</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  50%         5         5         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 100%         5         5         5</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

