<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>derfinder users guide • derfinder</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="derfinder users guide">
<meta property="og:description" content="derfinder">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">derfinder</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.35.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/derfinder-quickstart.html">derfinder quick start guide</a>
    </li>
    <li>
      <a href="../articles/derfinder-users-guide.html">derfinder users guide</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lcolladotor/derfinder/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>derfinder users guide</h1>
                        <h4 data-toc-skip class="author">Leonardo
Collado-Torres</h4>
            <address class="author_afil">
      Lieber Institute for Brain Development, Johns Hopkins Medical
CampusCenter for Computational Biology, Johns Hopkins
University<br><a class="author_email" href="mailto:#"></a><a href="mailto:lcolladotor@gmail.com" class="email">lcolladotor@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">7 May 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/lcolladotor/derfinder/blob/HEAD/vignettes/derfinder-users-guide.Rmd" class="external-link"><code>vignettes/derfinder-users-guide.Rmd</code></a></small>
      <div class="hidden name"><code>derfinder-users-guide.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="asking-for-help">Asking for help<a class="anchor" aria-label="anchor" href="#asking-for-help"></a>
</h2>
<p>Please read the <a href="http://bioconductor.org/packages/release/bioc/vignettes/derfinder/inst/doc/derfinder-quickstart.html#basics" class="external-link">basics
of using <code>derfinder</code></a> in the quick start guide. Thank
you.</p>
</div>
<div class="section level2">
<h2 id="derfinder-users-guide">
<em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
users guide<a class="anchor" aria-label="anchor" href="#derfinder-users-guide"></a>
</h2>
<p>If you haven’t already, please read the <a href="quickstart-derfinder.html">quick start to using derfinder</a>
vignette. It explains the basics of using <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>,
how to ask for help, and showcases an example analysis.</p>
<p>The <em>derfinder users guide</em> goes into more depth about what
you can do with <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>.
It covers the two implementations of the DER Finder approach (Frazee,
Sabunciyan, Hansen, Irizarry, and Leek, 2014). That is, the (A)
expressed regions-level and (B) single base-level F-statistics
implementations. The vignette also includes advanced material for fine
tuning some options, working with non-human data, and example scripts
for a high-performance computing cluster.</p>
</div>
<div class="section level2">
<h2 id="expressed-regions-level">Expressed regions-level<a class="anchor" aria-label="anchor" href="#expressed-regions-level"></a>
</h2>
<p>The expressed regions-level implementation is based on summarizing
the coverage information for all your samples and applying a cutoff to
that summary. For example, by calculating the mean coverage at every
base and then checking if it’s greater than some cutoff value of
interest. Contiguous bases passing the cutoff become a candidate
Expressed Region (ER). We can then construct a matrix summarizing the
base-level coverage for each sample for the set of ERs. This matrix can
then be using with packages developed for feature counting (<em><a href="https://bioconductor.org/packages/3.17/limma" class="external-link">limma</a></em>,
<em><a href="https://bioconductor.org/packages/3.17/DESeq2" class="external-link">DESeq2</a></em>,
<em><a href="https://bioconductor.org/packages/3.17/edgeR" class="external-link">edgeR</a></em>, etc)
to determine which ERs have differential expression signal. That is, to
identify the Differentially Expressed Regions (DERs).</p>
<div class="section level3">
<h3 id="regionmatrix">
<code>regionMatrix()</code><a class="anchor" aria-label="anchor" href="#regionmatrix"></a>
</h3>
<p>Commonly, users have aligned their raw RNA-seq data and saved the
alignments in BAM files. Some might have chosen to compress this
information into BigWig files. BigWig files are much faster to load than
BAM files and are the type of file we prefer to work with. However, we
can still identify the expressed regions from BAM files.</p>
<p>The function <code><a href="../reference/regionMatrix.html">regionMatrix()</a></code> will require you to load the
data (either from BAM or BigWig files) and store it all in memory. It
will then calculate the mean coverage across all samples, apply the
cutoff you chose, and determine the expressed regions.</p>
<p>This is the path you will want to follow in most scenarios.</p>
</div>
<div class="section level3">
<h3 id="railmatrix">
<code>railMatrix()</code><a class="anchor" aria-label="anchor" href="#railmatrix"></a>
</h3>
<p><a href="http://rail.bio" class="external-link"><img src="https://github.com/nellore/rail/blob/master/assets/railrnalogodark.png?raw=true" alt="Rail logo"></a></p>
<p>Our favorite method for identifying expressed regions is based on
pre-computed summary coverage files (mean or median) as well as coverage
files by sample. <a href="http://rail.bio" class="external-link">Rail</a> is a cloud-enabled
aligner that will generate:</p>
<ol style="list-style-type: decimal">
<li>Per chromosome, a summary (mean or median) coverage BigWig file.
Note that the summary is adjusted for library size and is by default set
to libraries with 40 million reads.</li>
<li>Per sample, a coverage BigWig file with the un-adjusted
coverage.</li>
</ol>
<p><a href="http://rail.bio" class="external-link">Rail</a> does a great job in creating these
files for us, saving time and reducing the memory load needed for this
type of analysis with <code>R</code>.</p>
<p>If you have this type of data or can generate it from BAM files with
other tools, you will be interested in using the
<code><a href="../reference/railMatrix.html">railMatrix()</a></code> function. The output is identical to the one
from <code><a href="../reference/regionMatrix.html">regionMatrix()</a></code>. It’s just much faster and memory
efficient. The only drawback is that BigWig files are not fully
supported in Windows as of <em><a href="https://bioconductor.org/packages/3.17/rtracklayer" class="external-link">rtracklayer</a></em>
version 1.25.16.</p>
<p>We highly recommend this approach. <a href="http://rail.bio" class="external-link">Rail</a>
has also other significant features such as: scalability, reduced
redundancy, integrative analysis, mode agnosticism, and inexpensive
cloud implementation. For more information, visit <a href="http://rail.bio" class="external-link">rail.bio</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="single-base-level-f-statistics">Single base-level F-statistics<a class="anchor" aria-label="anchor" href="#single-base-level-f-statistics"></a>
</h2>
<p>The DER Finder approach was originally implemented by calculating
t-statistics between two groups and using a hidden markov model to
determine the expression states: not expressed, expressed,
differentially expressed (Frazee, Sabunciyan, Hansen et al., 2014). The
original software works but had areas where we worked to improve it,
which lead to the single base-level F-statistics implementation. Also
note that the original software is no longer maintained.</p>
<p>This type of analysis first loads the data and preprocess it in a
format that saves time and reduces memory later. It then fits two nested
models (an alternative and a null model) with the coverage information
for every single base-pair of the genome. Using the two fitted models,
it calculates an F-statistic. Basically, it generates a vector along the
genome with F-statistics.</p>
<p>A cutoff is then applied to the F-statistics and contiguous
base-pairs of the genome passing the cutoff are considered a candidate
Differentially Expressed Region (DER).</p>
<p>Calculating F-statistics along the genome is computationally
intensive, but doable. The major resource drain comes from assigning
p-values to the DERs. To do so, we permute the model matrices and
re-calculate the F-statistics generating a set of null DERs. Once we
have enough null DERs, we compare the observed DERs against the null
DERs to calculate p-values for the observed DERs.</p>
<p>This type of analysis at the chromosome level is done by the
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> function, which is a high level function using
many other pieces of <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>.
Once all chromosomes have been processed, <code><a href="../reference/mergeResults.html">mergeResults()</a></code>
combines them.</p>
<p>Which implementation of the DER Finder approach you will want to use
depends on your specific use case and computational resources available.
But in general, we recommend starting with the expressed regions-level
implementation.</p>
</div>
<div class="section level2">
<h2 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h2>
<p>In this vignette we we will analyze a small subset of the samples
from the <em>BrainSpan Atlas of the Human Brain</em> (BrainSpan, 2011)
publicly available data.</p>
<p>We first load the required packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/lcolladotor/derfinder" class="external-link">"derfinder"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/leekgroup/derfinderData" class="external-link">"derfinderData"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">"GenomicRanges"</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="phenotype-data">Phenotype data<a class="anchor" aria-label="anchor" href="#phenotype-data"></a>
</h3>
<p>For this example, we created a small table with the relevant
phenotype data for 12 samples: 6 from fetal samples and 6 from adult
samples. We chose at random a brain region, in this case the amygdaloid
complex. For this example we will only look at data from chromosome 21.
Other variables include the age in years and the gender. The data is
shown below.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://yihui.org/knitr/" class="external-link">"knitr"</a></span><span class="op">)</span></span>
<span><span class="co">## Get pheno table</span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">brainspanPheno</span>, <span class="va">structure_acronym</span> <span class="op">==</span> <span class="st">"AMY"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Display the main information</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">pheno</span><span class="op">[</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"structure_acronym"</span>,</span>
<span>    <span class="st">"structure_name"</span>, <span class="st">"file"</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">p</span>, format <span class="op">=</span> <span class="st">"html"</span>, row.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
gender
</th>
<th style="text-align:left;">
lab
</th>
<th style="text-align:right;">
Age
</th>
<th style="text-align:left;">
group
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
HSB97.AMY
</td>
<td style="text-align:right;">
-0.4423077
</td>
<td style="text-align:left;">
fetal
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
HSB92.AMY
</td>
<td style="text-align:right;">
-0.3653846
</td>
<td style="text-align:left;">
fetal
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
HSB178.AMY
</td>
<td style="text-align:right;">
-0.4615385
</td>
<td style="text-align:left;">
fetal
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
HSB159.AMY
</td>
<td style="text-align:right;">
-0.3076923
</td>
<td style="text-align:left;">
fetal
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
HSB153.AMY
</td>
<td style="text-align:right;">
-0.5384615
</td>
<td style="text-align:left;">
fetal
</td>
</tr>
<tr>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
HSB113.AMY
</td>
<td style="text-align:right;">
-0.5384615
</td>
<td style="text-align:left;">
fetal
</td>
</tr>
<tr>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
HSB130.AMY
</td>
<td style="text-align:right;">
21.0000000
</td>
<td style="text-align:left;">
adult
</td>
</tr>
<tr>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
HSB136.AMY
</td>
<td style="text-align:right;">
23.0000000
</td>
<td style="text-align:left;">
adult
</td>
</tr>
<tr>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
HSB126.AMY
</td>
<td style="text-align:right;">
30.0000000
</td>
<td style="text-align:left;">
adult
</td>
</tr>
<tr>
<td style="text-align:left;">
10
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
HSB145.AMY
</td>
<td style="text-align:right;">
36.0000000
</td>
<td style="text-align:left;">
adult
</td>
</tr>
<tr>
<td style="text-align:left;">
11
</td>
<td style="text-align:left;">
M
</td>
<td style="text-align:left;">
HSB123.AMY
</td>
<td style="text-align:right;">
37.0000000
</td>
<td style="text-align:left;">
adult
</td>
</tr>
<tr>
<td style="text-align:left;">
12
</td>
<td style="text-align:left;">
F
</td>
<td style="text-align:left;">
HSB135.AMY
</td>
<td style="text-align:right;">
40.0000000
</td>
<td style="text-align:left;">
adult
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="load-the-data">Load the data<a class="anchor" aria-label="anchor" href="#load-the-data"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
offers three functions related to loading raw data. The first one,
<code><a href="../reference/rawFiles.html">rawFiles()</a></code>, is a helper function for identifying the full
paths to the input files. Next, <code><a href="../reference/loadCoverage.html">loadCoverage()</a></code> loads the
base-level coverage data from either BAM or BigWig files for a specific
chromosome. Finally, <code><a href="../reference/fullCoverage.html">fullCoverage()</a></code> will load the coverage
for a set of chromosomes using <code><a href="../reference/loadCoverage.html">loadCoverage()</a></code>.</p>
<p>We can load the data from <em><a href="https://bioconductor.org/packages/3.17/derfinderData" class="external-link">derfinderData</a></em>
(Collado-Torres, Jaffe, and Leek, 2023) by first identifying the paths
to the BigWig files with <code><a href="../reference/rawFiles.html">rawFiles()</a></code> and then loading the
data with <code><a href="../reference/fullCoverage.html">fullCoverage()</a></code>. Note that the <em>BrainSpan</em>
data is already normalized by the total number of mapped reads in each
sample. However, that won’t be the case with most data sets in which
case you might want to use the <code>totalMapped</code> and
<code>targetSize</code> arguments. The function
<code><a href="../reference/getTotalMapped.html">getTotalMapped()</a></code> will be helpful to get this
information.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Determine the files to use and fix the names</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rawFiles.html">rawFiles</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"AMY"</span>, package <span class="op">=</span> <span class="st">"derfinderData"</span><span class="op">)</span>,</span>
<span>    samplepatt <span class="op">=</span> <span class="st">"bw"</span>, fileterm <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".bw"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load the data from disk</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">fullCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fullCoverage.html">fullCoverage</a></span><span class="op">(</span></span>
<span>    files <span class="op">=</span> <span class="va">files</span>, chrs <span class="op">=</span> <span class="st">"chr21"</span>,</span>
<span>    totalMapped <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span><span class="op">)</span>, targetSize <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2023-05-07 03:54:56.460636 fullCoverage: processing chromosome chr21</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:56.481253 loadCoverage: finding chromosome lengths</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:56.519525 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB113.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:56.780866 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB123.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:57.045148 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB126.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:57.173166 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB130.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:57.326525 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB135.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:57.454771 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB136.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:57.607898 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB145.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:57.746853 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB153.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:57.904444 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB159.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:58.04743 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB178.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:58.185444 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB92.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:58.357274 loadCoverage: loading BigWig file /__w/_temp/Library/derfinderData/extdata/AMY/HSB97.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:58.532032 loadCoverage: applying the cutoff to the merged data</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:58.552644 filterData: normalizing coverage</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:59.798006 filterData: done normalizing coverage</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:54:59.81619 filterData: originally there were 48129895 rows, now there are 48129895 rows. Meaning that 0 percent was filtered.</span></span></code></pre>
<pre><code><span><span class="co">##    user  system elapsed </span></span>
<span><span class="co">##   3.369   0.056   3.434</span></span></code></pre>
<p>Alternatively, since the BigWig files are publicly available from
<em>BrainSpan</em> (see <a href="http://download.alleninstitute.org/brainspan/MRF_BigWig_Gencode_v10/bigwig/" class="external-link">here</a>),
we can extract the relevant coverage data using
<code><a href="../reference/fullCoverage.html">fullCoverage()</a></code>. Note that as of <em><a href="https://bioconductor.org/packages/3.17/rtracklayer" class="external-link">rtracklayer</a></em>
1.25.16 BigWig files are not supported on Windows: you can find the
<code>fullCov</code> object inside <em><a href="https://bioconductor.org/packages/3.17/derfinderData" class="external-link">derfinderData</a></em>
to follow the examples.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Determine the files to use and fix the names</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="va">pheno</span><span class="op">$</span><span class="va">file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".AMY"</span>, <span class="st">""</span>, <span class="va">pheno</span><span class="op">$</span><span class="va">lab</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load the data from the web</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">fullCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fullCoverage.html">fullCoverage</a></span><span class="op">(</span></span>
<span>    files <span class="op">=</span> <span class="va">files</span>, chrs <span class="op">=</span> <span class="st">"chr21"</span>,</span>
<span>    totalMapped <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span><span class="op">)</span>, targetSize <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Note how loading the coverage for 12 samples from the web was quite
fast. Although in this case we only retained the information for
chromosome 21.</p>
<p>The result of <code>fullCov</code> is a list with one element per
chromosome. If no filtering was performed, each chromosome has a
<code>DataFrame</code> with the number of rows equaling the number of
bases in the chromosome with one column per sample.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Lets explore it</span></span>
<span><span class="va">fullCov</span></span></code></pre></div>
<pre><code><span><span class="co">## $chr21</span></span>
<span><span class="co">## DataFrame with 48129895 rows and 12 columns</span></span>
<span><span class="co">##          HSB113 HSB123 HSB126 HSB130 HSB135 HSB136 HSB145 HSB153 HSB159 HSB178</span></span>
<span><span class="co">##           &lt;Rle&gt;  &lt;Rle&gt;  &lt;Rle&gt;  &lt;Rle&gt;  &lt;Rle&gt;  &lt;Rle&gt;  &lt;Rle&gt;  &lt;Rle&gt;  &lt;Rle&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">## 1             0      0      0      0      0      0      0      0      0      0</span></span>
<span><span class="co">## 2             0      0      0      0      0      0      0      0      0      0</span></span>
<span><span class="co">## 3             0      0      0      0      0      0      0      0      0      0</span></span>
<span><span class="co">## 4             0      0      0      0      0      0      0      0      0      0</span></span>
<span><span class="co">## 5             0      0      0      0      0      0      0      0      0      0</span></span>
<span><span class="co">## ...         ...    ...    ...    ...    ...    ...    ...    ...    ...    ...</span></span>
<span><span class="co">## 48129891      0      0      0      0      0      0      0      0      0      0</span></span>
<span><span class="co">## 48129892      0      0      0      0      0      0      0      0      0      0</span></span>
<span><span class="co">## 48129893      0      0      0      0      0      0      0      0      0      0</span></span>
<span><span class="co">## 48129894      0      0      0      0      0      0      0      0      0      0</span></span>
<span><span class="co">## 48129895      0      0      0      0      0      0      0      0      0      0</span></span>
<span><span class="co">##          HSB92 HSB97</span></span>
<span><span class="co">##          &lt;Rle&gt; &lt;Rle&gt;</span></span>
<span><span class="co">## 1            0     0</span></span>
<span><span class="co">## 2            0     0</span></span>
<span><span class="co">## 3            0     0</span></span>
<span><span class="co">## 4            0     0</span></span>
<span><span class="co">## 5            0     0</span></span>
<span><span class="co">## ...        ...   ...</span></span>
<span><span class="co">## 48129891     0     0</span></span>
<span><span class="co">## 48129892     0     0</span></span>
<span><span class="co">## 48129893     0     0</span></span>
<span><span class="co">## 48129894     0     0</span></span>
<span><span class="co">## 48129895     0     0</span></span></code></pre>
<p>If filtering was performed, each chromosome also has a logical
<code>Rle</code> indicating which bases of the chromosome passed the
filtering. This information is useful later on to map back the results
to the genome coordinates.</p>
</div>
<div class="section level3">
<h3 id="filter-coverage">Filter coverage<a class="anchor" aria-label="anchor" href="#filter-coverage"></a>
</h3>
<p>Depending on the use case, you might want to filter the base-level
coverage at the time of reading it, or you might want to keep an
unfiltered version. By default both <code><a href="../reference/loadCoverage.html">loadCoverage()</a></code> and
<code><a href="../reference/fullCoverage.html">fullCoverage()</a></code> will not filter.</p>
<p>If you decide to filter, set the <code>cutoff</code> argument to a
positive value. This will run <code><a href="../reference/filterData.html">filterData()</a></code>. Note that you
might want to standardize the library sizes prior to filtering if you
didn’t already do it when creating the <code>fullCov</code> object. You
can do so by by supplying the <code>totalMapped</code> and
<code>targetSize</code> arguments to <code><a href="../reference/filterData.html">filterData()</a></code>. Note:
don’t use these arguments twice in <code><a href="../reference/fullCoverage.html">fullCoverage()</a></code> and
<code><a href="../reference/filterData.html">filterData()</a></code>.</p>
<p>In this example, we prefer to keep both an unfiltered and filtered
version. For the filtered version, we will retain the bases where at
least one sample has coverage greater than 2.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Filter coverage</span></span>
<span><span class="va">filteredCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fullCov</span>, <span class="va">filterData</span>, cutoff <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2023-05-07 03:55:00.561219 filterData: originally there were 48129895 rows, now there are 130356 rows. Meaning that 99.73 percent was filtered.</span></span></code></pre>
<p>The result is similar to <code>fullCov</code> but with the genomic
position index as shown below.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Similar to fullCov but with $position</span></span>
<span><span class="va">filteredCov</span></span></code></pre></div>
<pre><code><span><span class="co">## $chr21</span></span>
<span><span class="co">## $chr21$coverage</span></span>
<span><span class="co">## DataFrame with 130356 rows and 12 columns</span></span>
<span><span class="co">##                  HSB113           HSB123             HSB126             HSB130</span></span>
<span><span class="co">##                   &lt;Rle&gt;            &lt;Rle&gt;              &lt;Rle&gt;              &lt;Rle&gt;</span></span>
<span><span class="co">## 1      2.00999999046326                0                  0 0.0399999991059303</span></span>
<span><span class="co">## 2      2.17000007629395                0                  0 0.0399999991059303</span></span>
<span><span class="co">## 3      2.21000003814697                0                  0 0.0399999991059303</span></span>
<span><span class="co">## 4      2.36999988555908                0                  0 0.0399999991059303</span></span>
<span><span class="co">## 5      2.36999988555908                0 0.0599999986588955 0.0399999991059303</span></span>
<span><span class="co">## ...                 ...              ...                ...                ...</span></span>
<span><span class="co">## 130352             1.25 1.27999997138977   2.04999995231628  0.790000021457672</span></span>
<span><span class="co">## 130353 1.21000003814697 1.24000000953674   2.04999995231628  0.790000021457672</span></span>
<span><span class="co">## 130354 1.21000003814697 1.20000004768372   1.92999994754791  0.790000021457672</span></span>
<span><span class="co">## 130355 1.16999995708466 1.20000004768372   1.92999994754791  0.790000021457672</span></span>
<span><span class="co">## 130356 1.16999995708466  1.0900000333786   1.87000000476837  0.790000021457672</span></span>
<span><span class="co">##                   HSB135            HSB136            HSB145            HSB153</span></span>
<span><span class="co">##                    &lt;Rle&gt;             &lt;Rle&gt;             &lt;Rle&gt;             &lt;Rle&gt;</span></span>
<span><span class="co">## 1      0.230000004172325 0.129999995231628                 0 0.589999973773956</span></span>
<span><span class="co">## 2      0.230000004172325 0.129999995231628                 0 0.629999995231628</span></span>
<span><span class="co">## 3      0.230000004172325 0.129999995231628                 0 0.670000016689301</span></span>
<span><span class="co">## 4      0.230000004172325 0.129999995231628                 0 0.709999978542328</span></span>
<span><span class="co">## 5      0.230000004172325 0.129999995231628                 0              0.75</span></span>
<span><span class="co">## ...                  ...               ...               ...               ...</span></span>
<span><span class="co">## 130352  1.62999999523163  1.37000000476837  1.02999997138977  2.21000003814697</span></span>
<span><span class="co">## 130353  1.62999999523163  1.37000000476837  1.02999997138977  2.21000003814697</span></span>
<span><span class="co">## 130354  1.62999999523163  1.27999997138977 0.730000019073486  2.00999999046326</span></span>
<span><span class="co">## 130355  1.62999999523163  1.27999997138977 0.600000023841858  2.00999999046326</span></span>
<span><span class="co">## 130356  1.54999995231628  1.01999998092651 0.560000002384186  1.92999994754791</span></span>
<span><span class="co">##                   HSB159             HSB178              HSB92            HSB97</span></span>
<span><span class="co">##                    &lt;Rle&gt;              &lt;Rle&gt;              &lt;Rle&gt;            &lt;Rle&gt;</span></span>
<span><span class="co">## 1      0.150000005960464 0.0500000007450581 0.0700000002980232 1.58000004291534</span></span>
<span><span class="co">## 2      0.150000005960464 0.0500000007450581 0.0700000002980232 1.58000004291534</span></span>
<span><span class="co">## 3      0.150000005960464 0.0500000007450581  0.100000001490116 1.61000001430511</span></span>
<span><span class="co">## 4      0.150000005960464 0.0500000007450581  0.129999995231628 1.64999997615814</span></span>
<span><span class="co">## 5                   0.25  0.100000001490116  0.129999995231628 1.67999994754791</span></span>
<span><span class="co">## ...                  ...                ...                ...              ...</span></span>
<span><span class="co">## 130352  2.46000003814697    2.0699999332428   2.23000001907349 1.50999999046326</span></span>
<span><span class="co">## 130353  2.46000003814697    2.0699999332428   2.23000001907349 1.50999999046326</span></span>
<span><span class="co">## 130354  2.21000003814697   2.10999989509583   2.13000011444092 1.47000002861023</span></span>
<span><span class="co">## 130355  2.10999989509583   2.10999989509583   2.13000011444092 1.47000002861023</span></span>
<span><span class="co">## 130356  1.96000003814697   1.77999997138977   2.05999994277954 1.47000002861023</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $chr21$position</span></span>
<span><span class="co">## logical-Rle of length 48129895 with 3235 runs</span></span>
<span><span class="co">##   Lengths: 9825448     149       2       9 ...     172     740     598   45091</span></span>
<span><span class="co">##   Values :   FALSE    TRUE   FALSE    TRUE ...    TRUE   FALSE    TRUE   FALSE</span></span></code></pre>
<p>In terms of memory, the filtered version requires less resources.
Although this will depend on how rich the data set is and how aggressive
was the filtering step.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Compare the size in Mb</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    fullCov <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">fullCov</span><span class="op">)</span>,</span>
<span>    filteredCov <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="va">filteredCov</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">/</span> <span class="fl">1024</span><span class="op">^</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     fullCov filteredCov </span></span>
<span><span class="co">##        22.7         8.5</span></span></code></pre>
<p>Note that with your own data, filtering for bases where at least one
sample has coverage greater than 2 might not make sense: maybe you need
a higher or lower filter. The amount of bases remaining after filtering
will impact how long the analysis will take to complete. We thus
recommend exploring this before proceeding.</p>
</div>
</div>
<div class="section level2">
<h2 id="expressed-region-level-analysis">Expressed region-level analysis<a class="anchor" aria-label="anchor" href="#expressed-region-level-analysis"></a>
</h2>
<div class="section level3">
<h3 id="via-regionmatrix">Via <code>regionMatrix()</code><a class="anchor" aria-label="anchor" href="#via-regionmatrix"></a>
</h3>
<p>Now that we have the data, we can identify expressed regions (ERs) by
using a cutoff of 30 on the base-level mean coverage from these 12
samples. Once the regions have been identified, we can calculate a
coverage matrix with one row per ER and one column per sample (12 in
this case). For doing this calculation we need to know the length of the
sequence reads, which in this study were 76 bp long.</p>
<p>Note that for this type of analysis there is no major benefit of
filtering the data. Although it can be done if needed.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Use regionMatrix()</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">regionMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regionMatrix.html">regionMatrix</a></span><span class="op">(</span><span class="va">fullCov</span>, cutoff <span class="op">=</span> <span class="fl">30</span>, L <span class="op">=</span> <span class="fl">76</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## By using totalMapped equal to targetSize, regionMatrix() assumes that you have normalized the data already in fullCoverage(), loadCoverage() or filterData().</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:01.793646 regionMatrix: processing chr21</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:01.794034 filterData: normalizing coverage</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:02.106444 filterData: done normalizing coverage</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:03.370267 filterData: originally there were 48129895 rows, now there are 2256 rows. Meaning that 100 percent was filtered.</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:03.371542 findRegions: identifying potential segments</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:03.379901 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:03.389818 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:03.450733 findRegions: identifying region clusters</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:03.695329 getRegionCoverage: processing chr21</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:03.754335 getRegionCoverage: done processing chr21</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:03.758607 regionMatrix: calculating coverageMatrix</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:03.765978 regionMatrix: adjusting coverageMatrix for 'L'</span></span></code></pre>
<pre><code><span><span class="co">##    user  system elapsed </span></span>
<span><span class="co">##   1.917   0.068   1.985</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Explore results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "list"</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "regions"        "coverageMatrix" "bpCoverage"</span></span></code></pre>
<p><code><a href="../reference/regionMatrix.html">regionMatrix()</a></code> returns a list of elements of length
equal to the number of chromosomes analyzed. For each chromosome, there
are three pieces of output. The actual ERs are arranged in a
<code>GRanges</code> object named <code>regions</code>.</p>
<ul>
<li>
<em>regions</em> is the result from filtering with
<code><a href="../reference/filterData.html">filterData()</a></code> and then defining the regions with
<code><a href="../reference/findRegions.html">findRegions()</a></code>. Note that the metadata variable
<code>value</code> represents the mean coverage for the given region
while <code>area</code> is the sum of the base-level coverage (before
adjusting for read length) from all samples.</li>
<li>
<em>bpCoverage</em> is a list with the base-level coverage from all
the regions. This information can then be used with
<code><a href="https://rdrr.io/pkg/derfinderPlot/man/plotRegionCoverage.html" class="external-link">plotRegionCoverage()</a></code> from <em><a href="https://bioconductor.org/packages/3.17/derfinderPlot" class="external-link">derfinderPlot</a></em>.</li>
<li>
<em>coverageMatrix</em> is the matrix with one row per region and
one column per sample. The different matrices for each of the
chromosomes can then be merged prior to using <em><a href="https://bioconductor.org/packages/3.17/limma" class="external-link">limma</a></em>,
<em><a href="https://bioconductor.org/packages/3.17/DESeq2" class="external-link">DESeq2</a></em> or
other packages. Note that the counts are generally not integers, but can
easily be rounded if necessary.</li>
</ul>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## regions output</span></span>
<span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 45 ranges and 6 metadata columns:</span></span>
<span><span class="co">##      seqnames            ranges strand |     value      area indexStart</span></span>
<span><span class="co">##         &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;  &lt;integer&gt;</span></span>
<span><span class="co">##    1    chr21   9827018-9827582      * |  313.6717 177224.53          1</span></span>
<span><span class="co">##    2    chr21 15457301-15457438      * |  215.0846  29681.68        566</span></span>
<span><span class="co">##    3    chr21 20230140-20230192      * |   38.8325   2058.12        704</span></span>
<span><span class="co">##    4    chr21 20230445-20230505      * |   41.3245   2520.80        757</span></span>
<span><span class="co">##    5    chr21 27253318-27253543      * |   34.9131   7890.37        818</span></span>
<span><span class="co">##   ..      ...               ...    ... .       ...       ...        ...</span></span>
<span><span class="co">##   41    chr21 33039644-33039688      * |   34.4705 1551.1742       2180</span></span>
<span><span class="co">##   42    chr21 33040784-33040798      * |   32.1342  482.0133       2225</span></span>
<span><span class="co">##   43    chr21          33040890      * |   30.0925   30.0925       2240</span></span>
<span><span class="co">##   44    chr21 33040900-33040901      * |   30.1208   60.2417       2241</span></span>
<span><span class="co">##   45    chr21 48019401-48019414      * |   31.1489  436.0850       2243</span></span>
<span><span class="co">##       indexEnd cluster clusterL</span></span>
<span><span class="co">##      &lt;integer&gt;   &lt;Rle&gt;    &lt;Rle&gt;</span></span>
<span><span class="co">##    1       565       1      565</span></span>
<span><span class="co">##    2       703       2      138</span></span>
<span><span class="co">##    3       756       3      366</span></span>
<span><span class="co">##    4       817       3      366</span></span>
<span><span class="co">##    5      1043       4      765</span></span>
<span><span class="co">##   ..       ...     ...      ...</span></span>
<span><span class="co">##   41      2224      17       45</span></span>
<span><span class="co">##   42      2239      18      118</span></span>
<span><span class="co">##   43      2240      18      118</span></span>
<span><span class="co">##   44      2242      18      118</span></span>
<span><span class="co">##   45      2256      19       14</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Number of regions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 45</span></span></code></pre>
<p><code>bpCoverage</code> is the base-level coverage list which can
then be used for plotting.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Base-level coverage matrices for each of the regions</span></span>
<span><span class="co">## Useful for plotting</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">bpCoverage</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="va">head</span>, n <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $`1`</span></span>
<span><span class="co">##   HSB113 HSB123 HSB126 HSB130 HSB135 HSB136 HSB145 HSB153 HSB159 HSB178 HSB92</span></span>
<span><span class="co">## 1  93.20   3.32  28.22   5.62 185.17  98.34   5.88  16.71   3.52  15.71 47.40</span></span>
<span><span class="co">## 2 124.76   7.25  63.68  11.32 374.85 199.28  10.39  30.53   5.83  29.35 65.04</span></span>
<span><span class="co">##   HSB97</span></span>
<span><span class="co">## 1 36.54</span></span>
<span><span class="co">## 2 51.42</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`2`</span></span>
<span><span class="co">##     HSB113 HSB123 HSB126 HSB130 HSB135 HSB136 HSB145 HSB153 HSB159 HSB178 HSB92</span></span>
<span><span class="co">## 566  45.59   7.94  15.92  34.75 141.61 104.21  19.87  38.61   4.97   23.2 13.95</span></span>
<span><span class="co">## 567  45.59   7.94  15.92  35.15 141.64 104.30  19.87  38.65   4.97   23.2 13.95</span></span>
<span><span class="co">##     HSB97</span></span>
<span><span class="co">## 566 22.21</span></span>
<span><span class="co">## 567 22.21</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Check dimensions. First region is 565 long, second one is 138 bp long.</span></span>
<span><span class="co">## The columns match the number of samples (12 in this case).</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">bpCoverage</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="va">dim</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $`1`</span></span>
<span><span class="co">## [1] 565  12</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`2`</span></span>
<span><span class="co">## [1] 138  12</span></span></code></pre>
<p>The end result of the coverage matrix is shown below. Note that the
coverage has been adjusted for read length. Because reads might not
fully align inside a given region, the numbers are generally not
integers but can be rounded if needed.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Dimensions of the coverage matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">coverageMatrix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 45 12</span></span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Coverage for each region. This matrix can then be used with limma or other pkgs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">coverageMatrix</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         HSB113     HSB123      HSB126      HSB130      HSB135      HSB136</span></span>
<span><span class="co">## 1 3653.1093346 277.072106 1397.068687 1106.722895 8987.460401 5570.221054</span></span>
<span><span class="co">## 2  333.3740816  99.987237  463.909476  267.354342 1198.713552 1162.313418</span></span>
<span><span class="co">## 3   35.3828948  20.153553   30.725394   23.483947   16.786842   17.168947</span></span>
<span><span class="co">## 4   42.3398681  29.931579   41.094474   24.724736   32.634080   19.309606</span></span>
<span><span class="co">## 5   77.7402631 168.939342  115.059342  171.861974  180.638684   93.503158</span></span>
<span><span class="co">## 6    0.7988158   1.770263    1.473421    2.231053    1.697368    1.007895</span></span>
<span><span class="co">##        HSB145       HSB153     HSB159      HSB178       HSB92        HSB97</span></span>
<span><span class="co">## 1 1330.158818 1461.2986829 297.939342 1407.288552 1168.519079 1325.9622371</span></span>
<span><span class="co">## 2  257.114210  313.8513139  67.940131  193.695657  127.543553  200.7834228</span></span>
<span><span class="co">## 3   22.895921   52.8756585  28.145395   33.127368   23.758816   20.4623685</span></span>
<span><span class="co">## 4   33.802632   51.6146040  31.244343   33.576974   29.546183   28.2011836</span></span>
<span><span class="co">## 5   90.950526   36.3046051  78.069605   97.151316  100.085790   35.5428946</span></span>
<span><span class="co">## 6    1.171316    0.4221053   1.000132    1.139079    1.136447    0.3956579</span></span></code></pre>
<p>We can then use the coverage matrix and packages such as <em><a href="https://bioconductor.org/packages/3.17/limma" class="external-link">limma</a></em>,
<em><a href="https://bioconductor.org/packages/3.17/DESeq2" class="external-link">DESeq2</a></em> or
<em><a href="https://bioconductor.org/packages/3.17/edgeR" class="external-link">edgeR</a></em> to
identify which ERs are differentially expressed.</p>
<div class="section level4">
<h4 id="find-ders-with-deseq2">Find DERs with <em><a href="https://bioconductor.org/packages/3.17/DESeq2" class="external-link">DESeq2</a></em><a class="anchor" aria-label="anchor" href="#find-ders-with-deseq2"></a>
</h4>
<p>Here we’ll use <em><a href="https://bioconductor.org/packages/3.17/DESeq2" class="external-link">DESeq2</a></em> to
identify the DERs. To use it we need to round the coverage data.</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Required</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/mikelove/DESeq2" class="external-link">"DESeq2"</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: SummarizedExperiment</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: MatrixGenerics</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: matrixStats</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'MatrixGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">##     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">##     rowWeightedSds, rowWeightedVars</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Biobase</span></span></code></pre>
<pre><code><span><span class="co">## Welcome to Bioconductor</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Vignettes contain introductory material; view with</span></span>
<span><span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biobase'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyMissing, rowMedians</span></span></code></pre>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Round matrix</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">coverageMatrix</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Round matrix and specify design</span></span>
<span><span class="va">dse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></span><span class="op">(</span><span class="va">counts</span>, <span class="va">pheno</span>, <span class="op">~</span> <span class="va">group</span> <span class="op">+</span> <span class="va">gender</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## converting counts to integer mode</span></span></code></pre>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Perform DE analysis</span></span>
<span><span class="va">dse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dse</span>, test <span class="op">=</span> <span class="st">"LRT"</span>, reduced <span class="op">=</span> <span class="op">~</span><span class="va">gender</span>, fitType <span class="op">=</span> <span class="st">"local"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## estimating size factors</span></span></code></pre>
<pre><code><span><span class="co">## estimating dispersions</span></span></code></pre>
<pre><code><span><span class="co">## gene-wise dispersion estimates</span></span></code></pre>
<pre><code><span><span class="co">## mean-dispersion relationship</span></span></code></pre>
<pre><code><span><span class="co">## final dispersion estimates</span></span></code></pre>
<pre><code><span><span class="co">## fitting model and testing</span></span></code></pre>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract results</span></span>
<span><span class="va">deseq</span> <span class="op">&lt;-</span> <span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">deseq</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">deseq</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Explore the results</span></span>
<span><span class="va">deseq</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 45 ranges and 12 metadata columns:</span></span>
<span><span class="co">##      seqnames            ranges strand |     value      area indexStart</span></span>
<span><span class="co">##         &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;  &lt;integer&gt;</span></span>
<span><span class="co">##    1    chr21   9827018-9827582      * |  313.6717 177224.53          1</span></span>
<span><span class="co">##    2    chr21 15457301-15457438      * |  215.0846  29681.68        566</span></span>
<span><span class="co">##    3    chr21 20230140-20230192      * |   38.8325   2058.12        704</span></span>
<span><span class="co">##    4    chr21 20230445-20230505      * |   41.3245   2520.80        757</span></span>
<span><span class="co">##    5    chr21 27253318-27253543      * |   34.9131   7890.37        818</span></span>
<span><span class="co">##   ..      ...               ...    ... .       ...       ...        ...</span></span>
<span><span class="co">##   41    chr21 33039644-33039688      * |   34.4705 1551.1742       2180</span></span>
<span><span class="co">##   42    chr21 33040784-33040798      * |   32.1342  482.0133       2225</span></span>
<span><span class="co">##   43    chr21          33040890      * |   30.0925   30.0925       2240</span></span>
<span><span class="co">##   44    chr21 33040900-33040901      * |   30.1208   60.2417       2241</span></span>
<span><span class="co">##   45    chr21 48019401-48019414      * |   31.1489  436.0850       2243</span></span>
<span><span class="co">##       indexEnd cluster clusterL  baseMean log2FoldChange     lfcSE      stat</span></span>
<span><span class="co">##      &lt;integer&gt;   &lt;Rle&gt;    &lt;Rle&gt; &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##    1       565       1      565 2846.2872     -1.6903182  0.831959  0.215262</span></span>
<span><span class="co">##    2       703       2      138  451.5196     -1.1640426  0.757490  0.871126</span></span>
<span><span class="co">##    3       756       3      366   29.5781      0.0461488  0.458097  3.132082</span></span>
<span><span class="co">##    4       817       3      366   36.0603     -0.1866200  0.390920  2.225708</span></span>
<span><span class="co">##    5      1043       4      765  101.6468     -0.1387377  0.320166  3.957987</span></span>
<span><span class="co">##   ..       ...     ...      ...       ...            ...       ...       ...</span></span>
<span><span class="co">##   41      2224      17       45 20.782035      -0.642056  0.427661 0.6047814</span></span>
<span><span class="co">##   42      2239      18      118  6.410542      -0.634321  0.512262 0.5454039</span></span>
<span><span class="co">##   43      2240      18      118  0.129717      -0.859549  3.116540 0.0206273</span></span>
<span><span class="co">##   44      2242      18      118  0.702291      -0.628285  2.247378 0.5825105</span></span>
<span><span class="co">##   45      2256      19       14  5.293293      -1.694563  1.252290 5.7895910</span></span>
<span><span class="co">##         pvalue      padj</span></span>
<span><span class="co">##      &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##    1 0.6426743  0.997155</span></span>
<span><span class="co">##    2 0.3506436  0.997155</span></span>
<span><span class="co">##    3 0.0767657  0.863614</span></span>
<span><span class="co">##    4 0.1357305  0.997155</span></span>
<span><span class="co">##    5 0.0466495  0.862040</span></span>
<span><span class="co">##   ..       ...       ...</span></span>
<span><span class="co">##   41 0.4367595  0.997155</span></span>
<span><span class="co">##   42 0.4602018  0.997155</span></span>
<span><span class="co">##   43 0.8857989  0.997155</span></span>
<span><span class="co">##   44 0.4453299  0.997155</span></span>
<span><span class="co">##   45 0.0161213  0.725460</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></span></code></pre>
<p>You can get similar results using <em><a href="https://bioconductor.org/packages/3.17/edgeR" class="external-link">edgeR</a></em> using
these functions: <code>DGEList()</code>, <code>calcNormFactors()</code>,
<code>estimateGLMRobustDisp()</code>, <code>glmFit()</code>, and
<code>glmLRT()</code>.</p>
</div>
<div class="section level4">
<h4 id="find-ders-with-limma">Find DERs with <em><a href="https://bioconductor.org/packages/3.17/limma" class="external-link">limma</a></em><a class="anchor" aria-label="anchor" href="#find-ders-with-limma"></a>
</h4>
<p>Alternatively, we can find DERs using <em><a href="https://bioconductor.org/packages/3.17/limma" class="external-link">limma</a></em>. Here
we’ll exemplify a type of test closer to what we’ll do later with the
F-statistics approach. First of all, we need to define our models.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Build models</span></span>
<span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">pheno</span><span class="op">$</span><span class="va">group</span> <span class="op">+</span> <span class="va">pheno</span><span class="op">$</span><span class="va">gender</span><span class="op">)</span></span>
<span><span class="va">mod0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">pheno</span><span class="op">$</span><span class="va">gender</span><span class="op">)</span></span></code></pre></div>
<p>Next, we’ll transform the coverage information using the same default
transformation from <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Transform coverage</span></span>
<span><span class="va">transformedCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">coverageMatrix</span> <span class="op">+</span> <span class="fl">32</span><span class="op">)</span></span></code></pre></div>
<p>We can then fit the models and get the F-statistic p-values and
control the FDR.</p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Example using limma</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">"limma"</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'limma'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:DESeq2':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     plotMA</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     plotMA</span></span></code></pre>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run limma</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">transformedCov</span>, <span class="va">mod</span><span class="op">)</span></span>
<span><span class="va">fit0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">transformedCov</span>, <span class="va">mod0</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Determine DE status for the regions</span></span>
<span><span class="co">## Also in https://github.com/LieberInstitute/jaffelab with help and examples</span></span>
<span><span class="va">getF</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">fit</span>, <span class="va">fit0</span>, <span class="va">theData</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">rss1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/fitted.values.html" class="external-link">fitted</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span> <span class="op">-</span> <span class="va">theData</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="va">df1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">coef</span><span class="op">)</span></span>
<span>    <span class="va">rss0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/fitted.values.html" class="external-link">fitted</a></span><span class="op">(</span><span class="va">fit0</span><span class="op">)</span> <span class="op">-</span> <span class="va">theData</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="va">df0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">fit0</span><span class="op">$</span><span class="va">coef</span><span class="op">)</span></span>
<span>    <span class="va">fstat</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="op">(</span><span class="va">rss0</span> <span class="op">-</span> <span class="va">rss1</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">df1</span> <span class="op">-</span> <span class="va">df0</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">rss1</span> <span class="op">/</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">theData</span><span class="op">)</span> <span class="op">-</span> <span class="va">df1</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">f_pval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Fdist.html" class="external-link">pf</a></span><span class="op">(</span><span class="va">fstat</span>, <span class="va">df1</span> <span class="op">-</span> <span class="va">df0</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">theData</span><span class="op">)</span> <span class="op">-</span> <span class="va">df1</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>    <span class="va">fout</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">fstat</span>, <span class="va">df1</span> <span class="op">-</span> <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">theData</span><span class="op">)</span> <span class="op">-</span> <span class="va">df1</span>, <span class="va">f_pval</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">fout</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"df1"</span>, <span class="st">"df0"</span><span class="op">)</span></span>
<span>    <span class="va">fout</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">fout</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">fout</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">ff</span> <span class="op">&lt;-</span> <span class="fu">getF</span><span class="op">(</span><span class="va">fit</span>, <span class="va">fit0</span>, <span class="va">transformedCov</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Get the p-value and assign it to the regions</span></span>
<span><span class="va">limma</span> <span class="op">&lt;-</span> <span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span></span>
<span><span class="va">limma</span><span class="op">$</span><span class="va">fstat</span> <span class="op">&lt;-</span> <span class="va">ff</span><span class="op">$</span><span class="va">fstat</span></span>
<span><span class="va">limma</span><span class="op">$</span><span class="va">pvalue</span> <span class="op">&lt;-</span> <span class="va">ff</span><span class="op">$</span><span class="va">f_pval</span></span>
<span><span class="va">limma</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">ff</span><span class="op">$</span><span class="va">f_pval</span>, <span class="st">"BH"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Explore the results</span></span>
<span><span class="va">limma</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 45 ranges and 9 metadata columns:</span></span>
<span><span class="co">##      seqnames            ranges strand |     value      area indexStart</span></span>
<span><span class="co">##         &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;  &lt;integer&gt;</span></span>
<span><span class="co">##    1    chr21   9827018-9827582      * |  313.6717 177224.53          1</span></span>
<span><span class="co">##    2    chr21 15457301-15457438      * |  215.0846  29681.68        566</span></span>
<span><span class="co">##    3    chr21 20230140-20230192      * |   38.8325   2058.12        704</span></span>
<span><span class="co">##    4    chr21 20230445-20230505      * |   41.3245   2520.80        757</span></span>
<span><span class="co">##    5    chr21 27253318-27253543      * |   34.9131   7890.37        818</span></span>
<span><span class="co">##   ..      ...               ...    ... .       ...       ...        ...</span></span>
<span><span class="co">##   41    chr21 33039644-33039688      * |   34.4705 1551.1742       2180</span></span>
<span><span class="co">##   42    chr21 33040784-33040798      * |   32.1342  482.0133       2225</span></span>
<span><span class="co">##   43    chr21          33040890      * |   30.0925   30.0925       2240</span></span>
<span><span class="co">##   44    chr21 33040900-33040901      * |   30.1208   60.2417       2241</span></span>
<span><span class="co">##   45    chr21 48019401-48019414      * |   31.1489  436.0850       2243</span></span>
<span><span class="co">##       indexEnd cluster clusterL     fstat    pvalue      padj</span></span>
<span><span class="co">##      &lt;integer&gt;   &lt;Rle&gt;    &lt;Rle&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##    1       565       1      565  1.638455 0.2325446  0.581362</span></span>
<span><span class="co">##    2       703       2      138  4.307443 0.0677644  0.324601</span></span>
<span><span class="co">##    3       756       3      366  1.323342 0.2796406  0.629191</span></span>
<span><span class="co">##    4       817       3      366  0.380332 0.5527044  0.863074</span></span>
<span><span class="co">##    5      1043       4      765  7.249519 0.0246955  0.309532</span></span>
<span><span class="co">##   ..       ...     ...      ...       ...       ...       ...</span></span>
<span><span class="co">##   41      2224      17       45   3.11799 0.1112440  0.385075</span></span>
<span><span class="co">##   42      2239      18      118   3.66184 0.0879543  0.329829</span></span>
<span><span class="co">##   43      2240      18      118   3.87860 0.0804175  0.328981</span></span>
<span><span class="co">##   44      2242      18      118   4.39338 0.0655381  0.324601</span></span>
<span><span class="co">##   45      2256      19       14   6.80915 0.0282970  0.309532</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></span></code></pre>
<p>In this simple example, none of the ERs have strong differential
expression signal when adjusting for an FDR of 5%.</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">limma</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">deseq</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        </span></span>
<span><span class="co">##         FALSE</span></span>
<span><span class="co">##   FALSE    45</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="via-railmatrix">Via <code>railMatrix()</code><a class="anchor" aria-label="anchor" href="#via-railmatrix"></a>
</h3>
<p>If you have <a href="http://rail.bio" class="external-link">Rail</a> output, you can get
the same results faster than with <code><a href="../reference/regionMatrix.html">regionMatrix()</a></code>. <a href="http://rail.bio" class="external-link">Rail</a> will create the summarized coverage
BigWig file for you, but we are not including it in this package due to
its size. So, lets create it.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Calculate the mean: this step takes a long time with many samples</span></span>
<span><span class="va">meanCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="va">fullCov</span><span class="op">$</span><span class="va">chr21</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">fullCov</span><span class="op">$</span><span class="va">chr21</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Save it on a bigwig file called meanChr21.bw</span></span>
<span><span class="fu"><a href="../reference/createBw.html">createBw</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"chr21"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span><span class="st">"meanChr21"</span> <span class="op">=</span> <span class="va">meanCov</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    keepGR <span class="op">=</span></span>
<span>        <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2023-05-07 03:55:10.104309 coerceGR: coercing sample meanChr21</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:10.242256 createBwSample: exporting bw for sample meanChr21</span></span></code></pre>
<p>Now that we have the files <a href="http://rail.bio" class="external-link">Rail</a> creates
for us, we can use <code><a href="../reference/railMatrix.html">railMatrix()</a></code>.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Identify files to use</span></span>
<span><span class="va">summaryFile</span> <span class="op">&lt;-</span> <span class="st">"meanChr21.bw"</span></span>
<span><span class="co">## We had already found the sample BigWig files and saved it in the object 'files'</span></span>
<span><span class="co">## Lets just rename it to sampleFiles for clarity.</span></span>
<span><span class="va">sampleFiles</span> <span class="op">&lt;-</span> <span class="va">files</span></span>
<span></span>
<span><span class="co">## Get the regions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span></span>
<span>    <span class="va">regionMat.rail</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/railMatrix.html">railMatrix</a></span><span class="op">(</span></span>
<span>        chrs <span class="op">=</span> <span class="st">"chr21"</span>, summaryFiles <span class="op">=</span> <span class="va">summaryFile</span>,</span>
<span>        sampleFiles <span class="op">=</span> <span class="va">sampleFiles</span>, L <span class="op">=</span> <span class="fl">76</span>, cutoff <span class="op">=</span> <span class="fl">30</span>, maxClusterGap <span class="op">=</span> <span class="fl">3000L</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2023-05-07 03:55:12.31958 loadCoverage: finding chromosome lengths</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:12.3288 loadCoverage: loading BigWig file meanChr21.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:12.739384 loadCoverage: applying the cutoff to the merged data</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.031253 filterData: originally there were 48129895 rows, now there are 48129895 rows. Meaning that 0 percent was filtered.</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.259395 filterData: originally there were 48129895 rows, now there are 2256 rows. Meaning that 100 percent was filtered.</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.262671 findRegions: identifying potential segments</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.266207 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.266513 .getSegmentsRle: segmenting with cutoff(s) 30</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.275871 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.344693 findRegions: identifying region clusters</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.418387 railMatrix: processing regions 1 to 45</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.424341 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB113.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.511727 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB123.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.598803 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB126.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.684181 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB130.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.770387 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB135.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.855568 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB136.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:15.942099 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB145.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:16.028471 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB153.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:16.122663 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB159.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:16.216498 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB178.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:16.312569 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB92.bw</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:16.417099 railMatrix: processing file /__w/_temp/Library/derfinderData/extdata/AMY/HSB97.bw</span></span></code></pre>
<pre><code><span><span class="co">##    user  system elapsed </span></span>
<span><span class="co">##   4.199   0.008   4.209</span></span></code></pre>
<p>When you take into account the time needed to load the data
(<code><a href="../reference/fullCoverage.html">fullCoverage()</a></code>) and then creating the matrix
(<code><a href="../reference/regionMatrix.html">regionMatrix()</a></code>), <code><a href="../reference/railMatrix.html">railMatrix()</a></code> is faster and
less memory intensive.</p>
<p>The objects are not identical due to small rounding errors, but it’s
nothing to worry about.</p>
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Overall not identical due to small rounding errors</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">regionMat</span>, <span class="va">regionMat.rail</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] FALSE</span></span></code></pre>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Actual regions are the same</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/range-squeezers.html" class="external-link">ranges</a></span><span class="op">(</span><span class="va">regionMat.rail</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## When you round, the small differences go away</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">value</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">regionMat.rail</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">value</span>, <span class="fl">4</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">area</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">regionMat.rail</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">area</span>, <span class="fl">4</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="single-base-level-f-statistics-analysis">Single base-level F-statistics analysis<a class="anchor" aria-label="anchor" href="#single-base-level-f-statistics-analysis"></a>
</h2>
<p>One form of base-level differential expression analysis implemented
in <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
is to calculate F-statistics for every base and use them to define
candidate differentially expressed regions. This type of analysis is
further explained in this section.</p>
<div class="section level3">
<h3 id="models">Models<a class="anchor" aria-label="anchor" href="#models"></a>
</h3>
<p>Once we have the base-level coverage data for all 12 samples, we can
construct the models. In this case, we want to find differences between
fetal and adult samples while adjusting for gender and a proxy of the
library size.</p>
<p>We can use <code><a href="../reference/sampleDepth.html">sampleDepth()</a></code> and it’s helper function
<code><a href="../reference/collapseFullCoverage.html">collapseFullCoverage()</a></code> to get a proxy of the library size.
Note that you would normally use the unfiltered data from all the
chromosomes in this step and not just one.</p>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get some idea of the library sizes</span></span>
<span><span class="va">sampleDepths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sampleDepth.html">sampleDepth</a></span><span class="op">(</span><span class="fu"><a href="../reference/collapseFullCoverage.html">collapseFullCoverage</a></span><span class="op">(</span><span class="va">fullCov</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2023-05-07 03:55:16.715657 sampleDepth: Calculating sample quantiles</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:16.94321 sampleDepth: Calculating sample adjustments</span></span></code></pre>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleDepths</span></span></code></pre></div>
<pre><code><span><span class="co">## HSB113.100% HSB123.100% HSB126.100% HSB130.100% HSB135.100% HSB136.100% </span></span>
<span><span class="co">##    19.82106    19.40505    19.53045    19.52017    20.33392    19.97758 </span></span>
<span><span class="co">## HSB145.100% HSB153.100% HSB159.100% HSB178.100%  HSB92.100%  HSB97.100% </span></span>
<span><span class="co">##    19.49827    19.41285    19.24186    19.44252    19.55904    19.47733</span></span></code></pre>
<p><code><a href="../reference/sampleDepth.html">sampleDepth()</a></code> is similar to
<code>calcNormFactors()</code> from <a href="http://www.bioconductor.org/packages/release/bioc/html/metagenomeSeq.html" class="external-link">metagenomeSeq</a>
with some code underneath tailored for the type of data we are using.
<code><a href="../reference/collapseFullCoverage.html">collapseFullCoverage()</a></code> is only needed to deal with the size
of the data.</p>
<p>We can then define the nested models we want to use using
<code><a href="../reference/makeModels.html">makeModels()</a></code>. This is a helper function that assumes that
you will always adjust for the library size. You then need to define the
variable to test, in this case we are comparing fetal vs adult samples.
Optionally, you can adjust for other sample covariates, such as the
gender in this case.</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Define models</span></span>
<span><span class="va">models</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeModels.html">makeModels</a></span><span class="op">(</span><span class="va">sampleDepths</span>,</span>
<span>    testvars <span class="op">=</span> <span class="va">pheno</span><span class="op">$</span><span class="va">group</span>,</span>
<span>    adjustvars <span class="op">=</span> <span class="va">pheno</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gender"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Explore the models</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">models</span>, <span class="va">head</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $mod</span></span>
<span><span class="co">##   (Intercept) testvarsadult sampleDepths adjustVar1M</span></span>
<span><span class="co">## 1           1             0     19.82106           0</span></span>
<span><span class="co">## 2           1             0     19.40505           1</span></span>
<span><span class="co">## 3           1             0     19.53045           1</span></span>
<span><span class="co">## 4           1             0     19.52017           1</span></span>
<span><span class="co">## 5           1             0     20.33392           0</span></span>
<span><span class="co">## 6           1             0     19.97758           0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $mod0</span></span>
<span><span class="co">##   (Intercept) sampleDepths adjustVar1M</span></span>
<span><span class="co">## 1           1     19.82106           0</span></span>
<span><span class="co">## 2           1     19.40505           1</span></span>
<span><span class="co">## 3           1     19.53045           1</span></span>
<span><span class="co">## 4           1     19.52017           1</span></span>
<span><span class="co">## 5           1     20.33392           0</span></span>
<span><span class="co">## 6           1     19.97758           0</span></span></code></pre>
<p>Note how the null model (<code>mod0</code>) is nested in the
alternative model (<code>mod</code>). Use the same models for all your
chromosomes unless you have a specific reason to use chromosome-specific
models. Note that <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
is very flexible and works with any type of nested model.</p>
</div>
<div class="section level3">
<h3 id="find-candidate-ders">Find candidate DERs<a class="anchor" aria-label="anchor" href="#find-candidate-ders"></a>
</h3>
<p>Next, we can find candidate differentially expressed regions (DERs)
using as input the segments of the genome where at least one sample has
coverage greater than 2. That is, the filtered coverage version we
created previously.</p>
<p>The <strong>main</strong> function in <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
for this type of analysis is <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>. It works at a
chromosome level and runs behinds the scenes several other <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
functions. To use it, you have to provide the models, the grouping
information, how to calculate the F-statistic cutoff and most
importantly, the number of permutations.</p>
<p>By default <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> will use a theoretical cutoff.
In this example, we use the cutoff that would correspond to a p-value of
0.05. To assign p-values to the candidate DERs, <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
permutes the rows of the model matrices, re-calculates the F-statistics
and identifies null regions. Then it compares the area of the observed
regions versus the areas from the null regions to assign an empirical
p-value.</p>
<p>In this example we will use twenty permutations, although in a real
case scenario you might consider a larger number of permutations.</p>
<p>In real scenario, you might consider saving the results from all the
chromosomes in a given directory. Here we will use
<em>analysisResults</em>. For each chromosome you analyze, a new
directory with the chromosome-specific data will be created. So in this
case, we will have <em>analysisResults/chr21</em>.</p>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create a analysis directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"analysisResults"</span><span class="op">)</span></span>
<span><span class="va">originalWd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">originalWd</span>, <span class="st">"analysisResults"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Perform differential expression analysis</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/analyzeChr.html">analyzeChr</a></span><span class="op">(</span></span>
<span>    chr <span class="op">=</span> <span class="st">"chr21"</span>, <span class="va">filteredCov</span><span class="op">$</span><span class="va">chr21</span>, <span class="va">models</span>,</span>
<span>    groupInfo <span class="op">=</span> <span class="va">pheno</span><span class="op">$</span><span class="va">group</span>, writeOutput <span class="op">=</span> <span class="cn">TRUE</span>, cutoffFstat <span class="op">=</span> <span class="fl">5e-02</span>,</span>
<span>    nPermute <span class="op">=</span> <span class="fl">20</span>, seeds <span class="op">=</span> <span class="fl">20140923</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span>, returnOutput <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2023-05-07 03:55:18.181992 analyzeChr: Pre-processing the coverage data</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:20.719012 analyzeChr: Calculating statistics</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:20.722853 calculateStats: calculating the F-statistics</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:21.046211 analyzeChr: Calculating pvalues</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:21.046578 analyzeChr: Using the following theoretical cutoff for the F-statistics 5.31765507157871</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:21.04777 calculatePvalues: identifying data segments</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:21.051841 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:22.05669 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:22.106606 findRegions: identifying region clusters</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:22.314168 calculatePvalues: calculating F-statistics for permutation 1 and seed 20140924</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:22.520378 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:22.554569 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:22.617906 calculatePvalues: calculating F-statistics for permutation 2 and seed 20140925</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:22.845918 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:22.879316 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:22.93353 calculatePvalues: calculating F-statistics for permutation 3 and seed 20140926</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:23.139586 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:23.175848 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:23.233863 calculatePvalues: calculating F-statistics for permutation 4 and seed 20140927</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:23.450427 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:23.48377 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:23.53723 calculatePvalues: calculating F-statistics for permutation 5 and seed 20140928</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:23.760179 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:23.793616 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:23.851702 calculatePvalues: calculating F-statistics for permutation 6 and seed 20140929</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:24.073765 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:24.107848 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:24.162272 calculatePvalues: calculating F-statistics for permutation 7 and seed 20140930</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:24.390271 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:24.424194 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:24.478669 calculatePvalues: calculating F-statistics for permutation 8 and seed 20140931</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:24.684326 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:24.717958 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:24.77324 calculatePvalues: calculating F-statistics for permutation 9 and seed 20140932</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:25.002691 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:25.035959 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:25.093563 calculatePvalues: calculating F-statistics for permutation 10 and seed 20140933</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:25.32385 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:25.35723 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:25.415801 calculatePvalues: calculating F-statistics for permutation 11 and seed 20140934</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:25.638287 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:25.671856 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:25.725825 calculatePvalues: calculating F-statistics for permutation 12 and seed 20140935</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:25.942798 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:25.976187 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:26.030626 calculatePvalues: calculating F-statistics for permutation 13 and seed 20140936</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:26.247059 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:26.281352 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:26.334144 calculatePvalues: calculating F-statistics for permutation 14 and seed 20140937</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:26.552751 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:26.596143 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:26.652122 calculatePvalues: calculating F-statistics for permutation 15 and seed 20140938</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:26.869282 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:26.902771 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:26.956482 calculatePvalues: calculating F-statistics for permutation 16 and seed 20140939</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:27.194407 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:27.228281 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:27.281702 calculatePvalues: calculating F-statistics for permutation 17 and seed 20140940</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:27.493447 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:27.526183 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:27.579976 calculatePvalues: calculating F-statistics for permutation 18 and seed 20140941</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:27.803567 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:27.836589 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:27.896491 calculatePvalues: calculating F-statistics for permutation 19 and seed 20140942</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:28.117963 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:28.151563 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:28.205596 calculatePvalues: calculating F-statistics for permutation 20 and seed 20140943</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:28.441663 findRegions: segmenting information</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:28.474771 findRegions: identifying candidate regions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:28.546054 calculatePvalues: calculating the p-values</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:55:28.633447 analyzeChr: Annotating regions</span></span></code></pre>
<pre><code><span><span class="co">## No annotationPackage supplied. Trying org.Hs.eg.db.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: org.Hs.eg.db</span></span></code></pre>
<pre><code><span><span class="co">## Warning in library(package, lib.loc = lib.loc, character.only = TRUE,</span></span>
<span><span class="co">## logical.return = TRUE, : there is no package called 'org.Hs.eg.db'</span></span></code></pre>
<pre><code><span><span class="co">## Could not load org.Hs.eg.db. Will continue without annotation</span></span></code></pre>
<pre><code><span><span class="co">## Getting TSS and TSE.</span></span></code></pre>
<pre><code><span><span class="co">## Getting CSS and CSE.</span></span></code></pre>
<pre><code><span><span class="co">## Getting exons.</span></span></code></pre>
<pre><code><span><span class="co">## .....</span></span></code></pre>
<pre><code><span><span class="co">##    user  system elapsed </span></span>
<span><span class="co">##  55.765   1.973  54.235</span></span></code></pre>
<p>To speed up <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>, you might need to use several
cores via the <code>mc.cores</code> argument. If memory is limiting, you
might want to use a smaller <code>chunksize</code> (default is 5
million). Note that if you use too many cores, you might hit the
input/output ceiling of your data network and/or hard drives speed.</p>
<p>Before running with a large number of permutations we recommend
exploring how long each permutation cycle takes using a single
permutation.</p>
<p>Note that analyzing each chromosome with a large number of
permutations and a rich data set can take several hours, so we recommend
running one job running <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> per chromosome, and
then merging the results via <code><a href="../reference/mergeResults.html">mergeResults()</a></code>. This process
is further described in the advanced <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
vignette.</p>
</div>
<div class="section level3">
<h3 id="explore-results">Explore results<a class="anchor" aria-label="anchor" href="#explore-results"></a>
</h3>
<p>When using <code>returnOutput = TRUE</code>,
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> will return a list with the results to explore
interactively. However, by default it writes the results to disk (one
.Rdata file per result).</p>
<p>The following code explores the results.</p>
<div class="sourceCode" id="cb219"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Explore</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "timeinfo"     "optionsStats" "coveragePrep" "fstats"       "regions"     </span></span>
<span><span class="co">## [6] "annotation"</span></span></code></pre>
<div class="section level4">
<h4 id="optionstats">optionStats<a class="anchor" aria-label="anchor" href="#optionstats"></a>
</h4>
<p><code>optionStats</code> stores the main options used in the
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> call including the models used, the type of
cutoff, number of permutations, seeds for the permutations. All this
information can be useful to reproduce the analysis.</p>
<div class="sourceCode" id="cb221"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Explore optionsStats</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">optionsStats</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "models"          "cutoffPre"       "scalefac"        "chunksize"      </span></span>
<span><span class="co">##  [5] "cutoffFstat"     "cutoffType"      "nPermute"        "seeds"          </span></span>
<span><span class="co">##  [9] "groupInfo"       "lowMemDir"       "analyzeCall"     "cutoffFstatUsed"</span></span>
<span><span class="co">## [13] "smooth"          "smoothFunction"  "weights"         "returnOutput"</span></span></code></pre>
<div class="sourceCode" id="cb223"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Call used</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">optionsStats</span><span class="op">$</span><span class="va">analyzeCall</span></span></code></pre></div>
<pre><code><span><span class="co">## analyzeChr(chr = "chr21", coverageInfo = filteredCov$chr21, models = models, </span></span>
<span><span class="co">##     cutoffFstat = 0.05, nPermute = 20, seeds = 20140923 + seq_len(20), </span></span>
<span><span class="co">##     groupInfo = pheno$group, writeOutput = TRUE, returnOutput = TRUE)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="coverageprep">coveragePrep<a class="anchor" aria-label="anchor" href="#coverageprep"></a>
</h4>
<p><code>coveragePrep</code> has the result from the
<code><a href="../reference/preprocessCoverage.html">preprocessCoverage()</a></code> step. This includes the genomic
position index, the mean coverage (after scaling and the log2
transformation) for all the samples, and the group mean coverages. By
default, the chunks are written to disk in
<code>optionsStats$lowMemDir</code> (chr21/chunksDir in this example) to
help reduce the required memory resources. Otherwise it is stored in
<code>coveragePrep$coverageProcessed</code>.</p>
<div class="sourceCode" id="cb225"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Explore coveragePrep</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">coveragePrep</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "coverageProcessed" "mclapplyIndex"     "position"         </span></span>
<span><span class="co">## [4] "meanCoverage"      "groupMeans"</span></span></code></pre>
<div class="sourceCode" id="cb227"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Group means</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">coveragePrep</span><span class="op">$</span><span class="va">groupMeans</span></span></code></pre></div>
<pre><code><span><span class="co">## $fetal</span></span>
<span><span class="co">## numeric-Rle of length 130356 with 116452 runs</span></span>
<span><span class="co">##   Lengths:        1        1        1        1 ...        1        1        1</span></span>
<span><span class="co">##   Values : 0.401667 0.428333 0.435000 0.461667 ...  1.34000  1.33333  1.24833</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $adult</span></span>
<span><span class="co">## numeric-Rle of length 130356 with 119226 runs</span></span>
<span><span class="co">##   Lengths:        1        1        1        1 ...        1        1        1</span></span>
<span><span class="co">##   Values : 0.406667 0.413333 0.430000 0.448333 ...  1.77667  1.73833  1.62667</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="fstats">fstats<a class="anchor" aria-label="anchor" href="#fstats"></a>
</h4>
<p>The F-statistics are then stored in <code>fstats</code>. These are
calculated using <code><a href="../reference/calculateStats.html">calculateStats()</a></code>.</p>
<div class="sourceCode" id="cb229"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Explore optionsStats</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">fstats</span></span></code></pre></div>
<pre><code><span><span class="co">## numeric-Rle of length 130356 with 126807 runs</span></span>
<span><span class="co">##   Lengths:          1          1          1 ...          1          1</span></span>
<span><span class="co">##   Values : 0.01922610 0.02996937 0.02066332 ...   2.507611   2.324638</span></span></code></pre>
<div class="sourceCode" id="cb231"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Note that the length matches the number of bases used</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">fstats</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">coveragePrep</span><span class="op">$</span><span class="va">position</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="regions">regions<a class="anchor" aria-label="anchor" href="#regions"></a>
</h4>
<p>The candidate DERs and summary results from the permutations is then
stored in <code>regions</code>. This is the output from
<code><a href="../reference/calculatePvalues.html">calculatePvalues()</a></code> which uses several underneath other
functions including <code><a href="../reference/calculateStats.html">calculateStats()</a></code> and
<code><a href="../reference/findRegions.html">findRegions()</a></code>.</p>
<div class="sourceCode" id="cb233"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Explore regions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "regions"         "nullStats"       "nullWidths"      "nullPermutation"</span></span></code></pre>
<p>For the null regions, the summary information is composed of the mean
F-statistic for the null regions (<code>regions$nullStats</code>), the
width of the null regions (<code>regions$nullWidths</code>), and the
permutation number under which they were identified
(<code>regions$nullPermutation</code>).</p>
<div class="sourceCode" id="cb235"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Permutation summary information</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## $nullStats</span></span>
<span><span class="co">## numeric-Rle of length 13994 with 13994 runs</span></span>
<span><span class="co">##   Lengths:        1        1        1        1 ...        1        1        1</span></span>
<span><span class="co">##   Values :  5.43461  5.71738  6.37821  6.33171 ...  5.35554  5.36614  5.62516</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $nullWidths</span></span>
<span><span class="co">## integer-Rle of length 13994 with 12365 runs</span></span>
<span><span class="co">##   Lengths:   2   1   1   1   1   1   3   1 ...   1   1   1   2   1   1   4   1</span></span>
<span><span class="co">##   Values :   1  24   7   1  32   2   1  11 ...   1   3  45   4  28   6   1   2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $nullPermutation</span></span>
<span><span class="co">## integer-Rle of length 13994 with 20 runs</span></span>
<span><span class="co">##   Lengths:  246  350  574  554  396  462  482 ...  746  114 1460  428  802  278</span></span>
<span><span class="co">##   Values :    1    2    3    4    5    6    7 ...   15   16   17   18   19   20</span></span></code></pre>
<p>The most important part of the output is the <code>GRanges</code>
object with the candidate DERs shown below.</p>
<div class="sourceCode" id="cb237"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Candidate DERs</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">regions</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 591 ranges and 14 metadata columns:</span></span>
<span><span class="co">##      seqnames            ranges strand |     value      area indexStart</span></span>
<span><span class="co">##         &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt; &lt;numeric&gt;  &lt;integer&gt;</span></span>
<span><span class="co">##   up    chr21 47610386-47610682      * |  11.10304   3297.60     122158</span></span>
<span><span class="co">##   up    chr21 40196145-40196444      * |  10.06142   3018.43      76110</span></span>
<span><span class="co">##   up    chr21 27253616-27253948      * |   8.43488   2808.82      22019</span></span>
<span><span class="co">##   up    chr21 22115534-22115894      * |   7.23645   2612.36      12274</span></span>
<span><span class="co">##   up    chr21 22914853-22915064      * |   9.78066   2073.50      17318</span></span>
<span><span class="co">##   ..      ...               ...    ... .       ...       ...        ...</span></span>
<span><span class="co">##   up    chr21          35889784      * |   5.31952   5.31952      60088</span></span>
<span><span class="co">##   up    chr21          47610093      * |   5.31912   5.31912     121865</span></span>
<span><span class="co">##   up    chr21          16333728      * |   5.31881   5.31881       5048</span></span>
<span><span class="co">##   up    chr21          34001896      * |   5.31871   5.31871      32577</span></span>
<span><span class="co">##   up    chr21          34809571      * |   5.31801   5.31801      43694</span></span>
<span><span class="co">##       indexEnd cluster clusterL meanCoverage meanfetal meanadult</span></span>
<span><span class="co">##      &lt;integer&gt;   &lt;Rle&gt;    &lt;Rle&gt;    &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   up    122454     138      933     1.597952   0.82289  2.373013</span></span>
<span><span class="co">##   up     76409      71     1323     1.303508   2.02532  0.581694</span></span>
<span><span class="co">##   up     22351      28      407    33.657858  42.46704 24.848674</span></span>
<span><span class="co">##   up     12634       9      694     0.964464   1.71906  0.209871</span></span>
<span><span class="co">##   up     17529      21      217     2.838978   4.23593  1.442028</span></span>
<span><span class="co">##   ..       ...     ...      ...          ...       ...       ...</span></span>
<span><span class="co">##   up     60088      51      742      2.75417   3.36000   2.14833</span></span>
<span><span class="co">##   up    121865     138      933      1.45583   0.77500   2.13667</span></span>
<span><span class="co">##   up      5048       1        9      1.19500   1.23167   1.15833</span></span>
<span><span class="co">##   up     32577      38     1428      1.71250   2.33333   1.09167</span></span>
<span><span class="co">##   up     43694      46      149      2.95000   2.87833   3.02167</span></span>
<span><span class="co">##      log2FoldChangeadultvsfetal    pvalues significant   qvalues</span></span>
<span><span class="co">##                       &lt;numeric&gt;  &lt;numeric&gt;    &lt;factor&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   up                   1.527949 0.00278671        TRUE  0.738407</span></span>
<span><span class="co">##   up                  -1.799818 0.00378707        TRUE  0.738407</span></span>
<span><span class="co">##   up                  -0.773175 0.00464452        TRUE  0.738407</span></span>
<span><span class="co">##   up                  -3.034045 0.00535906        TRUE  0.738407</span></span>
<span><span class="co">##   up                  -1.554578 0.00793140        TRUE  0.738407</span></span>
<span><span class="co">##   ..                        ...        ...         ...       ...</span></span>
<span><span class="co">##   up                 -0.6452433   0.997856       FALSE  0.974463</span></span>
<span><span class="co">##   up                  1.4630937   0.998285       FALSE  0.974463</span></span>
<span><span class="co">##   up                 -0.0885613   0.998714       FALSE  0.974463</span></span>
<span><span class="co">##   up                 -1.0958600   0.998714       FALSE  0.974463</span></span>
<span><span class="co">##   up                  0.0701108   0.999571       FALSE  0.974463</span></span>
<span><span class="co">##      significantQval</span></span>
<span><span class="co">##             &lt;factor&gt;</span></span>
<span><span class="co">##   up           FALSE</span></span>
<span><span class="co">##   up           FALSE</span></span>
<span><span class="co">##   up           FALSE</span></span>
<span><span class="co">##   up           FALSE</span></span>
<span><span class="co">##   up           FALSE</span></span>
<span><span class="co">##   ..             ...</span></span>
<span><span class="co">##   up           FALSE</span></span>
<span><span class="co">##   up           FALSE</span></span>
<span><span class="co">##   up           FALSE</span></span>
<span><span class="co">##   up           FALSE</span></span>
<span><span class="co">##   up           FALSE</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span></span></code></pre>
<p>The metadata columns are:</p>
<ul>
<li>
<em>value</em> is the mean F-statistics for the candidate DER.</li>
<li>
<em>area</em> is the sum of the F-statistics for the candidate
DER.</li>
<li>
<em>indexStart</em> Relates the genomic start coordinate with the
filtered genomic index start coordinate.</li>
<li>
<em>indexEnd</em> Similarly as above but for the end
coordinates.</li>
<li>
<em>cluster</em> The cluster id to which this candidate DER belongs
to.</li>
<li>
<em>clusterL</em> The length of the cluster to which this candidate
DER belongs to.</li>
<li>
<em>meanCoverage</em> The base level mean coverage for the candidate
DER.</li>
<li>
<em>meanfetal</em> In this example, the mean coverage for the fetal
samples.</li>
<li>
<em>meanadult</em> In this example, the mean coverage for the adult
samples.</li>
<li>
<em>log2FoldChangeadultvsfetal</em> In this example, the log2 fold
change between adult vs fetal samples.</li>
<li>
<em>pvalues</em> The p-value for the candidate DER.</li>
<li>
<em>significant</em> By default, whether the p-value is less than
0.05 or not.</li>
<li>
<em>qvalues</em> The q-value for the candidate DER calculated with
<a href="http://www.bioconductor.org/packages/release/bioc/html/qvalue.html" class="external-link">qvalue</a>.</li>
<li>
<em>significantQval</em> By default, whether the q-value is less
than 0.10 or not.</li>
</ul>
<p>Note that for this type of analysis you might want to try a few
coverage cutoffs and/or F-statistic cutoffs. One quick way to evaluate
the results is to compare the width of the regions.</p>
<div class="sourceCode" id="cb239"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Width of potential DERs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">##    1.00    1.00    4.00   17.98   17.00  361.00</span></span></code></pre>
<div class="sourceCode" id="cb241"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 68</span></span></code></pre>
<div class="sourceCode" id="cb243"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Width of candidate DERs</span></span>
<span><span class="va">sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">as.logical</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">significant</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span><span class="va">sig</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">##    65.0    81.5    97.0   127.8   122.0   361.0</span></span></code></pre>
<div class="sourceCode" id="cb245"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span><span class="va">sig</span><span class="op">]</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 35</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="nearest-annotation">Nearest annotation<a class="anchor" aria-label="anchor" href="#nearest-annotation"></a>
</h4>
<p><code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> will find the nearest annotation feature
using <code><a href="https://rdrr.io/pkg/bumphunter/man/matchGenes.html" class="external-link">matchGenes()</a></code> from <em><a href="https://bioconductor.org/packages/3.17/bumphunter" class="external-link">bumphunter</a></em>
(version &gt;= 1.7.3). This information is useful considering that the
candidate DERs were identified without relying on annotation. Yet at the
end, we are interested to check if they are inside a known exon,
upstream a gene, etc.</p>
<div class="sourceCode" id="cb247"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Nearest annotation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">annotation</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   name annotation description     region distance   subregion insideDistance</span></span>
<span><span class="co">## 1 &lt;NA&gt;       &lt;NA&gt; inside exon     inside    38056 inside exon              0</span></span>
<span><span class="co">## 2 &lt;NA&gt;       &lt;NA&gt; inside exon     inside    18390 inside exon              0</span></span>
<span><span class="co">## 3 &lt;NA&gt;       &lt;NA&gt; inside exon     inside   289498 inside exon              0</span></span>
<span><span class="co">## 4 &lt;NA&gt;       &lt;NA&gt; inside exon     inside    59532 inside exon              0</span></span>
<span><span class="co">## 5 &lt;NA&gt;       &lt;NA&gt;  downstream downstream   544220        &lt;NA&gt;             NA</span></span>
<span><span class="co">## 6 &lt;NA&gt;       &lt;NA&gt; inside exon     inside    59315 inside exon              0</span></span>
<span><span class="co">##   exonnumber nexons                         UTR strand  geneL codingL Geneid</span></span>
<span><span class="co">## 1         22     22                       3'UTR      -  39700      NA   4047</span></span>
<span><span class="co">## 2         10     10                       3'UTR      +  19123      NA   2114</span></span>
<span><span class="co">## 3         16     16                       3'UTR      - 290585      NA    351</span></span>
<span><span class="co">## 4          7      7 inside transcription region      -  60513      NA 387486</span></span>
<span><span class="co">## 5         NA     15                        &lt;NA&gt;      + 541884      NA   4685</span></span>
<span><span class="co">## 6          7      7 inside transcription region      -  60513      NA 387486</span></span>
<span><span class="co">##   subjectHits</span></span>
<span><span class="co">## 1       34356</span></span>
<span><span class="co">## 2       17340</span></span>
<span><span class="co">## 3       31161</span></span>
<span><span class="co">## 4       32699</span></span>
<span><span class="co">## 5       37133</span></span>
<span><span class="co">## 6       32699</span></span></code></pre>
<p>For more details on the output please check the <em><a href="https://bioconductor.org/packages/3.17/bumphunter" class="external-link">bumphunter</a></em>
package.</p>
<p>Check the section about non-human data (specifically, using
annotation different from <strong>hg19</strong>) on the advanced
vignette.</p>
</div>
<div class="section level4">
<h4 id="time-spent">Time spent<a class="anchor" aria-label="anchor" href="#time-spent"></a>
</h4>
<p>The final piece is the wallclock time spent during each of the steps
in <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>. You can then make a plot with this
information as shown in Figure @ref(fig:exploreTime).</p>
<div class="sourceCode" id="cb249"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Time spent</span></span>
<span><span class="va">results</span><span class="op">$</span><span class="va">timeinfo</span></span></code></pre></div>
<pre><code><span><span class="co">##                      init                     setup                  prepData </span></span>
<span><span class="co">## "2023-05-07 03:55:18 UTC" "2023-05-07 03:55:18 UTC" "2023-05-07 03:55:20 UTC" </span></span>
<span><span class="co">##                  savePrep            calculateStats                 saveStats </span></span>
<span><span class="co">## "2023-05-07 03:55:20 UTC" "2023-05-07 03:55:20 UTC" "2023-05-07 03:55:21 UTC" </span></span>
<span><span class="co">##             saveStatsOpts          calculatePvalues                  saveRegs </span></span>
<span><span class="co">## "2023-05-07 03:55:21 UTC" "2023-05-07 03:55:28 UTC" "2023-05-07 03:55:28 UTC" </span></span>
<span><span class="co">##                  annotate                  saveAnno </span></span>
<span><span class="co">## "2023-05-07 03:56:12 UTC" "2023-05-07 03:56:12 UTC"</span></span></code></pre>
<div class="sourceCode" id="cb251"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Use this information to make a plot</span></span>
<span><span class="va">timed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">results</span><span class="op">$</span><span class="va">timeinfo</span><span class="op">)</span></span>
<span><span class="va">timed.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Seconds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">timed</span><span class="op">)</span>, Step <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">timed</span><span class="op">)</span>,</span>
<span>    levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">timed</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">timed.df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">Step</span>, x <span class="op">=</span> <span class="va">Seconds</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="derfinder-users-guide_files/figure-html/exploreTime-1.png" alt="Seconds used to run each step in analyzeChr()." width="700"><p class="caption">
Seconds used to run each step in analyzeChr().
</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="merge-results">Merge results<a class="anchor" aria-label="anchor" href="#merge-results"></a>
</h3>
<p>Once you have analyzed each chromosome using
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>, you can use <code><a href="../reference/mergeResults.html">mergeResults()</a></code> to
merge the results. This function does not return an object in R but
instead creates several Rdata files with the main results from the
different chromosomes.</p>
<div class="sourceCode" id="cb252"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Go back to the original directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">originalWd</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Merge results from several chromosomes. In this case we only have one.</span></span>
<span><span class="fu"><a href="../reference/mergeResults.html">mergeResults</a></span><span class="op">(</span></span>
<span>    chrs <span class="op">=</span> <span class="st">"chr21"</span>, prefix <span class="op">=</span> <span class="st">"analysisResults"</span>,</span>
<span>    genomicState <span class="op">=</span> <span class="va">genomicState</span><span class="op">$</span><span class="va">fullGenome</span>,</span>
<span>    optionsStats <span class="op">=</span> <span class="va">results</span><span class="op">$</span><span class="va">optionsStats</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2023-05-07 03:56:13.868673 mergeResults: Saving options used</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:13.870671 Loading chromosome chr21</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:13.948135 mergeResults: calculating FWER</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:13.990788 mergeResults: Saving fullNullSummary</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:14.001698 mergeResults: Re-calculating the p-values</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:14.079212 mergeResults: Saving fullRegions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:14.086161 mergeResults: assigning genomic states</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:14.201496 annotateRegions: counting</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:14.29637 annotateRegions: annotating</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:14.329989 mergeResults: Saving fullAnnotatedRegions</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:14.332013 mergeResults: Saving fullFstats</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:14.390113 mergeResults: Saving fullTime</span></span></code></pre>
<div class="sourceCode" id="cb265"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Files created by mergeResults()</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="st">"analysisResults"</span>, pattern <span class="op">=</span> <span class="st">".Rdata"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "fullAnnotatedRegions.Rdata" "fullFstats.Rdata"          </span></span>
<span><span class="co">## [3] "fullNullSummary.Rdata"      "fullRegions.Rdata"         </span></span>
<span><span class="co">## [5] "fullTime.Rdata"             "optionsMerge.Rdata"</span></span></code></pre>
<ul>
<li>
<em>fullFstats.Rdata</em> contains a list with one element per
chromosome. Per chromosome it has the F-statistics.</li>
<li>
<em>fullNullSummary.Rdata</em> is a list with the summary
information from the null regions stored for each chromosome.</li>
<li>
<em>fullTime.Rdata</em> has the timing information for each
chromosome as a list.</li>
</ul>
<div class="section level4">
<h4 id="optionsmerge">optionsMerge<a class="anchor" aria-label="anchor" href="#optionsmerge"></a>
</h4>
<p>For reproducibility purposes, the options used the merge the results
are stored in <code>optionsMerge</code>.</p>
<div class="sourceCode" id="cb267"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Options used to merge</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"analysisResults"</span>, <span class="st">"optionsMerge.Rdata"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Contents</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">optionsMerge</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "chrs"            "significantCut"  "minoverlap"      "mergeCall"      </span></span>
<span><span class="co">## [5] "cutoffFstatUsed" "optionsStats"</span></span></code></pre>
<div class="sourceCode" id="cb269"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Merge call</span></span>
<span><span class="va">optionsMerge</span><span class="op">$</span><span class="va">mergeCall</span></span></code></pre></div>
<pre><code><span><span class="co">## mergeResults(chrs = "chr21", prefix = "analysisResults", genomicState = genomicState$fullGenome, </span></span>
<span><span class="co">##     optionsStats = results$optionsStats)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="fullregions">fullRegions<a class="anchor" aria-label="anchor" href="#fullregions"></a>
</h4>
<p>The main result from <code><a href="../reference/mergeResults.html">mergeResults()</a></code> is in
<code>fullRegions</code>. This is a <code>GRanges</code> object with the
candidate DERs from all the chromosomes. It also includes the nearest
annotation metadata as well as FWER adjusted p-values (<em>fwer</em>)
and whether the FWER adjusted p-value is less than 0.05
(<em>significantFWER</em>).</p>
<div class="sourceCode" id="cb271"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load all the regions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"analysisResults"</span>, <span class="st">"fullRegions.Rdata"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Metadata columns</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">fullRegions</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "value"                      "area"                      </span></span>
<span><span class="co">##  [3] "indexStart"                 "indexEnd"                  </span></span>
<span><span class="co">##  [5] "cluster"                    "clusterL"                  </span></span>
<span><span class="co">##  [7] "meanCoverage"               "meanfetal"                 </span></span>
<span><span class="co">##  [9] "meanadult"                  "log2FoldChangeadultvsfetal"</span></span>
<span><span class="co">## [11] "pvalues"                    "significant"               </span></span>
<span><span class="co">## [13] "qvalues"                    "significantQval"           </span></span>
<span><span class="co">## [15] "name"                       "annotation"                </span></span>
<span><span class="co">## [17] "description"                "region"                    </span></span>
<span><span class="co">## [19] "distance"                   "subregion"                 </span></span>
<span><span class="co">## [21] "insideDistance"             "exonnumber"                </span></span>
<span><span class="co">## [23] "nexons"                     "UTR"                       </span></span>
<span><span class="co">## [25] "annoStrand"                 "geneL"                     </span></span>
<span><span class="co">## [27] "codingL"                    "Geneid"                    </span></span>
<span><span class="co">## [29] "subjectHits"                "fwer"                      </span></span>
<span><span class="co">## [31] "significantFWER"</span></span></code></pre>
<p>Note that <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> only has the information for a
given chromosome at a time, so <code><a href="../reference/mergeResults.html">mergeResults()</a></code> re-calculates
the p-values and q-values using the information from all the
chromosomes.</p>
</div>
<div class="section level4">
<h4 id="fullannotatedregions">fullAnnotatedRegions<a class="anchor" aria-label="anchor" href="#fullannotatedregions"></a>
</h4>
<p>In preparation for visually exploring the results,
<code><a href="../reference/mergeResults.html">mergeResults()</a></code> will run <code><a href="../reference/annotateRegions.html">annotateRegions()</a></code>
which counts how many known exons, introns and intergenic segments each
candidate DER overlaps (by default with a minimum overlap of 20bp).
<code><a href="../reference/annotateRegions.html">annotateRegions()</a></code> uses a summarized version of the genome
annotation created with <code><a href="../reference/makeGenomicState.html">makeGenomicState()</a></code>. For this
example, we can use the data included in <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
which is the summarized annotation of hg19 for chromosome 21.</p>
<div class="sourceCode" id="cb273"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load annotateRegions() output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"analysisResults"</span>, <span class="st">"fullAnnotatedRegions.Rdata"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Information stored</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fullAnnotatedRegions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "countTable"     "annotationList"</span></span></code></pre>
<div class="sourceCode" id="cb275"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Take a peak</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fullAnnotatedRegions</span>, <span class="va">head</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $countTable</span></span>
<span><span class="co">##   exon intergenic intron</span></span>
<span><span class="co">## 1    1          0      0</span></span>
<span><span class="co">## 2    1          0      0</span></span>
<span><span class="co">## 3    1          0      0</span></span>
<span><span class="co">## 4    1          0      0</span></span>
<span><span class="co">## 5    0          1      0</span></span>
<span><span class="co">## 6    1          0      0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $annotationList</span></span>
<span><span class="co">## GRangesList object of length 6:</span></span>
<span><span class="co">## $`1`</span></span>
<span><span class="co">## GRanges object with 1 range and 4 metadata columns:</span></span>
<span><span class="co">##        seqnames            ranges strand |   theRegion                 tx_id</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;IntegerList&gt;</span></span>
<span><span class="co">##   4871    chr21 47609038-47611149      - |        exon 73448,73449,73450,...</span></span>
<span><span class="co">##                                     tx_name          gene</span></span>
<span><span class="co">##                             &lt;CharacterList&gt; &lt;IntegerList&gt;</span></span>
<span><span class="co">##   4871 uc002zij.3,uc002zik.2,uc002zil.2,...           170</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from hg19 genome</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`2`</span></span>
<span><span class="co">## GRanges object with 1 range and 4 metadata columns:</span></span>
<span><span class="co">##        seqnames            ranges strand |   theRegion         tx_id</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;IntegerList&gt;</span></span>
<span><span class="co">##   1189    chr21 40194598-40196878      + |        exon   72757,72758</span></span>
<span><span class="co">##                      tx_name          gene</span></span>
<span><span class="co">##              &lt;CharacterList&gt; &lt;IntegerList&gt;</span></span>
<span><span class="co">##   1189 uc002yxf.3,uc002yxg.4            77</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from hg19 genome</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`3`</span></span>
<span><span class="co">## GRanges object with 1 range and 4 metadata columns:</span></span>
<span><span class="co">##        seqnames            ranges strand |   theRegion                 tx_id</span></span>
<span><span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;         &lt;IntegerList&gt;</span></span>
<span><span class="co">##   2965    chr21 27252861-27254082      - |        exon 73066,73067,73068,...</span></span>
<span><span class="co">##                                     tx_name          gene</span></span>
<span><span class="co">##                             &lt;CharacterList&gt; &lt;IntegerList&gt;</span></span>
<span><span class="co">##   2965 uc002ylz.3,uc002yma.3,uc002ymb.3,...           139</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from hg19 genome</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ...</span></span>
<span><span class="co">## &lt;3 more elements&gt;</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="chip-seq-differential-binding">ChIP-seq differential binding<a class="anchor" aria-label="anchor" href="#chip-seq-differential-binding"></a>
</h2>
<p>As of version 1.5.27 <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
has parameters that allow smoothing of the single base-level
F-statistics before determining DERs. This allows finding differentially
bounded regions (peaks) using ChIP-seq data. In general, ChIP-seq
studies are smaller than RNA-seq studies which means that the single
base-level F-statistics approach is well suited for differential binding
analysis.</p>
<p>To smooth the F-statistics use <code>smooth = TRUE</code> in
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>. The default smoothing function is
<code><a href="https://rdrr.io/pkg/bumphunter/man/locfitByCluster.html" class="external-link">bumphunter::locfitByCluster()</a></code> and all its parameters can be
passed specified in the call to <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>. In
particular, the <code>minNum</code> and <code>bpSpan</code> arguments
are important. We recommend setting <code>minNum</code> to the minimum
read length and <code>bpSpan</code> to the average peak length expected
in the ChIP-seq data being analyzed. Smoothing the F-statistics will
take longer but not use significantly more memory than the default
behavior. So take this into account when choosing the number of
permutations to run.</p>
</div>
<div class="section level2">
<h2 id="visually-explore-results">Visually explore results<a class="anchor" aria-label="anchor" href="#visually-explore-results"></a>
</h2>
<p>Optionally, we can use the addon package <em><a href="https://bioconductor.org/packages/3.17/derfinderPlot" class="external-link">derfinderPlot</a></em>
to visually explore the results.</p>
<p>To make the region level plots, we will need to extract the region
level coverage data. We can do so using <code><a href="../reference/getRegionCoverage.html">getRegionCoverage()</a></code>
as shown below.</p>
<div class="sourceCode" id="cb277"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Find overlaps between regions and summarized genomic annotation</span></span>
<span><span class="va">annoRegs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotateRegions.html">annotateRegions</a></span><span class="op">(</span><span class="va">fullRegions</span>, <span class="va">genomicState</span><span class="op">$</span><span class="va">fullGenome</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2023-05-07 03:56:14.983116 annotateRegions: counting</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:15.075128 annotateRegions: annotating</span></span></code></pre>
<div class="sourceCode" id="cb280"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Indeed, the result is the same because we only used chr21</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">annoRegs</span>, <span class="va">fullAnnotatedRegions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] FALSE</span></span></code></pre>
<div class="sourceCode" id="cb282"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the region coverage</span></span>
<span><span class="va">regionCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getRegionCoverage.html">getRegionCoverage</a></span><span class="op">(</span><span class="va">fullCov</span>, <span class="va">fullRegions</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2023-05-07 03:56:15.201005 getRegionCoverage: processing chr21</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:15.309516 getRegionCoverage: done processing chr21</span></span></code></pre>
<div class="sourceCode" id="cb285"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Explore the result</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">regionCov</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   HSB113 HSB123 HSB126 HSB130 HSB135 HSB136 HSB145 HSB153 HSB159 HSB178 HSB92</span></span>
<span><span class="co">## 1   0.68   0.44   0.48   0.36   0.19   2.34   1.29   1.77   2.21   2.69  1.89</span></span>
<span><span class="co">## 2   0.60   0.44   0.48   0.36   0.19   2.30   1.29   1.77   2.21   2.64  1.86</span></span>
<span><span class="co">## 3   0.60   0.40   0.48   0.32   0.19   2.39   1.37   1.81   2.31   2.69  1.89</span></span>
<span><span class="co">## 4   0.64   0.40   0.48   0.32   0.19   2.61   1.42   1.89   2.36   2.88  1.89</span></span>
<span><span class="co">## 5   0.64   0.40   0.48   0.36   0.19   2.65   1.42   1.93   2.36   2.88  1.96</span></span>
<span><span class="co">## 6   0.60   0.44   0.48   0.39   0.23   2.65   1.59   1.93   2.36   2.83  1.93</span></span>
<span><span class="co">##   HSB97</span></span>
<span><span class="co">## 1  3.57</span></span>
<span><span class="co">## 2  3.60</span></span>
<span><span class="co">## 3  3.60</span></span>
<span><span class="co">## 4  3.60</span></span>
<span><span class="co">## 5  3.70</span></span>
<span><span class="co">## 6  3.70</span></span></code></pre>
<p>With this, we are all set to visually explore the results.</p>
<div class="sourceCode" id="cb287"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/leekgroup/derfinderPlot" class="external-link">"derfinderPlot"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Overview of the candidate DERs in the genome</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/derfinderPlot/man/plotOverview.html" class="external-link">plotOverview</a></span><span class="op">(</span></span>
<span>    regions <span class="op">=</span> <span class="va">fullRegions</span>, annotation <span class="op">=</span> <span class="va">results</span><span class="op">$</span><span class="va">annotation</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"fwer"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"TxDb.Hsapiens.UCSC.hg19.knownGene"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span></span>
<span></span>
<span><span class="co">## Base-levle coverage plots for the first 10 regions</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/derfinderPlot/man/plotRegionCoverage.html" class="external-link">plotRegionCoverage</a></span><span class="op">(</span></span>
<span>    regions <span class="op">=</span> <span class="va">fullRegions</span>, regionCoverage <span class="op">=</span> <span class="va">regionCov</span>,</span>
<span>    groupInfo <span class="op">=</span> <span class="va">pheno</span><span class="op">$</span><span class="va">group</span>, nearestAnnotation <span class="op">=</span> <span class="va">results</span><span class="op">$</span><span class="va">annotation</span>,</span>
<span>    annotatedRegions <span class="op">=</span> <span class="va">annoRegs</span>, whichRegions <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, txdb <span class="op">=</span> <span class="va">txdb</span>, scalefac <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    ask <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Cluster plot for the first region</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/derfinderPlot/man/plotCluster.html" class="external-link">plotCluster</a></span><span class="op">(</span></span>
<span>    idx <span class="op">=</span> <span class="fl">1</span>, regions <span class="op">=</span> <span class="va">fullRegions</span>, annotation <span class="op">=</span> <span class="va">results</span><span class="op">$</span><span class="va">annotation</span>,</span>
<span>    coverageInfo <span class="op">=</span> <span class="va">fullCov</span><span class="op">$</span><span class="va">chr21</span>, txdb <span class="op">=</span> <span class="va">txdb</span>, groupInfo <span class="op">=</span> <span class="va">pheno</span><span class="op">$</span><span class="va">group</span>,</span>
<span>    titleUse <span class="op">=</span> <span class="st">"fwer"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <a href="quickstart-derfinder.html">quick start to using
derfinder</a> has example plots for the expressed regions-level
approach. The vignette for <em><a href="https://bioconductor.org/packages/3.17/derfinderPlot" class="external-link">derfinderPlot</a></em>
has even more examples.</p>
</div>
<div class="section level2">
<h2 id="interactive-html-reports">Interactive HTML reports<a class="anchor" aria-label="anchor" href="#interactive-html-reports"></a>
</h2>
<p>We have also developed an addon package called <em><a href="https://bioconductor.org/packages/3.17/regionReport" class="external-link">regionReport</a></em>
available via <a href="http://bioconductor.org/packages/regionReport" class="external-link">Bioconductor</a>.</p>
<p>The function <code>derfinderReport()</code> in <em><a href="https://bioconductor.org/packages/3.17/regionReport" class="external-link">regionReport</a></em>
basically takes advantage of the results from
<code><a href="../reference/mergeResults.html">mergeResults()</a></code> and plotting functions available in <em><a href="https://bioconductor.org/packages/3.17/derfinderPlot" class="external-link">derfinderPlot</a></em>
as well as other neat features from <em><a href="https://CRAN.R-project.org/package=knitrBootstrap" class="external-link">knitrBootstrap</a></em>.
It then generates a customized report for single-base level F-statistics
DER finding analyses.</p>
<p>For results from <code><a href="../reference/regionMatrix.html">regionMatrix()</a></code> or
<code><a href="../reference/railMatrix.html">railMatrix()</a></code> use <code>renderReport()</code> from <em><a href="https://bioconductor.org/packages/3.17/regionReport" class="external-link">regionReport</a></em>.
In both cases, the resulting HTML report promotes reproducibility of the
analysis and allows you to explore in more detail the results through
some diagnostic plots.</p>
<p>We think that these reports are very important when you are exploring
the resulting DERs after changing a key parameter in
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>, <code><a href="../reference/regionMatrix.html">regionMatrix()</a></code> or
<code><a href="../reference/railMatrix.html">railMatrix()</a></code>.</p>
<p>Check out the vignette for <em><a href="https://bioconductor.org/packages/3.17/regionReport" class="external-link">regionReport</a></em>
for example reports generated with it.</p>
</div>
<div class="section level2">
<h2 id="miscellaneous-features">Miscellaneous features<a class="anchor" aria-label="anchor" href="#miscellaneous-features"></a>
</h2>
<p>In this section we go over some other features of <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
which can be useful for performing feature-counts based analyses,
exploring the results, or exporting data.</p>
<div class="section level3">
<h3 id="feature-level-analysis">Feature level analysis<a class="anchor" aria-label="anchor" href="#feature-level-analysis"></a>
</h3>
<p>Similar to the expressed region-level analysis, you might be
interested in performing a feature-level analysis. More specifically,
this means getting a count matrix at the exon-level (or gene-level).
<code><a href="../reference/coverageToExon.html">coverageToExon()</a></code> allows you to get such a matrix by taking
advantage of the summarized annotation produced by
<code><a href="../reference/makeGenomicState.html">makeGenomicState()</a></code>.</p>
<p>In this example, we use the genomic state included in the package
which has the information for chr21 <em><a href="https://bioconductor.org/packages/3.17/TxDb.Hsapiens.UCSC.hg19.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg19.knownGene</a></em>
annotation.</p>
<div class="sourceCode" id="cb288"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the exon-level matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">exonCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coverageToExon.html">coverageToExon</a></span><span class="op">(</span><span class="va">fullCov</span>, <span class="va">genomicState</span><span class="op">$</span><span class="va">fullGenome</span>, L <span class="op">=</span> <span class="fl">76</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SerialParam</span></span>
<span><span class="co">##   bpisup: FALSE; bpnworkers: 1; bptasks: 0; bpjobname: BPJOB</span></span>
<span><span class="co">##   bplog: FALSE; bpthreshold: INFO; bpstopOnError: TRUE</span></span>
<span><span class="co">##   bpRNGseed: ; bptimeout: NA; bpprogressbar: FALSE</span></span>
<span><span class="co">##   bpexportglobals: FALSE; bpexportvariables: FALSE; bpforceGC: FALSE</span></span>
<span><span class="co">##   bpfallback: FALSE</span></span>
<span><span class="co">##   bplogdir: NA</span></span>
<span><span class="co">##   bpresultdir: NA</span></span>
<span><span class="co">## class: SerialParam</span></span>
<span><span class="co">##   bpisup: FALSE; bpnworkers: 1; bptasks: 0; bpjobname: BPJOB</span></span>
<span><span class="co">##   bplog: FALSE; bpthreshold: INFO; bpstopOnError: TRUE</span></span>
<span><span class="co">##   bpRNGseed: ; bptimeout: NA; bpprogressbar: FALSE</span></span>
<span><span class="co">##   bpexportglobals: FALSE; bpexportvariables: FALSE; bpforceGC: FALSE</span></span>
<span><span class="co">##   bpfallback: FALSE</span></span>
<span><span class="co">##   bplogdir: NA</span></span>
<span><span class="co">##   bpresultdir: NA</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:21.7708 coverageToExon: processing chromosome chr21</span></span></code></pre>
<pre><code><span><span class="co">## class: SerialParam</span></span>
<span><span class="co">##   bpisup: FALSE; bpnworkers: 1; bptasks: 0; bpjobname: BPJOB</span></span>
<span><span class="co">##   bplog: FALSE; bpthreshold: INFO; bpstopOnError: TRUE</span></span>
<span><span class="co">##   bpRNGseed: ; bptimeout: NA; bpprogressbar: FALSE</span></span>
<span><span class="co">##   bpexportglobals: FALSE; bpexportvariables: FALSE; bpforceGC: FALSE</span></span>
<span><span class="co">##   bpfallback: FALSE</span></span>
<span><span class="co">##   bplogdir: NA</span></span>
<span><span class="co">##   bpresultdir: NA</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:24.454374 coverageToExon: processing chromosome chr21</span></span></code></pre>
<pre><code><span><span class="co">##    user  system elapsed </span></span>
<span><span class="co">##   7.790   0.660   8.453</span></span></code></pre>
<div class="sourceCode" id="cb294"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Dimensions of the matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">exonCov</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 2658   12</span></span></code></pre>
<div class="sourceCode" id="cb296"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Explore a little bit</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">exonCov</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         HSB113      HSB123      HSB126      HSB130       HSB135       HSB136</span></span>
<span><span class="co">## 4983 0.1173684   0.1382895   0.0000000   0.0000000   0.07894737   0.03947368</span></span>
<span><span class="co">## 4985 4.6542105   1.4557895   1.6034210   0.7798684   1.29592103   0.72986842</span></span>
<span><span class="co">## 4986 7.1510526 291.0205261 252.4892106 152.1905258 439.69500020 263.63486853</span></span>
<span><span class="co">## 4988 0.0000000   0.7063158   0.7223684   0.2639474   0.48657895   0.37657895</span></span>
<span><span class="co">## 4990 2.3064474  64.9423686  69.6584212  35.5769737 101.76394746  76.96736838</span></span>
<span><span class="co">## 4992 0.1652632   8.1834211   9.7538158   2.4193421  10.08618420  14.41013157</span></span>
<span><span class="co">##            HSB145     HSB153      HSB159    HSB178       HSB92       HSB97</span></span>
<span><span class="co">## 4983 5.263158e-03  0.1052632  0.04934211 0.1289474  0.03986842  0.44605264</span></span>
<span><span class="co">## 4985 9.964474e-01 10.4906578  4.39013167 7.3119736  1.65184211 10.87723678</span></span>
<span><span class="co">## 4986 2.345414e+02  4.5310526 10.64368415 4.9807895 11.40447366  6.23315790</span></span>
<span><span class="co">## 4988 6.156579e-01  0.0000000  0.00000000 0.0000000  0.05644737  0.00000000</span></span>
<span><span class="co">## 4990 5.978842e+01  1.2284210  2.67486840 0.9542105  3.98013159  1.96131579</span></span>
<span><span class="co">## 4992 5.325658e+00  0.2402632  0.73039474 0.2601316  0.66907895  0.07473684</span></span></code></pre>
<p>With this matrix, rounded if necessary, you can proceed to use
packages such as <em><a href="https://bioconductor.org/packages/3.17/limma" class="external-link">limma</a></em>,
<em><a href="https://bioconductor.org/packages/3.17/DESeq2" class="external-link">DESeq2</a></em>,
<em><a href="https://bioconductor.org/packages/3.17/edgeR" class="external-link">edgeR</a></em> among
others.</p>
</div>
<div class="section level3">
<h3 id="compare-results-visually">Compare results visually<a class="anchor" aria-label="anchor" href="#compare-results-visually"></a>
</h3>
<p>We can certainly make region-level plots using
<code><a href="https://rdrr.io/pkg/derfinderPlot/man/plotRegionCoverage.html" class="external-link">plotRegionCoverage()</a></code> or cluster plots using
<code><a href="https://rdrr.io/pkg/derfinderPlot/man/plotCluster.html" class="external-link">plotCluster()</a></code> or overview plots using
<code>plotOveview()</code>, all from <em><a href="https://bioconductor.org/packages/3.17/derfinderPlot" class="external-link">derfinderPlot</a></em>.</p>
<p>First we need to get the relevant annotation information.</p>
<div class="sourceCode" id="cb298"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Annotate regions as exonic, intronic or intergenic</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">annoGenome</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/annotateRegions.html">annotateRegions</a></span><span class="op">(</span></span>
<span>    <span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span>,</span>
<span>    <span class="va">genomicState</span><span class="op">$</span><span class="va">fullGenome</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2023-05-07 03:56:26.17906 annotateRegions: counting</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:26.273836 annotateRegions: annotating</span></span></code></pre>
<pre><code><span><span class="co">##    user  system elapsed </span></span>
<span><span class="co">##   0.220   0.004   0.225</span></span></code></pre>
<div class="sourceCode" id="cb302"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Note that the genomicState object included in derfinder only has information</span></span>
<span><span class="co">## for chr21 (hg19).</span></span>
<span></span>
<span><span class="co">## Identify closest genes to regions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/rafalab/bumphunter" class="external-link">"bumphunter"</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"TxDb.Hsapiens.UCSC.hg19.knownGene"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bumphunter/man/annotateTranscripts.html" class="external-link">annotateTranscripts</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## No annotationPackage supplied. Trying org.Hs.eg.db.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: org.Hs.eg.db</span></span></code></pre>
<pre><code><span><span class="co">## Warning in library(package, lib.loc = lib.loc, character.only = TRUE,</span></span>
<span><span class="co">## logical.return = TRUE, : there is no package called 'org.Hs.eg.db'</span></span></code></pre>
<pre><code><span><span class="co">## Could not load org.Hs.eg.db. Will continue without annotation</span></span></code></pre>
<pre><code><span><span class="co">## Getting TSS and TSE.</span></span></code></pre>
<pre><code><span><span class="co">## Getting CSS and CSE.</span></span></code></pre>
<pre><code><span><span class="co">## Getting exons.</span></span></code></pre>
<div class="sourceCode" id="cb310"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">annoNear</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/bumphunter/man/matchGenes.html" class="external-link">matchGenes</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span>, <span class="va">genes</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    user  system elapsed </span></span>
<span><span class="co">##   2.787   0.000   2.788</span></span></code></pre>
<p>Now we can proceed to use <em><a href="https://bioconductor.org/packages/3.17/derfinderPlot" class="external-link">derfinderPlot</a></em>
to make the region-level plots for the top 100 regions.</p>
<div class="sourceCode" id="cb312"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Identify the top regions by highest total coverage</span></span>
<span><span class="va">top</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span><span class="op">$</span><span class="va">area</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Base-level plots for the top 100 regions with transcript information</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/leekgroup/derfinderPlot" class="external-link">"derfinderPlot"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/derfinderPlot/man/plotRegionCoverage.html" class="external-link">plotRegionCoverage</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span>,</span>
<span>    regionCoverage <span class="op">=</span> <span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">bpCoverage</span>,</span>
<span>    groupInfo <span class="op">=</span> <span class="va">pheno</span><span class="op">$</span><span class="va">group</span>, nearestAnnotation <span class="op">=</span> <span class="va">annoNear</span>,</span>
<span>    annotatedRegions <span class="op">=</span> <span class="va">annoGenome</span>, whichRegions <span class="op">=</span> <span class="va">top</span>, scalefac <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    txdb <span class="op">=</span> <span class="va">txdb</span>, ask <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>However, we can alternatively use <em><a href="https://bioconductor.org/packages/3.17/epivizr" class="external-link">epivizr</a></em>
to view the candidate DERs and the region matrix results in a genome
browser.</p>
<div class="sourceCode" id="cb313"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load epivizr, it's available from Bioconductor</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"epivizr"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load data to your browser</span></span>
<span><span class="va">mgr</span> <span class="op">&lt;-</span> <span class="fu">startEpiviz</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ders_dev</span> <span class="op">&lt;-</span> <span class="va">mgr</span><span class="op">$</span><span class="fu">addDevice</span><span class="op">(</span></span>
<span>    <span class="va">fullRegions</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">as.logical</a></span><span class="op">(</span><span class="va">fullRegions</span><span class="op">$</span><span class="va">significantFWER</span><span class="op">)</span><span class="op">]</span>, <span class="st">"Candidate DERs"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ders_potential_dev</span> <span class="op">&lt;-</span> <span class="va">mgr</span><span class="op">$</span><span class="fu">addDevice</span><span class="op">(</span></span>
<span>    <span class="va">fullRegions</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">as.logical</a></span><span class="op">(</span><span class="va">fullRegions</span><span class="op">$</span><span class="va">significantFWER</span><span class="op">)</span><span class="op">]</span>, <span class="st">"Potential DERs"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">regs_dev</span> <span class="op">&lt;-</span> <span class="va">mgr</span><span class="op">$</span><span class="fu">addDevice</span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span>, <span class="st">"Region Matrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Go to a place you like in the genome</span></span>
<span><span class="va">mgr</span><span class="op">$</span><span class="fu">navigate</span><span class="op">(</span></span>
<span>    <span class="st">"chr21"</span>, <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">start</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span><span class="va">top</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">-</span> <span class="fl">100</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/start.html" class="external-link">end</a></span><span class="op">(</span><span class="va">regionMat</span><span class="op">$</span><span class="va">chr21</span><span class="op">$</span><span class="va">regions</span><span class="op">[</span><span class="va">top</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Stop the navigation</span></span>
<span><span class="va">mgr</span><span class="op">$</span><span class="fu">stopServer</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="export-coverage-to-bigwig-files">Export coverage to BigWig files<a class="anchor" aria-label="anchor" href="#export-coverage-to-bigwig-files"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
also includes <code><a href="../reference/createBw.html">createBw()</a></code> with related functions
<code><a href="../reference/createBwSample.html">createBwSample()</a></code> and <code><a href="../reference/coerceGR.html">coerceGR()</a></code> to export the
output of <code><a href="../reference/fullCoverage.html">fullCoverage()</a></code> to BigWig files. These functions
can be useful in the case where you start with BAM files and later on
want to save the coverage data into BigWig files, which are generally
smaller.</p>
<div class="sourceCode" id="cb314"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Subset only the first sample</span></span>
<span><span class="va">fullCovSmall</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fullCov</span>, <span class="st">"["</span>, <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Export to BigWig</span></span>
<span><span class="va">bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createBw.html">createBw</a></span><span class="op">(</span><span class="va">fullCovSmall</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2023-05-07 03:56:36.60597 coerceGR: coercing sample HSB113</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:36.657117 createBwSample: exporting bw for sample HSB113</span></span></code></pre>
<div class="sourceCode" id="cb317"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## See the file. Note that the sample name is used to name the file.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".bw"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "HSB113.bw"    "meanChr21.bw"</span></span></code></pre>
<div class="sourceCode" id="cb319"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Internally createBw() coerces each sample to a GRanges object before</span></span>
<span><span class="co">## exporting to a BigWig file. If more than one sample was exported, the</span></span>
<span><span class="co">## GRangesList would have more elements.</span></span>
<span><span class="va">bw</span></span></code></pre></div>
<pre><code><span><span class="co">## GRangesList object of length 1:</span></span>
<span><span class="co">## $HSB113</span></span>
<span><span class="co">## GRanges object with 155950 ranges and 1 metadata column:</span></span>
<span><span class="co">##         seqnames            ranges strand |     score</span></span>
<span><span class="co">##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;</span></span>
<span><span class="co">##   chr21    chr21   9458667-9458741      * |      0.04</span></span>
<span><span class="co">##   chr21    chr21   9540957-9540971      * |      0.04</span></span>
<span><span class="co">##   chr21    chr21   9543719-9543778      * |      0.04</span></span>
<span><span class="co">##   chr21    chr21   9651480-9651554      * |      0.04</span></span>
<span><span class="co">##   chr21    chr21   9653397-9653471      * |      0.04</span></span>
<span><span class="co">##     ...      ...               ...    ... .       ...</span></span>
<span><span class="co">##   chr21    chr21 48093246-48093255      * |      0.04</span></span>
<span><span class="co">##   chr21    chr21 48093257-48093331      * |      0.04</span></span>
<span><span class="co">##   chr21    chr21 48093350-48093424      * |      0.04</span></span>
<span><span class="co">##   chr21    chr21 48112194-48112268      * |      0.04</span></span>
<span><span class="co">##   chr21    chr21 48115056-48115130      * |      0.04</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 1 sequence from an unspecified genome</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="advanced-arguments">Advanced arguments<a class="anchor" aria-label="anchor" href="#advanced-arguments"></a>
</h2>
<p>If you are interested in using advanced arguments in <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>,
they are described in the manual pages of each function. Some of the
most common advanced arguments are:</p>
<ul>
<li>
<code>chrsStyle</code> (default is <code>UCSC</code>)</li>
<li>
<code>verbose</code> (by default <code>TRUE</code>).</li>
</ul>
<p><code>verbose</code> controls whether to print status updates for
nearly all the functions. <code>chrsStyle</code> is used to determine
the chromosome naming style and is powered by <em><a href="https://bioconductor.org/packages/3.17/GenomeInfoDb" class="external-link">GenomeInfoDb</a></em>.
Note that <code>chrsStyle</code> is used in any of the functions that
call <code><a href="../reference/extendedMapSeqlevels.html">extendedMapSeqlevels()</a></code>. If you are working with a
different organism than <em>Homo sapiens</em> set the global
<code>species</code> option using
<code>options(species = 'your species')</code> with the syntax used in
<code>names(GenomeInfoDb::genomeStyles())</code>. If you want to disable
<code><a href="../reference/extendedMapSeqlevels.html">extendedMapSeqlevels()</a></code> set <code>chrsStyle</code> to
<code>NULL</code>, which can be useful if your organism is not part of
<em><a href="https://bioconductor.org/packages/3.17/GenomeInfoDb" class="external-link">GenomeInfoDb</a></em>.</p>
<p>The third commonly used advanced argument is <code>mc.cores</code>.
It controls the number of cores to use for the functions that can run
with more than one core to speed up. In nearly all the cases, the
maximum number of cores depends on the number of chromosomes. One
notable exception is <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> where the maximum number
of cores depends on the <code>chunksize</code> used and the dimensions
of the data for the chromosome under study.</p>
<p>Note that using the <code>...</code> argument allows you to specify
some of the documented arguments. For example, you might want to control
the <code>maxClusterGap</code> from <code><a href="../reference/findRegions.html">findRegions()</a></code> in the
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> call.</p>
</div>
<div class="section level2">
<h2 id="non-human-data">Non-human data<a class="anchor" aria-label="anchor" href="#non-human-data"></a>
</h2>
<p>If you are working with data from an organism that is not <em>Homo
sapiens</em>, then set the global options defining the
<code>species</code> and the <code>chrsStyle</code> used. For example,
if you are working with <em>Arabidopsis Thaliana</em> and the
<em>NCBI</em> naming style, then set the options using the following
code:</p>
<div class="sourceCode" id="cb321"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Set global species and chrsStyle options</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"arabidopsis_thaliana"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>chrsStyle <span class="op">=</span> <span class="st">"NCBI"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Then proceed to load and analyze the data</span></span></code></pre></div>
<p>Internally <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
uses <code><a href="../reference/extendedMapSeqlevels.html">extendedMapSeqlevels()</a></code> to use the appropriate
chromosome naming style given a species in all functions involving
chromosome names.</p>
<p>Further note that the argument <code>txdb</code> from
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> is passed to
<code>bumphunter::annotateTranscripts(txdb)</code>. So if you are using
a genome different from <strong>hg19</strong> remember to provide the
appropriate annotation data or simply use
<code>analyzeChr(runAnnotation = FALSE)</code>.</p>
<p>So, in the <em>Arabidopsis Thaliana</em> example, your
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> call would look like this:</p>
<div class="sourceCode" id="cb322"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load transcript database information</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"TxDb.Athaliana.BioMart.plantsmart28"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Set organism options</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"arabidopsis_thaliana"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>chrsStyle <span class="op">=</span> <span class="st">"NCBI"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run command with more arguments</span></span>
<span><span class="fu"><a href="../reference/analyzeChr.html">analyzeChr</a></span><span class="op">(</span>txdb <span class="op">=</span> <span class="va">TxDb.Athaliana.BioMart.plantsmart28</span><span class="op">)</span></span></code></pre></div>
<p>You might find the discussion <a href="https://support.bioconductor.org/p/62781/" class="external-link">Using bumphunter with
non-human genomes</a> useful.</p>
</div>
<div class="section level2">
<h2 id="functions-that-use-multiple-cores">Functions that use multiple cores<a class="anchor" aria-label="anchor" href="#functions-that-use-multiple-cores"></a>
</h2>
<p>Currently, the following functions can use multiple cores, several of
which are called inside <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>.</p>
<ul>
<li>
<code><a href="../reference/calculatePvalues.html">calculatePvalues()</a></code>: 1 core per chunk of data to
process.</li>
<li>
<code><a href="../reference/calculateStats.html">calculateStats()</a></code>: 1 core per chunk of data to
process.</li>
<li>
<code><a href="../reference/coerceGR.html">coerceGR()</a></code>: 1 core per chromosome. This function is
used by <code><a href="../reference/createBw.html">createBw()</a></code>.</li>
<li>
<code><a href="../reference/coverageToExon.html">coverageToExon()</a></code>: 1 core per strand, then 1 core per
chromosome.</li>
<li>
<code><a href="../reference/loadCoverage.html">loadCoverage()</a></code>: up to 1 core per tile when loading the
data with <a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicFiles.html" class="external-link">GenomicFiles</a>.
Otherwise, no parallelization is used.</li>
<li>
<code><a href="../reference/fullCoverage.html">fullCoverage()</a></code>: 1 core per chromosome. In general, try
to avoid using more than 10 cores as you might reach your maximum
network speed and/or hard disk input/output seed. For the case described
in <code><a href="../reference/loadCoverage.html">loadCoverage()</a></code>, you can specify how many cores to use
per chromosome for the tiles using the <code>mc.cores.load</code>
argument effectively resulting in <code>mc.cores</code> times
<code>mc.cores.load</code> used (otherwise it’s <code>mc.cores</code>
squared).</li>
<li>
<code><a href="../reference/getRegionCoverage.html">getRegionCoverage()</a></code>: 1 core per chromosome.</li>
<li>
<code><a href="../reference/regionMatrix.html">regionMatrix()</a></code>: 1 core per chromosome.</li>
<li>
<code><a href="../reference/railMatrix.html">railMatrix()</a></code>: 1 core per chromosome.</li>
</ul>
<p>All parallel operations use <code><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam()</a></code> from <em><a href="https://bioconductor.org/packages/3.17/BiocParallel" class="external-link">BiocParallel</a></em>
when more than 1 core is being used. Otherwise,
<code><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam()</a></code> is used. Note that if you prefer to specify
other types of parallelization you can do so by specifying the
<code>BPPARAM.custom</code> advanced argument.</p>
<p>Because <code><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam()</a></code> requires <code>R</code> to load the
necessary packages on each worker, the key function
<code>fstats.apply()</code> was isolated in the <em><a href="https://bioconductor.org/packages/3.17/derfinderHelper" class="external-link">derfinderHelper</a></em>
package. This package has much faster loading speeds than <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
which greatly impacts performance on cases where the actual step of
calculating the F-statistics is fast.</p>
<p>You may prefer to use <code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam()</a></code> described in the
<em><a href="https://bioconductor.org/packages/3.17/BiocParallel" class="external-link">BiocParallel</a></em>
vignette. In that case, when using these functions use
<code>BPPARAM.custom = MulticoreParam(workers = x)</code> where
<code>x</code> is the number of cores you want to use. Note that in some
systems, as is the case of the cluster used by <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>’s
developers, the system tools for assessing memory usage can be
misleading, thus resulting in much higher memory loads when using
<code><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam()</a></code> instead of the default
<code><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="loading-data-details">Loading data details<a class="anchor" aria-label="anchor" href="#loading-data-details"></a>
</h2>
<div class="section level3">
<h3 id="controlling-loading-from-bam-files">Controlling loading from BAM files<a class="anchor" aria-label="anchor" href="#controlling-loading-from-bam-files"></a>
</h3>
<p>If you are loading data from BAM files, you might want to specify
some criteria to decide which reads to include or not. For example, your
data might have been generated by a strand-specific protocol. You can do
so by specifying the arguments of <code>scanBamFlag()</code> from <em><a href="https://bioconductor.org/packages/3.17/Rsamtools" class="external-link">Rsamtools</a></em>.</p>
<p>You can also control whether to include or exclude bases with
<code>CIGAR</code> string <code>D</code> (deletion from the reference)
by setting the advanced argument <code>drop.D = TRUE</code> in your
<code><a href="../reference/fullCoverage.html">fullCoverage()</a></code> or <code><a href="../reference/loadCoverage.html">loadCoverage()</a></code> call.</p>
</div>
<div class="section level3">
<h3 id="unfiltered-base-level-coverage">Unfiltered base-level coverage<a class="anchor" aria-label="anchor" href="#unfiltered-base-level-coverage"></a>
</h3>
<p>Note that in most scenarios, the <code>fullCov</code> object
illustrated in the introductory vignette can be large in memory. When
making plots or calculating the region-level coverage, we don’t need the
full information. In such situations, it might pay off to create a
smaller version by loading only the required data. This can be achieved
using the advanced argument <code>which</code> to
<code><a href="../reference/fullCoverage.html">fullCoverage()</a></code> or <code><a href="../reference/loadCoverage.html">loadCoverage()</a></code>.</p>
<p>However, it is important to consider that when reading the data from
BAM files, a read might align partially inside the region of interest.
By default such a read would be discarded and thus the base-level
coverage would be lower than what it is in reality. The advanced
argument <code>protectWhich</code> extends regions by 30 kbp (15 kbp
each side) to help mitigate this issue.</p>
<p>We can illustrate this issue with the example data from <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>.
First, we load in the data and generate some regions of interest.</p>
<div class="sourceCode" id="cb323"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Find some regions to work with</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="st">"loadCoverage"</span>, <span class="st">"derfinder"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## ldCvrg&gt; datadir &lt;- system.file("extdata", "genomeData", package = "derfinder")</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ldCvrg&gt; files &lt;- rawFiles(</span></span>
<span><span class="co">## ldCvrg+     datadir = datadir, samplepatt = "*accepted_hits.bam$",</span></span>
<span><span class="co">## ldCvrg+     fileterm = NULL</span></span>
<span><span class="co">## ldCvrg+ )</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ldCvrg&gt; ## Shorten the column names</span></span>
<span><span class="co">## ldCvrg&gt; names(files) &lt;- gsub("_accepted_hits.bam", "", names(files))</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## ldCvrg&gt; ## Read and filter the data, only for 2 files</span></span>
<span><span class="co">## ldCvrg&gt; dataSmall &lt;- loadCoverage(files = files[1:2], chr = "21", cutoff = 0)</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:37.235364 loadCoverage: finding chromosome lengths</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:37.260011 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009101_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:37.329389 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009102_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:37.388443 loadCoverage: applying the cutoff to the merged data</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:37.411901 filterData: originally there were 48129895 rows, now there are 454 rows. Meaning that 100 percent was filtered.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## ldCvrg&gt; ## Not run: </span></span>
<span><span class="co">## ldCvrg&gt; ##D ## Export to BigWig files</span></span>
<span><span class="co">## ldCvrg&gt; ##D createBw(list("chr21" = dataSmall))</span></span>
<span><span class="co">## ldCvrg&gt; ##D </span></span>
<span><span class="co">## ldCvrg&gt; ##D ## Load data from BigWig files</span></span>
<span><span class="co">## ldCvrg&gt; ##D dataSmall.bw &lt;- loadCoverage(c(</span></span>
<span><span class="co">## ldCvrg&gt; ##D     ERR009101 = "ERR009101.bw", ERR009102 =</span></span>
<span><span class="co">## ldCvrg&gt; ##D         "ERR009102.bw"</span></span>
<span><span class="co">## ldCvrg&gt; ##D ), chr = "chr21")</span></span>
<span><span class="co">## ldCvrg&gt; ##D </span></span>
<span><span class="co">## ldCvrg&gt; ##D ## Compare them</span></span>
<span><span class="co">## ldCvrg&gt; ##D mapply(function(x, y) {</span></span>
<span><span class="co">## ldCvrg&gt; ##D     x - y</span></span>
<span><span class="co">## ldCvrg&gt; ##D }, dataSmall$coverage, dataSmall.bw$coverage)</span></span>
<span><span class="co">## ldCvrg&gt; ##D </span></span>
<span><span class="co">## ldCvrg&gt; ##D ## Note that the only difference is the type of Rle (integer vs numeric) used</span></span>
<span><span class="co">## ldCvrg&gt; ##D ## to store the data.</span></span>
<span><span class="co">## ldCvrg&gt; ## End(Not run)</span></span>
<span><span class="co">## ldCvrg&gt; </span></span>
<span><span class="co">## ldCvrg&gt; </span></span>
<span><span class="co">## ldCvrg&gt; </span></span>
<span><span class="co">## ldCvrg&gt;</span></span></code></pre>
<div class="sourceCode" id="cb331"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="st">"getRegionCoverage"</span>, <span class="st">"derfinder"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## gtRgnC&gt; ## Obtain fullCov object</span></span>
<span><span class="co">## gtRgnC&gt; fullCov &lt;- list("21" = genomeDataRaw$coverage)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## gtRgnC&gt; ## Assign chr lengths using hg19 information, use only first two regions</span></span>
<span><span class="co">## gtRgnC&gt; library("GenomicRanges")</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## gtRgnC&gt; regions &lt;- genomeRegions$regions[1:2]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## gtRgnC&gt; seqlengths(regions) &lt;- seqlengths(getChromInfoFromUCSC("hg19",</span></span>
<span><span class="co">## gtRgnC+     as.Seqinfo = TRUE</span></span>
<span><span class="co">## gtRgnC+ ))[</span></span>
<span><span class="co">## gtRgnC+     mapSeqlevels(names(seqlengths(regions)), "UCSC")</span></span>
<span><span class="co">## gtRgnC+ ]</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## gtRgnC&gt; ## Finally, get the region coverage</span></span>
<span><span class="co">## gtRgnC&gt; regionCov &lt;- getRegionCoverage(fullCov = fullCov, regions = regions)</span></span></code></pre>
<pre><code><span><span class="co">## extendedMapSeqlevels: sequence names mapped from NCBI to UCSC for species homo_sapiens</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:37.749671 getRegionCoverage: processing chr21</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:37.793381 getRegionCoverage: done processing chr21</span></span></code></pre>
<p>Next, we load the coverage again using <code>which</code> but without
any padding. We can see how the coverage is not the same by looking at
the maximum coverage for each sample.</p>
<div class="sourceCode" id="cb336"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Illustrate reading data from a set of regions</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadCoverage.html">loadCoverage</a></span><span class="op">(</span></span>
<span>    files <span class="op">=</span> <span class="va">files</span>, chr <span class="op">=</span> <span class="st">"21"</span>, cutoff <span class="op">=</span> <span class="cn">NULL</span>, which <span class="op">=</span> <span class="va">regions</span>,</span>
<span>    protectWhich <span class="op">=</span> <span class="fl">0</span>, fileStyle <span class="op">=</span> <span class="st">"NCBI"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## extendedMapSeqlevels: sequence names mapped from UCSC to NCBI for species homo_sapiens</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.012712 loadCoverage: finding chromosome lengths</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.019247 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009101_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.081123 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009102_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.14354 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009105_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.206271 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009107_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.266149 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009108_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.326288 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009112_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.385673 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009115_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.446993 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009116_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.505968 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009131_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.564666 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009138_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.625793 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009144_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.68761 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009145_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.776551 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009148_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.834394 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009151_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.89537 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009152_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:38.957485 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009153_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.018284 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009159_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.078111 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009161_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.138299 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009163_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.19697 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009164_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.255811 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009167_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.317813 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031812_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.37905 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031835_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.438092 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031867_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.499357 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031868_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.559783 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031900_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.618632 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031904_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.67836 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031914_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.738021 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031936_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.797695 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031958_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.855202 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031960_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.913489 loadCoverage: applying the cutoff to the merged data</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:39.965801 filterData: originally there were 48129895 rows, now there are 48129895 rows. Meaning that 0 percent was filtered.</span></span></code></pre>
<div class="sourceCode" id="cb372"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Some reads were ignored and thus the coverage is lower as can be seen below:</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">test</span><span class="op">$</span><span class="va">coverage</span>, <span class="va">max</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">genomeDataRaw</span><span class="op">$</span><span class="va">coverage</span>, <span class="va">max</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ERR009101 ERR009102 ERR009105 ERR009107 ERR009108 ERR009112 ERR009115 ERR009116 </span></span>
<span><span class="co">##         0         0         0         0        -1         0        -1        -2 </span></span>
<span><span class="co">## ERR009131 ERR009138 ERR009144 ERR009145 ERR009148 ERR009151 ERR009152 ERR009153 </span></span>
<span><span class="co">##        -2        -2        -1        -1        -3        -3         0        -3 </span></span>
<span><span class="co">## ERR009159 ERR009161 ERR009163 ERR009164 ERR009167 SRR031812 SRR031835 SRR031867 </span></span>
<span><span class="co">##        -3        -3        -1        -3        -3         0        -1         0 </span></span>
<span><span class="co">## SRR031868 SRR031900 SRR031904 SRR031914 SRR031936 SRR031958 SRR031960 </span></span>
<span><span class="co">##         0         0        -1         0         0         0         0</span></span></code></pre>
<p>When we re-load the data using some padding to the regions, we find
that the coverage matches at all the bases.</p>
<div class="sourceCode" id="cb374"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Illustrate reading data from a set of regions</span></span>
<span></span>
<span><span class="va">test2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadCoverage.html">loadCoverage</a></span><span class="op">(</span></span>
<span>    files <span class="op">=</span> <span class="va">files</span>, chr <span class="op">=</span> <span class="st">"21"</span>, cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    which <span class="op">=</span> <span class="va">regions</span>, protectWhich <span class="op">=</span> <span class="fl">3e4</span>, fileStyle <span class="op">=</span> <span class="st">"NCBI"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## extendedMapSeqlevels: sequence names mapped from UCSC to NCBI for species homo_sapiens</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.274935 loadCoverage: finding chromosome lengths</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.281707 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009101_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.339751 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009102_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.398396 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009105_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.456246 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009107_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.513938 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009108_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.573312 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009112_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.630943 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009115_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.692933 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009116_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.751515 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009131_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.812074 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009138_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.869164 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009144_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.92797 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009145_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:40.986571 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009148_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.045626 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009151_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.105466 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009152_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.165804 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009153_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.225944 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009159_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.285352 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009161_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.347123 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009163_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.409982 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009164_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.469272 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/ERR009167_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.553305 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031812_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.611227 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031835_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.669085 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031867_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.724602 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031868_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.780251 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031900_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.836367 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031904_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.892551 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031914_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:41.950506 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031936_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:42.007656 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031958_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:42.063886 loadCoverage: loading BAM file /__w/_temp/Library/derfinder/extdata/genomeData/SRR031960_accepted_hits.bam</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:42.121345 loadCoverage: applying the cutoff to the merged data</span></span></code></pre>
<pre><code><span><span class="co">## 2023-05-07 03:56:42.173839 filterData: originally there were 48129895 rows, now there are 48129895 rows. Meaning that 0 percent was filtered.</span></span></code></pre>
<div class="sourceCode" id="cb410"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Adding some padding to the regions helps get the same coverage</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">test2</span><span class="op">$</span><span class="va">coverage</span>, <span class="va">max</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">genomeDataRaw</span><span class="op">$</span><span class="va">coverage</span>, <span class="va">max</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb412"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## A more detailed test reveals that the coverage matches at every base</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span></span>
<span><span class="op">}</span>, <span class="va">test2</span><span class="op">$</span><span class="va">coverage</span>, <span class="va">genomeDataRaw</span><span class="op">$</span><span class="va">coverage</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>How much padding you need to use will depend on your specific data
set, and you might be comfortable getting approximately the same
coverage values for the sake of greatly reducing the memory resources
needed.</p>
</div>
<div class="section level3">
<h3 id="input-files-in-a-different-naming-style">Input files in a different naming style<a class="anchor" aria-label="anchor" href="#input-files-in-a-different-naming-style"></a>
</h3>
<p>If you are under the case where you like to use a specific chromosome
naming style but the raw data files use another one, you might need to
use the <code>fileStyle</code> argument.</p>
<p>For example, you could be working with <em>Homo sapiens</em> data and
your preferred naming style is <em>UCSC</em> (chr1, chr2, …, chrX, chrY)
but the raw data uses <em>NCBI</em> style names (1, 2, …, X, Y). In that
case, use <code>fullCoverage(fileStyle = 'NCBI')</code> or
<code>loadCoverage(fileStyle = 'NCBI')</code> depending if you are
loading one chromosome or multiple at a time.</p>
</div>
<div class="section level3">
<h3 id="loading-data-in-chunks">Loading data in chunks<a class="anchor" aria-label="anchor" href="#loading-data-in-chunks"></a>
</h3>
<p>If you prefer to do so, <code><a href="../reference/fullCoverage.html">fullCoverage()</a></code> and
<code><a href="../reference/loadCoverage.html">loadCoverage()</a></code> can load the data of a chromosome in chunks
using <em><a href="https://bioconductor.org/packages/3.17/GenomicFiles" class="external-link">GenomicFiles</a></em>.
This is controlled by whether you specify the <code>tilewidth</code>
argument.</p>
<p>Notice that you might run into slight coverage errors near the
borders of the tiles for the same reason that was illustrated previously
when loading specific regions.</p>
<p>This approach is not necessarily more efficient and can be
significantly time consuming if you use a small
<code>tilewidth</code>.</p>
</div>
<div class="section level3">
<h3 id="large-number-of-samples">Large number of samples<a class="anchor" aria-label="anchor" href="#large-number-of-samples"></a>
</h3>
<p>If you have a large number of samples (say thousands), it might be
best to submit cluster jobs that run <code><a href="../reference/loadCoverage.html">loadCoverage()</a></code> or
<code><a href="../reference/fullCoverage.html">fullCoverage()</a></code> for only one chromosome at a time.</p>
<p>In the case of working with <a href="http://rail.bio" class="external-link">Rail</a>
output, you can either use <code><a href="../reference/railMatrix.html">railMatrix()</a></code> with the argument
<code>file.cores</code> greater than 1 or specify the advanced argument
<code>BPPARAM.railChr</code> to control the parallel environment used
for loading the BigWig files. Doing so, you can then submit cluster jobs
that run <code><a href="../reference/railMatrix.html">railMatrix()</a></code> for one chromosome at a time, yet
read in the data fast. You can do all of from R by using
<code>BPPARAM.custom</code> and <code>BPPARAM.railChr</code> at the same
time where you use a <code>BatchJobsParam()</code> for the first
one.</p>
<p>Another option when working with <a href="http://rail.bio" class="external-link">Rail</a>
output is to simply load the summary BigWig data (mean, median), then
define the ERs using <code><a href="../reference/findRegions.html">findRegions()</a></code> and write them to a BED
file. You can then use <a href="https://github.com/CRG-Barcelona/bwtool/wiki" class="external-link">bwtool</a> to create
the coverage matrix.</p>
</div>
</div>
<div class="section level2">
<h2 id="flow-charts">Flow charts<a class="anchor" aria-label="anchor" href="#flow-charts"></a>
</h2>
<div class="section level3">
<h3 id="der-analysis-flow-chart">DER analysis flow chart<a class="anchor" aria-label="anchor" href="#der-analysis-flow-chart"></a>
</h3>
<p>Figure @ref(fig:Figure1) illustrates how most of <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>’s
functions interact when performing a base-level differential expression
analysis by calculating base-level F-statistics.</p>
<div class="figure" style="text-align: center">
<img src="http://lcolladotor.github.io/derfinder/fig/DERpathway.png" alt="derfinder F-statistics flow chart." width="100%"><p class="caption">
derfinder F-statistics flow chart.
</p>
</div>
<p>Flow chart of the different processing steps (black boxes) that can
be carried out using <em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
and the functions that perform these actions (in red). Input and output
is shown in green boxes. Functions in blue are those applied to the
results from multiple chromosomes (<code><a href="../reference/mergeResults.html">mergeResults()</a></code> and
<code>derfinderReport</code>). <em><a href="https://bioconductor.org/packages/3.17/regionReport" class="external-link">regionReport</a></em>
functions are shown in orange while <em><a href="https://bioconductor.org/packages/3.17/derfinderPlot" class="external-link">derfinderPlot</a></em>
functions are shown in dark purple. Purple dotted arrow marks functions
that require unfiltered base-level coverage.</p>
</div>
<div class="section level3">
<h3 id="analyzechr-flow-chart">
<code>analyzeChr()</code> flow chart<a class="anchor" aria-label="anchor" href="#analyzechr-flow-chart"></a>
</h3>
<div class="figure" style="text-align: center">
<img src="http://lcolladotor.github.io/derfinder/fig/analyzeChr.png" alt="analyzeChr() flow chart." width="100%"><p class="caption">
analyzeChr() flow chart.
</p>
</div>
<p>Figure @ref(fig:Figure2) shows in more detail the processing flow in
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>, which is the main function for identifying
candidate differentially expressed regions (DERs) from the base-level
F-statistics.</p>
<p>Many fine-tunning arguments can be passed to
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> to feed into the other functions. For example,
you might want to use a smaller <code>chunksize</code> when
pre-processing the coverage data (the default is 5 million): specially
if you have hundreds of samples.</p>
<p>Another useful argument is <code>scalefac</code> (by default it’s 32)
which controls the scaling factor to use before the log2
transformation.</p>
<p>Furthermore, you might want to specify <code>maxClusterGap</code> to
control the maximum gap between two regions before they are considered
to be part of the same cluster.</p>
</div>
<div class="section level3">
<h3 id="regionmatrix-flow-chart">
<code>regionMatrix()</code> flow chart<a class="anchor" aria-label="anchor" href="#regionmatrix-flow-chart"></a>
</h3>
<div class="figure" style="text-align: center">
<img src="http://lcolladotor.github.io/derfinder/fig/regionMatrix.png" alt="regionMatrix() flow chart." width="100%"><p class="caption">
regionMatrix() flow chart.
</p>
</div>
<p>Figure @ref(fig:Figure3) shows the functions used internally by
<code><a href="../reference/regionMatrix.html">regionMatrix()</a></code> and processing steps for identifying
expressed regions. Overall, it is much simpler than
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>. The flow is even simpler in
<code><a href="../reference/railMatrix.html">railMatrix()</a></code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="base-level-f-statistics-projects">Base-level F-statistics projects<a class="anchor" aria-label="anchor" href="#base-level-f-statistics-projects"></a>
</h2>
<p>For each project where you will calculating base-level F-statistics,
we recommend the following organization. You might be interested in a
similar structure when using <code><a href="../reference/regionMatrix.html">regionMatrix()</a></code> if you have BAM
files. The notable exception is when you are analyzing output data from
<a href="http://rail.bio" class="external-link">Rail</a> in which case you won’t be needing
this type of organization. Remember that <code><a href="../reference/railMatrix.html">railMatrix()</a></code> is
much less computationally intensive than <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>.</p>
<div class="section level3">
<h3 id="file-organization">File organization<a class="anchor" aria-label="anchor" href="#file-organization"></a>
</h3>
<p>This is our recommended file organization when using
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>.</p>
<pre><code>ProjectDir
|-derCoverageInfo
|-derAnalysis
|---analysis01
|---analysis02</code></pre>
<p>We start with a main project directory that has two initial
directories. One for storing the coverage data, and one for storing each
analysis: you might explore different models, cutoffs or other
parameters.</p>
<p>You can then use <code><a href="../reference/fullCoverage.html">fullCoverage()</a></code>, save the result and
also save the filtered coverage information for each chromosome
separately. Doing so will result in the following structure.</p>
<pre><code>ProjectDir
|-derCoverageInfo
|---Chr1Cov.Rdata
|---Chr2Cov.Rdata
...
|---ChrYCov.Rdata
|---fullCov.Rdata
|-derAnalysis
|---analysis01
|---analysis02</code></pre>
<p>Next, you can use <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> for each of the
chromosomes of a specific analysis (say <em>analysis01</em>). Doing so
will create several Rdata files per chromosome as shown below.
<code>bash</code> scripts can be useful if you wish to submit one
cluster job per chromosome. In general, you will use the same model and
group information for each chromosome, so saving the information can be
useful.</p>
<pre><code>ProjectDir
|-derCoverageInfo
|---Chr1Cov.Rdata
|---Chr2Cov.Rdata
...
|---ChrYCov.Rdata
|---fullCov.Rdata
|-derAnalysis
|---analysis01
|-----models.Rdata
|-----groupInfo.Rdata
|-----chr1/
|-------chunksDir/
|-------logs/
|-------annotation.Rdata
|-------coveragePrep.Rdata
|-------fstats.Rdata
|-------optionsStats.Rdata
|-------regions.Rdata
|-------timeinfo.Rdata
|-----chr2/
...
|-----chrY/
|---analysis02</code></pre>
<p>Then use <code><a href="../reference/mergeResults.html">mergeResults()</a></code> to pool together the results
from all the chromosomes for a given analysis (here
<em>analysis01</em>).</p>
<pre><code>ProjectDir
|-derCoverageInfo
|---Chr1Cov.Rdata
|---Chr2Cov.Rdata
...
|---ChrYCov.Rdata
|---fullCov.Rdata
|-derAnalysis
|---analysis01
|-----logs/
|-----fullAnnotatedRegions.Rdata
|-----fullFstats.Rdata
|-----fullNullSummary.Rdata
|-----fullRegions.Rdata
|-----fullTime.Rdata
|-----optionsMerge.Rdata
|-----chr1/
|-------chunksDir/
|-------logs/
|-------annotation.Rdata
|-------coveragePrep.Rdata
|-------fstats.Rdata
|-------optionsStats.Rdata
|-------regions.Rdata
|-------timeinfo.Rdata
|-----chr2/
...
|-----chrY/
|---analysis02</code></pre>
<p>Finally, you might want to use <code>derfinderReport()</code> from
<em><a href="https://bioconductor.org/packages/3.17/regionReport" class="external-link">regionReport</a></em>
to create a HTML report of the results.</p>
</div>
<div class="section level3">
<h3 id="bash-scripts">bash scripts<a class="anchor" aria-label="anchor" href="#bash-scripts"></a>
</h3>
<p>If you are following our recommended file organization for a
base-level F-statistics DER finding analysis, you might be interested in
the following <code>bash</code> scripts templates. Note that there are
plenty of other solutions, such as writing a master <code>R</code>
script and using <em><a href="https://bioconductor.org/packages/3.17/BiocParallel" class="external-link">BiocParallel</a></em>
to submit jobs and administer them for you. Check out the
<code>BatchJobsParam()</code> for more information. If you want to use
<code>bash</code> scripts, take a look below.</p>
<p>For interacting between <code>bash</code> and <code>R</code> we have
found quite useful the <em><a href="https://CRAN.R-project.org/package=getopt" class="external-link">getopt</a></em>
package. Here we include an example <code>R</code> script that is
controlled by a <code>bash</code> script which submits a job for each
chromosome to analyze for a given analysis.</p>
<p>The two files, <em>derfinderAnalysis.R</em> and
<em>derAnalysis.sh</em> should live under the <em>derAnalysis</em>
directory.</p>
<pre><code>ProjectDir
|-derCoverageInfo
|---Chr1Cov.Rdata
|---Chr2Cov.Rdata
...
|---ChrYCov.Rdata
|---fullCov.Rdata
|-derAnalysis
|---derfinder-Analysis.R
|---derAnalysis.sh
|---analysis01
|---analysis02</code></pre>
<p>Then, you can simply use:</p>
<div class="sourceCode" id="cb419"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb419-1"><a href="#cb419-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /ProjectDir/derAnalysis</span>
<span id="cb419-2"><a href="#cb419-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sh</span> derAnalysis.sh analysis01</span></code></pre></div>
<p>to run <code><a href="../reference/analyzeChr.html">analyzeChr()</a></code> on all your chromosomes.</p>
<div class="section level4">
<h4 id="derfinder-analysis-r">derfinder-analysis.R<a class="anchor" aria-label="anchor" href="#derfinder-analysis-r"></a>
</h4>
<div class="sourceCode" id="cb420"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run derfinder's analysis steps with timing info</span></span>
<span></span>
<span><span class="co">## Load libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/trevorld/r-getopt" class="external-link">"getopt"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Available at http:/bioconductor.org/packages/derfinder</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/lcolladotor/derfinder" class="external-link">"derfinder"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Specify parameters</span></span>
<span><span class="va">spec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"DFfile"</span>, <span class="st">"d"</span>, <span class="fl">1</span>, <span class="st">"character"</span>, <span class="st">"path to the .Rdata file with the results from loadCoverage()"</span>,</span>
<span>    <span class="st">"chr"</span>, <span class="st">"c"</span>, <span class="fl">1</span>, <span class="st">"character"</span>, <span class="st">"Chromosome under analysis. Use X instead of chrX."</span>,</span>
<span>    <span class="st">"mcores"</span>, <span class="st">"m"</span>, <span class="fl">1</span>, <span class="st">"integer"</span>, <span class="st">"Number of cores"</span>,</span>
<span>    <span class="st">"verbose"</span>, <span class="st">"v"</span>, <span class="fl">2</span>, <span class="st">"logical"</span>, <span class="st">"Print status updates"</span>,</span>
<span>    <span class="st">"help"</span>, <span class="st">"h"</span>, <span class="fl">0</span>, <span class="st">"logical"</span>, <span class="st">"Display help"</span></span>
<span><span class="op">)</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">opt</span> <span class="op">&lt;-</span> <span class="fu">getopt</span><span class="op">(</span><span class="va">spec</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Testing the script</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">test</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Speficy it using an interactive R session and testing</span></span>
<span>    <span class="va">test</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Test values</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">test</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">opt</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>    <span class="va">opt</span><span class="op">$</span><span class="va">DFfile</span> <span class="op">&lt;-</span> <span class="st">"/ProjectDir/derCoverageInfo/chr21Cov.Rdata"</span></span>
<span>    <span class="va">opt</span><span class="op">$</span><span class="va">chr</span> <span class="op">&lt;-</span> <span class="st">"21"</span></span>
<span>    <span class="va">opt</span><span class="op">$</span><span class="va">mcores</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span>    <span class="va">opt</span><span class="op">$</span><span class="va">verbose</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## if help was asked for print a friendly message</span></span>
<span><span class="co">## and exit with a non-zero error code</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">opt</span><span class="op">$</span><span class="va">help</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu">getopt</span><span class="op">(</span><span class="va">spec</span>, usage <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/quit.html" class="external-link">q</a></span><span class="op">(</span>status <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Default value for verbose = TRUE</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">opt</span><span class="op">$</span><span class="va">verbose</span><span class="op">)</span><span class="op">)</span> <span class="va">opt</span><span class="op">$</span><span class="va">verbose</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">opt</span><span class="op">$</span><span class="va">verbose</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Loading Rdata file with the output from loadCoverage()"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="va">opt</span><span class="op">$</span><span class="va">DFfile</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Make it easy to use the name later. Here I'm assuming the names were generated using output='auto' in loadCoverage()</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/parse.html" class="external-link">parse</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"data &lt;- "</span>, <span class="st">"chr"</span>, <span class="va">opt</span><span class="op">$</span><span class="va">chr</span>, <span class="st">"CovInfo"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/parse.html" class="external-link">parse</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rm(chr"</span>, <span class="va">opt</span><span class="op">$</span><span class="va">chr</span>, <span class="st">"CovInfo)"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Just for testing purposes</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">test</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tmp</span> <span class="op">&lt;-</span> <span class="va">data</span></span>
<span>    <span class="va">tmp</span><span class="op">$</span><span class="va">coverage</span> <span class="op">&lt;-</span> <span class="va">tmp</span><span class="op">$</span><span class="va">coverage</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1e6</span>, <span class="op">]</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/IRanges" class="external-link">"IRanges"</a></span><span class="op">)</span></span>
<span>    <span class="va">tmp</span><span class="op">$</span><span class="va">position</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">pos</span><span class="op">)</span><span class="op">[</span><span class="fl">1e6</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">$</span><span class="va">pos</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></span>
<span>    <span class="va">data</span> <span class="op">&lt;-</span> <span class="va">tmp</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Load the models</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"models.Rdata"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load group information</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"groupInfo.Rdata"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Run the analysis with lowMemDir</span></span>
<span><span class="fu"><a href="../reference/analyzeChr.html">analyzeChr</a></span><span class="op">(</span></span>
<span>    chr <span class="op">=</span> <span class="va">opt</span><span class="op">$</span><span class="va">chr</span>, coverageInfo <span class="op">=</span> <span class="va">data</span>, models <span class="op">=</span> <span class="va">models</span>,</span>
<span>    cutoffFstat <span class="op">=</span> <span class="fl">1e-06</span>, cutoffType <span class="op">=</span> <span class="st">"theoretical"</span>, nPermute <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    seeds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>, maxClusterGap <span class="op">=</span> <span class="fl">3000</span>, groupInfo <span class="op">=</span> <span class="va">groupInfo</span>,</span>
<span>    subject <span class="op">=</span> <span class="st">"hg19"</span>, mc.cores <span class="op">=</span> <span class="va">opt</span><span class="op">$</span><span class="va">mcores</span>,</span>
<span>    lowMemDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="va">opt</span><span class="op">$</span><span class="va">chr</span><span class="op">)</span>, <span class="st">"chunksDir"</span><span class="op">)</span>,</span>
<span>    verbose <span class="op">=</span> <span class="va">opt</span><span class="op">$</span><span class="va">verbose</span>, chunksize <span class="op">=</span> <span class="fl">1e5</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Done</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">opt</span><span class="op">$</span><span class="va">verbose</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/proc.time.html" class="external-link">proc.time</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>, locale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Remember to modify the the script to fit your project.</p>
</div>
<div class="section level4">
<h4 id="deranalysis-sh">derAnalysis.sh<a class="anchor" aria-label="anchor" href="#deranalysis-sh"></a>
</h4>
<div class="sourceCode" id="cb421"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb421-1"><a href="#cb421-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/sh</span></span>
<span id="cb421-2"><a href="#cb421-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-3"><a href="#cb421-3" aria-hidden="true" tabindex="-1"></a><span class="co">## Usage</span></span>
<span id="cb421-4"><a href="#cb421-4" aria-hidden="true" tabindex="-1"></a><span class="co"># sh derAnalysis.sh analysis01</span></span>
<span id="cb421-5"><a href="#cb421-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-6"><a href="#cb421-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Directories</span></span>
<span id="cb421-7"><a href="#cb421-7" aria-hidden="true" tabindex="-1"></a><span class="va">MAINDIR</span><span class="op">=</span>/ProjectDir</span>
<span id="cb421-8"><a href="#cb421-8" aria-hidden="true" tabindex="-1"></a><span class="va">WDIR</span><span class="op">=</span><span class="va">${MAINDIR}</span>/derAnalysis</span>
<span id="cb421-9"><a href="#cb421-9" aria-hidden="true" tabindex="-1"></a><span class="va">DATADIR</span><span class="op">=</span><span class="va">${MAINDIR}</span>/derCoverageInfo</span>
<span id="cb421-10"><a href="#cb421-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-11"><a href="#cb421-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Define variables</span></span>
<span id="cb421-12"><a href="#cb421-12" aria-hidden="true" tabindex="-1"></a><span class="va">SHORT</span><span class="op">=</span><span class="st">'derA-01'</span></span>
<span id="cb421-13"><a href="#cb421-13" aria-hidden="true" tabindex="-1"></a><span class="va">PREFIX</span><span class="op">=</span><span class="va">$1</span></span>
<span id="cb421-14"><a href="#cb421-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-15"><a href="#cb421-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Construct shell files</span></span>
<span id="cb421-16"><a href="#cb421-16" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> chrnum <span class="kw">in</span> 22 21 Y 20 19 18 17 16 15 14 13 12 11 10 9 8 X 7 6 5 4 3 2 1</span>
<span id="cb421-17"><a href="#cb421-17" aria-hidden="true" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb421-18"><a href="#cb421-18" aria-hidden="true" tabindex="-1"></a>    <span class="bu">echo</span> <span class="st">"Creating script for chromosome </span><span class="va">${chrnum}</span><span class="st">"</span></span>
<span id="cb421-19"><a href="#cb421-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb421-20"><a href="#cb421-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">[[</span> <span class="va">${chrnum}</span> <span class="ot">==</span> <span class="st">"Y"</span> <span class="kw">]]</span></span>
<span id="cb421-21"><a href="#cb421-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">then</span></span>
<span id="cb421-22"><a href="#cb421-22" aria-hidden="true" tabindex="-1"></a>        <span class="va">CORES</span><span class="op">=</span>2</span>
<span id="cb421-23"><a href="#cb421-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span></span>
<span id="cb421-24"><a href="#cb421-24" aria-hidden="true" tabindex="-1"></a>        <span class="va">CORES</span><span class="op">=</span>6</span>
<span id="cb421-25"><a href="#cb421-25" aria-hidden="true" tabindex="-1"></a>    <span class="cf">fi</span></span>
<span id="cb421-26"><a href="#cb421-26" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb421-27"><a href="#cb421-27" aria-hidden="true" tabindex="-1"></a>    <span class="va">chr</span><span class="op">=</span><span class="st">"chr</span><span class="va">${chrnum}</span><span class="st">"</span></span>
<span id="cb421-28"><a href="#cb421-28" aria-hidden="true" tabindex="-1"></a>    <span class="va">outdir</span><span class="op">=</span><span class="st">"</span><span class="va">${PREFIX}</span><span class="st">/</span><span class="va">${chr}</span><span class="st">"</span></span>
<span id="cb421-29"><a href="#cb421-29" aria-hidden="true" tabindex="-1"></a>    <span class="va">sname</span><span class="op">=</span><span class="st">"</span><span class="va">${SHORT}</span><span class="st">.</span><span class="va">${PREFIX}</span><span class="st">.</span><span class="va">${chr}</span><span class="st">"</span></span>
<span id="cb421-30"><a href="#cb421-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span> <span class="op">&gt;</span> <span class="va">${WDIR}</span>/.<span class="va">${sname}</span>.sh <span class="op">&lt;&lt;EOF</span></span>
<span id="cb421-31"><a href="#cb421-31" aria-hidden="true" tabindex="-1"></a><span class="st">#!/bin/bash</span></span>
<span id="cb421-32"><a href="#cb421-32" aria-hidden="true" tabindex="-1"></a><span class="st">#$ -cwd</span></span>
<span id="cb421-33"><a href="#cb421-33" aria-hidden="true" tabindex="-1"></a><span class="st">#$ -m e</span></span>
<span id="cb421-34"><a href="#cb421-34" aria-hidden="true" tabindex="-1"></a><span class="st">#$ -l mem_free=8G,h_vmem=10G,h_fsize=10G</span></span>
<span id="cb421-35"><a href="#cb421-35" aria-hidden="true" tabindex="-1"></a><span class="st">#$ -N </span><span class="va">${sname}</span></span>
<span id="cb421-36"><a href="#cb421-36" aria-hidden="true" tabindex="-1"></a><span class="st">#$ -pe local </span><span class="va">${CORES}</span></span>
<span id="cb421-37"><a href="#cb421-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-38"><a href="#cb421-38" aria-hidden="true" tabindex="-1"></a><span class="st">echo "**** Job starts ****"</span></span>
<span id="cb421-39"><a href="#cb421-39" aria-hidden="true" tabindex="-1"></a><span class="st">date</span></span>
<span id="cb421-40"><a href="#cb421-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-41"><a href="#cb421-41" aria-hidden="true" tabindex="-1"></a><span class="st"># Create output directory </span></span>
<span id="cb421-42"><a href="#cb421-42" aria-hidden="true" tabindex="-1"></a><span class="st">mkdir -p </span><span class="va">${WDIR}</span><span class="st">/</span><span class="va">${outdir}</span></span>
<span id="cb421-43"><a href="#cb421-43" aria-hidden="true" tabindex="-1"></a><span class="st"># Make logs directory</span></span>
<span id="cb421-44"><a href="#cb421-44" aria-hidden="true" tabindex="-1"></a><span class="st">mkdir -p </span><span class="va">${WDIR}</span><span class="st">/</span><span class="va">${outdir}</span><span class="st">/logs</span></span>
<span id="cb421-45"><a href="#cb421-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-46"><a href="#cb421-46" aria-hidden="true" tabindex="-1"></a><span class="st"># run derfinder-analysis.R</span></span>
<span id="cb421-47"><a href="#cb421-47" aria-hidden="true" tabindex="-1"></a><span class="st">cd </span><span class="va">${WDIR}</span><span class="st">/</span><span class="va">${PREFIX}</span><span class="st">/</span></span>
<span id="cb421-48"><a href="#cb421-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-49"><a href="#cb421-49" aria-hidden="true" tabindex="-1"></a><span class="st"># specific to our cluster</span></span>
<span id="cb421-50"><a href="#cb421-50" aria-hidden="true" tabindex="-1"></a><span class="st"># see http://www.jhpce.jhu.edu/knowledge-base/environment-modules/</span></span>
<span id="cb421-51"><a href="#cb421-51" aria-hidden="true" tabindex="-1"></a><span class="st">module load R/3.3.x</span></span>
<span id="cb421-52"><a href="#cb421-52" aria-hidden="true" tabindex="-1"></a><span class="st">Rscript </span><span class="va">${WDIR}</span><span class="st">/derfinder-analysis.R -d "</span><span class="va">${DATADIR}</span><span class="st">/</span><span class="va">${chr}</span><span class="st">Cov.Rdata" -c "</span><span class="va">${chrnum}</span><span class="st">" -m </span><span class="va">${CORES}</span><span class="st"> -v TRUE</span></span>
<span id="cb421-53"><a href="#cb421-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-54"><a href="#cb421-54" aria-hidden="true" tabindex="-1"></a><span class="st"># Move log files into the logs directory</span></span>
<span id="cb421-55"><a href="#cb421-55" aria-hidden="true" tabindex="-1"></a><span class="st">mv </span><span class="va">${WDIR}</span><span class="st">/</span><span class="va">${sname}</span><span class="st">.* </span><span class="va">${WDIR}</span><span class="st">/</span><span class="va">${outdir}</span><span class="st">/logs/</span></span>
<span id="cb421-56"><a href="#cb421-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb421-57"><a href="#cb421-57" aria-hidden="true" tabindex="-1"></a><span class="st">echo "**** Job ends ****"</span></span>
<span id="cb421-58"><a href="#cb421-58" aria-hidden="true" tabindex="-1"></a><span class="st">date</span></span>
<span id="cb421-59"><a href="#cb421-59" aria-hidden="true" tabindex="-1"></a><span class="op">EOF</span></span>
<span id="cb421-60"><a href="#cb421-60" aria-hidden="true" tabindex="-1"></a>    <span class="va">call</span><span class="op">=</span><span class="st">"qsub .</span><span class="va">${sname}</span><span class="st">.sh"</span></span>
<span id="cb421-61"><a href="#cb421-61" aria-hidden="true" tabindex="-1"></a>    <span class="va">$call</span></span>
<span id="cb421-62"><a href="#cb421-62" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
<p>Your cluster might specify memory requirements differently and you
might need to use fewer or more cores depending on your data set.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="expressed-regions-level-projects">Expressed regions-level projects<a class="anchor" aria-label="anchor" href="#expressed-regions-level-projects"></a>
</h2>
<p>If you are using <code><a href="../reference/regionMatrix.html">regionMatrix()</a></code>, an organization similar
to the one we mentioned for the single base-level F-statistics
implementation is recommended. If you have a very large number of
samples, it might be best to store the coverage data separately by
chromosome instead of a single <code>fullCov</code> object. This is help
with the memory load since you can then write a script for loading the
coverage data for the chromosome of interest and running
<code><a href="../reference/regionMatrix.html">regionMatrix()</a></code> on it. Once all your jobs are done, merge
the results from them.</p>
<p>If you are using <code><a href="../reference/railMatrix.html">railMatrix()</a></code>, you don’t need any fancy
setup. You only need the paths to the summary files and the sample
files, all of which are assumed to be BigWig files.</p>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>We have illustrated how to identify candidate differentially
expressed regions without using annotation by using
<code><a href="../reference/analyzeChr.html">analyzeChr()</a></code>. Furthermore, we covered how to perform the
expressed region-level matrix analysis with <code><a href="../reference/regionMatrix.html">regionMatrix()</a></code>
or <code><a href="../reference/railMatrix.html">railMatrix()</a></code>. We also highlighted other uses of the
<em><a href="https://bioconductor.org/packages/3.17/derfinder" class="external-link">derfinder</a></em>
package.</p>
<p>Latter sections of this vignette covered the most commonly used
advanced arguments, details on how to load data, flow charts explaining
the relationships between the functions, the recommended output
organization, and example shell scripts for running the analysis.</p>
</div>
<div class="section level2">
<h2 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h2>
<p>This package was made possible thanks to:</p>
<ul>
<li>R (R Core Team, 2023)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/AnnotationDbi" class="external-link">AnnotationDbi</a></em>
(Pagès, Carlson, Falcon, and Li, 2017)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/BiocParallel" class="external-link">BiocParallel</a></em>
(Morgan, Wang, Obenchain, Lang, Thompson, and Turaga, 2023)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/bumphunter" class="external-link">bumphunter</a></em>
(Jaffe, Murakami, Lee, Leek, Fallin, Feinberg, and Irizarry, 2012) and
(Jaffe, Murakami, Lee, Leek, Fallin, Feinberg, and Irizarry, 2012)</li>
<li><em><a href="https://bioconductor.org/packages/3.17/derfinderHelper" class="external-link">derfinderHelper</a></em></li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/GenomeInfoDb" class="external-link">GenomeInfoDb</a></em>
(Arora, Morgan, Carlson, and Pagès, 2017)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/GenomicAlignments" class="external-link">GenomicAlignments</a></em>
(Lawrence, Huber, Pagès, Aboyoun, Carlson, Gentleman, Morgan, and Carey,
2013)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/GenomicFeatures" class="external-link">GenomicFeatures</a></em>
(Lawrence, Huber, Pagès et al., 2013)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/GenomicFiles" class="external-link">GenomicFiles</a></em>
(Bioconductor Package Maintainer, Obenchain, Love, Shepherd, and Morgan,
2023)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/GenomicRanges" class="external-link">GenomicRanges</a></em>
(Lawrence, Huber, Pagès et al., 2013)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=Hmisc" class="external-link">Hmisc</a></em> (Harrell
Jr, 2023)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/IRanges" class="external-link">IRanges</a></em>
(Lawrence, Huber, Pagès et al., 2013)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/qvalue" class="external-link">qvalue</a></em>
(Storey, Bass, Dabney, and Robinson, 2023)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/Rsamtools" class="external-link">Rsamtools</a></em>
(Morgan, Pagès, Obenchain, and Hayden, 2023)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/rtracklayer" class="external-link">rtracklayer</a></em>
(Lawrence, Gentleman, and Carey, 2009)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/S4Vectors" class="external-link">S4Vectors</a></em>
(Pagès, Lawrence, and Aboyoun, 2017)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/derfinderData" class="external-link">derfinderData</a></em>
(Collado-Torres, Jaffe, and Leek, 2023)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=sessioninfo" class="external-link">sessioninfo</a></em>
(Wickham, Chang, Flight, Müller, and Hester, 2021)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=ggplot2" class="external-link">ggplot2</a></em>
(Wickham, 2016)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitr" class="external-link">knitr</a></em> (Xie,
2014)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/BiocStyle" class="external-link">BiocStyle</a></em>
(Oleś, 2023)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=RefManageR" class="external-link">RefManageR</a></em>
(McLean, 2017)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">rmarkdown</a></em>
(Allaire, Xie, Dervieux, McPherson, Luraschi, Ushey, Atkins, Wickham,
Cheng, Chang, and Iannone, 2023)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=testthat" class="external-link">testthat</a></em>
(Wickham, 2011)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.17/TxDb.Hsapiens.UCSC.hg19.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg19.knownGene</a></em>
(Carlson and Maintainer, 2015)</li>
</ul>
<p>Code for creating the vignette</p>
<div class="sourceCode" id="cb422"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create the vignette</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/rstudio/rmarkdown" class="external-link">"rmarkdown"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html" class="external-link">render</a></span><span class="op">(</span><span class="st">"derfinder-users-guide.Rmd"</span>, <span class="st">"BiocStyle::html_document"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract the R code</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://yihui.org/knitr/" class="external-link">"knitr"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/knit.html" class="external-link">knit</a></span><span class="op">(</span><span class="st">"derfinder-users-guide.Rmd"</span>, tangle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb423"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Clean up</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="st">"derfinderUsersGuideRef.bib"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in file.remove("derfinderUsersGuideRef.bib"): cannot remove file</span></span>
<span><span class="co">## 'derfinderUsersGuideRef.bib', reason 'No such file or directory'</span></span></code></pre>
<pre><code><span><span class="co">## [1] FALSE</span></span></code></pre>
<div class="sourceCode" id="cb426"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="st">"analysisResults"</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="st">"HSB113.bw"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb428"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="st">"meanChr21.bw"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>Date the vignette was generated.</p>
<pre><code><span><span class="co">## [1] "2023-05-07 03:56:43 UTC"</span></span></code></pre>
<p>Wallclock time spent generating the vignette.</p>
<pre><code><span><span class="co">## Time difference of 2.05 mins</span></span></code></pre>
<p><code>R</code> session information.</p>
<pre><code><span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">##  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">##  version  R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">##  os       Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">##  system   x86_64, linux-gnu</span></span>
<span><span class="co">##  ui       X11</span></span>
<span><span class="co">##  language en</span></span>
<span><span class="co">##  collate  en_US.UTF-8</span></span>
<span><span class="co">##  ctype    en_US.UTF-8</span></span>
<span><span class="co">##  tz       UTC</span></span>
<span><span class="co">##  date     2023-05-07</span></span>
<span><span class="co">##  pandoc   2.19.2 @ /usr/local/bin/ (via rmarkdown)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">##  <span style="color: #555555; font-style: italic;">package                          </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version  </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">##  AnnotationDbi                     * 1.62.1    <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  backports                           1.4.1     <span style="color: #555555;">2021-12-13</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  base64enc                           0.1-3     <span style="color: #555555;">2015-07-28</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  bibtex                              0.5.1     <span style="color: #555555;">2023-01-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  Biobase                           * 2.60.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  BiocFileCache                       2.8.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  BiocGenerics                      * 0.46.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  BiocIO                              1.10.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  BiocManager                         1.30.20   <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  BiocParallel                        1.34.1    <span style="color: #555555;">2023-05-05</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  BiocStyle                         * 2.28.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  biomaRt                             2.56.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  Biostrings                          2.68.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  bit                                 4.0.5     <span style="color: #555555;">2022-11-15</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  bit64                               4.0.5     <span style="color: #555555;">2020-08-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  bitops                              1.0-7     <span style="color: #555555;">2021-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  blob                                1.2.4     <span style="color: #555555;">2023-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  bookdown                            0.33      <span style="color: #555555;">2023-03-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  BSgenome                            1.68.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  bslib                               0.4.2     <span style="color: #555555;">2022-12-16</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  bumphunter                        * 1.42.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  cachem                              1.0.8     <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  checkmate                           2.2.0     <span style="color: #555555;">2023-04-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  cli                                 3.6.1     <span style="color: #555555;">2023-03-23</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  cluster                             2.1.4     <span style="color: #555555;">2022-08-22</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  codetools                           0.2-19    <span style="color: #555555;">2023-02-01</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  colorspace                          2.1-0     <span style="color: #555555;">2023-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  crayon                              1.5.2     <span style="color: #555555;">2022-09-29</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  curl                                5.0.0     <span style="color: #555555;">2023-01-12</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  data.table                          1.14.8    <span style="color: #555555;">2023-02-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  DBI                                 1.1.3     <span style="color: #555555;">2022-06-18</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  dbplyr                              2.3.2     <span style="color: #555555;">2023-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  DelayedArray                        0.26.2    <span style="color: #555555;">2023-05-05</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  derfinder                         * 1.35.0    <span style="color: #555555;">2023-05-07</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  derfinderData                     * 2.18.0    <span style="color: #555555;">2023-04-27</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  derfinderHelper                     1.34.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  desc                                1.4.2     <span style="color: #555555;">2022-09-08</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  DESeq2                            * 1.40.1    <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  digest                              0.6.31    <span style="color: #555555;">2022-12-11</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  doRNG                               1.8.6     <span style="color: #555555;">2023-01-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  dplyr                               1.1.2     <span style="color: #555555;">2023-04-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  evaluate                            0.20      <span style="color: #555555;">2023-01-17</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  fansi                               1.0.4     <span style="color: #555555;">2023-01-22</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  farver                              2.1.1     <span style="color: #555555;">2022-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  fastmap                             1.1.1     <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  filelock                            1.0.2     <span style="color: #555555;">2018-10-05</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  foreach                           * 1.5.2     <span style="color: #555555;">2022-02-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  foreign                             0.8-84    <span style="color: #555555;">2022-12-06</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  Formula                             1.2-5     <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  fs                                  1.6.2     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  generics                            0.1.3     <span style="color: #555555;">2022-07-05</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  GenomeInfoDb                      * 1.36.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  GenomeInfoDbData                    1.2.10    <span style="color: #555555;">2023-05-07</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  GenomicAlignments                   1.36.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  GenomicFeatures                   * 1.52.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  GenomicFiles                        1.36.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  GenomicRanges                     * 1.52.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  ggplot2                           * 3.4.2     <span style="color: #555555;">2023-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  glue                                1.6.2     <span style="color: #555555;">2022-02-24</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  gridExtra                           2.3       <span style="color: #555555;">2017-09-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  gtable                              0.3.3     <span style="color: #555555;">2023-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  highr                               0.10      <span style="color: #555555;">2022-12-22</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  Hmisc                               5.0-1     <span style="color: #555555;">2023-03-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  hms                                 1.1.3     <span style="color: #555555;">2023-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  htmlTable                           2.4.1     <span style="color: #555555;">2022-07-07</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  htmltools                           0.5.5     <span style="color: #555555;">2023-03-23</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  htmlwidgets                         1.6.2     <span style="color: #555555;">2023-03-17</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  httr                                1.4.5     <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  IRanges                           * 2.34.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  iterators                         * 1.0.14    <span style="color: #555555;">2022-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  jquerylib                           0.1.4     <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  jsonlite                            1.8.4     <span style="color: #555555;">2022-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  KEGGREST                            1.40.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  knitr                             * 1.42      <span style="color: #555555;">2023-01-25</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  labeling                            0.4.2     <span style="color: #555555;">2020-10-20</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  lattice                             0.21-8    <span style="color: #555555;">2023-04-05</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  lifecycle                           1.0.3     <span style="color: #555555;">2022-10-07</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  limma                             * 3.56.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  locfit                            * 1.5-9.7   <span style="color: #555555;">2023-01-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  lubridate                           1.9.2     <span style="color: #555555;">2023-02-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  magrittr                            2.0.3     <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  Matrix                              1.5-4     <span style="color: #555555;">2023-04-04</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  MatrixGenerics                    * 1.12.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  matrixStats                       * 0.63.0    <span style="color: #555555;">2022-11-18</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  memoise                             2.0.1     <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  munsell                             0.5.0     <span style="color: #555555;">2018-06-12</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  nnet                                7.3-19    <span style="color: #555555;">2023-05-03</span> <span style="color: #555555;">[3]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  pillar                              1.9.0     <span style="color: #555555;">2023-03-22</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  pkgconfig                           2.0.3     <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  pkgdown                             2.0.7     <span style="color: #555555;">2022-12-14</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  plyr                                1.8.8     <span style="color: #555555;">2022-11-11</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  png                                 0.1-8     <span style="color: #555555;">2022-11-29</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  prettyunits                         1.1.1     <span style="color: #555555;">2020-01-24</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  progress                            1.2.2     <span style="color: #555555;">2019-05-16</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  purrr                               1.0.1     <span style="color: #555555;">2023-01-10</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  qvalue                              2.32.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  R6                                  2.5.1     <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  ragg                                1.2.5     <span style="color: #555555;">2023-01-12</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  rappdirs                            0.3.3     <span style="color: #555555;">2021-01-31</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  Rcpp                                1.0.10    <span style="color: #555555;">2023-01-22</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  RCurl                               1.98-1.12 <span style="color: #555555;">2023-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  RefManageR                        * 1.4.0     <span style="color: #555555;">2022-09-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  reshape2                            1.4.4     <span style="color: #555555;">2020-04-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  restfulr                            0.0.15    <span style="color: #555555;">2022-06-16</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  rjson                               0.2.21    <span style="color: #555555;">2022-01-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  rlang                               1.1.1     <span style="color: #555555;">2023-04-28</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  rmarkdown                           2.21      <span style="color: #555555;">2023-03-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  rngtools                            1.5.2     <span style="color: #555555;">2021-09-20</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  rpart                               4.1.19    <span style="color: #555555;">2022-10-21</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  rprojroot                           2.0.3     <span style="color: #555555;">2022-04-02</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  Rsamtools                           2.16.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  RSQLite                             2.3.1     <span style="color: #555555;">2023-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  rstudioapi                          0.14      <span style="color: #555555;">2022-08-22</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  rtracklayer                         1.60.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  S4Arrays                            1.0.1     <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  S4Vectors                         * 0.38.1    <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  sass                                0.4.6     <span style="color: #555555;">2023-05-03</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  scales                              1.2.1     <span style="color: #555555;">2022-08-20</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  sessioninfo                       * 1.2.2     <span style="color: #555555;">2021-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  stringi                             1.7.12    <span style="color: #555555;">2023-01-11</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  stringr                             1.5.0     <span style="color: #555555;">2022-12-02</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  SummarizedExperiment              * 1.30.1    <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  systemfonts                         1.0.4     <span style="color: #555555;">2022-02-11</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  textshaping                         0.3.6     <span style="color: #555555;">2021-10-13</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  tibble                              3.2.1     <span style="color: #555555;">2023-03-20</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  tidyselect                          1.2.0     <span style="color: #555555;">2022-10-10</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">##  timechange                          0.2.0     <span style="color: #555555;">2023-01-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  TxDb.Hsapiens.UCSC.hg19.knownGene * 3.2.2     <span style="color: #555555;">2022-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  utf8                                1.2.3     <span style="color: #555555;">2023-01-31</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  VariantAnnotation                   1.46.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  vctrs                               0.6.2     <span style="color: #555555;">2023-04-19</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  withr                               2.5.0     <span style="color: #555555;">2022-03-03</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  xfun                                0.39      <span style="color: #555555;">2023-04-20</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  XML                                 3.99-0.14 <span style="color: #555555;">2023-03-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  xml2                                1.3.4     <span style="color: #555555;">2023-04-27</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  XVector                             0.40.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">##  yaml                                2.3.7     <span style="color: #555555;">2023-01-23</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">##  zlibbioc                            1.46.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #555555;"> [1] /__w/_temp/Library</span></span></span>
<span><span class="co">## <span style="color: #555555;"> [2] /usr/local/lib/R/site-library</span></span></span>
<span><span class="co">## <span style="color: #555555;"> [3] /usr/local/lib/R/library</span></span></span>
<span><span class="co">## </span></span>
<span><span class="co">## <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.17/BiocStyle" class="external-link">BiocStyle</a></em>
(Oleś, 2023) with <em><a href="https://CRAN.R-project.org/package=knitr" class="external-link">knitr</a></em> (Xie,
2014) and <em><a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">rmarkdown</a></em>
(Allaire, Xie, Dervieux et al., 2023) running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=RefManageR" class="external-link">RefManageR</a></em>
(McLean, 2017).</p>
<p>
<a id="bib-allaire2023rmarkdown"></a><a href="#cite-allaire2023rmarkdown">[1]</a><cite>
J. Allaire, Y. Xie, C. Dervieux, et al. <em>rmarkdown: Dynamic Documents
for R</em>. R package version 2.21. 2023. URL:
<a href="https://github.com/rstudio/rmarkdown" class="external-link">https://github.com/rstudio/rmarkdown</a>.</cite>
</p>
<p>
<a id="bib-GenomeInfoDb"></a><a href="#cite-GenomeInfoDb">[2]</a><cite>
S. Arora, M. Morgan, M. Carlson, et al. <em>GenomeInfoDb: Utilities for
manipulating chromosome and other ‘seqname’ identifiers</em>. 2017. DOI:
<a href="https://doi.org/10.18129/B9.bioc.GenomeInfoDb" class="external-link">10.18129/B9.bioc.GenomeInfoDb</a>.</cite>
</p>
<p>
<a id="bib-2023genomicfiles"></a><a href="#cite-2023genomicfiles">[3]</a><cite>
Bioconductor Package Maintainer, V. Obenchain, M. Love, et al.
<em>GenomicFiles: Distributed computing by file or by range</em>. R
package version 1.36.0. 2023. DOI:
<a href="https://doi.org/10.18129/B9.bioc.GenomicFiles" class="external-link">10.18129/B9.bioc.GenomicFiles</a>.
URL:
<a href="https://bioconductor.org/packages/GenomicFiles" class="external-link">https://bioconductor.org/packages/GenomicFiles</a>.</cite>
</p>
<p>
<a id="bib-brainspan"></a><a href="#cite-brainspan">[4]</a><cite>
BrainSpan. “Atlas of the Developing Human Brain [Internet]. Funded by
ARRA Awards 1RC2MH089921-01, 1RC2MH090047-01, and 1RC2MH089929-01.”
2011. URL:
<a href="http://www.brainspan.org/" class="external-link">http://www.brainspan.org/</a>.</cite>
</p>
<p>
<a id="bib-carlson2015txdb"></a><a href="#cite-carlson2015txdb">[5]</a><cite>
M. Carlson and B. P. Maintainer. <em>TxDb.Hsapiens.UCSC.hg19.knownGene:
Annotation package for TxDb object(s)</em>. R package version 3.2.2.
2015.</cite>
</p>
<p>
<a id="bib-colladotorres2023derfinderdata"></a><a href="#cite-colladotorres2023derfinderdata">[6]</a><cite>
L. Collado-Torres, A. Jaffe, and J. Leek. <em>derfinderData: Processed
BigWigs from BrainSpan for examples</em>. R package version 2.18.0.
2023. DOI:
<a href="https://doi.org/10.18129/B9.bioc.derfinderData" class="external-link">10.18129/B9.bioc.derfinderData</a>.
URL:
<a href="https://bioconductor.org/packages/derfinderData" class="external-link">https://bioconductor.org/packages/derfinderData</a>.</cite>
</p>
<p>
<a id="bib-frazee2014differential"></a><a href="#cite-frazee2014differential">[7]</a><cite>
A. C. Frazee, S. Sabunciyan, K. D. Hansen, et al. “Differential
expression analysis of RNA-seq data at single-base resolution”. In:
<em>Biostatistics</em> 15 (3) (2014), pp. 413-426. DOI:
<a href="https://doi.org/10.1093/biostatistics/kxt053" class="external-link">10.1093/biostatistics/kxt053</a>.
URL:
<a href="http://biostatistics.oxfordjournals.org/content/15/3/413.long" class="external-link">http://biostatistics.oxfordjournals.org/content/15/3/413.long</a>.</cite>
</p>
<p>
<a id="bib-harrelljr2023hmisc"></a><a href="#cite-harrelljr2023hmisc">[8]</a><cite>
F. Harrell Jr. <em>Hmisc: Harrell Miscellaneous</em>. R package version
5.0-1. 2023. URL:
<a href="https://hbiostat.org/R/Hmisc/" class="external-link">https://hbiostat.org/R/Hmisc/</a>.</cite>
</p>
<p>
<a id="bib-bumphunter"></a><a href="#cite-bumphunter">[9]</a><cite> A.
E. Jaffe, P. Murakami, H. Lee, et al. “Bump hunting to identify
differentially methylated regions in epigenetic epidemiology studies”.
In: <em>International journal of epidemiology</em> 41.1 (2012),
pp. 200–209. DOI:
<a href="https://doi.org/10.1093/ije/dyr238" class="external-link">10.1093/ije/dyr238</a>.</cite>
</p>
<p>
<a id="bib-bumphunterPaper"></a><a href="#cite-bumphunterPaper">[10]</a><cite>
A. E. Jaffe, P. Murakami, H. Lee, et al. “Bump hunting to identify
differentially methylated regions in epigenetic epidemiology studies”.
In: <em>International Journal of Epidemiology</em> (2012).</cite>
</p>
<p>
<a id="bib-lawrence2009rtracklayer"></a><a href="#cite-lawrence2009rtracklayer">[11]</a><cite>
M. Lawrence, R. Gentleman, and V. Carey. “rtracklayer: an R package for
interfacing with genome browsers”. In: <em>Bioinformatics</em> 25
(2009), pp. 1841-1842. DOI:
<a href="https://doi.org/10.1093/bioinformatics/btp328" class="external-link">10.1093/bioinformatics/btp328</a>.
URL:
<a href="http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract" class="external-link">http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract</a>.</cite>
</p>
<p>
<a id="bib-lawrence2013software"></a><a href="#cite-lawrence2013software">[12]</a><cite>
M. Lawrence, W. Huber, H. Pagès, et al. “Software for Computing and
Annotating Genomic Ranges”. In: <em>PLoS Computational Biology</em> 9 (8
2013). DOI:
<a href="https://doi.org/10.1371/journal.pcbi.1003118" class="external-link">10.1371/journal.pcbi.1003118</a>.
URL:
<a href="http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118%7D." class="external-link"></a><a href="http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118" class="external-link uri">http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118</a>}.
</cite></p>


<p>
<a id="bib-mclean2017refmanager"></a><a href="#cite-mclean2017refmanager">[13]</a><cite>
M. W. McLean. “RefManageR: Import and Manage BibTeX and BibLaTeX
References in R”. In: <em>The Journal of Open Source Software</em>
(2017). DOI:
<a href="https://doi.org/10.21105/joss.00338" class="external-link">10.21105/joss.00338</a>.</cite>
</p>
<p>
<a id="bib-morgan2023rsamtools"></a><a href="#cite-morgan2023rsamtools">[14]</a><cite>
M. Morgan, H. Pagès, V. Obenchain, et al. <em>Rsamtools: Binary
alignment (BAM), FASTA, variant call (BCF), and tabix file import</em>.
R package version 2.16.0. 2023. DOI:
<a href="https://doi.org/10.18129/B9.bioc.Rsamtools" class="external-link">10.18129/B9.bioc.Rsamtools</a>.
URL:
<a href="https://bioconductor.org/packages/Rsamtools" class="external-link">https://bioconductor.org/packages/Rsamtools</a>.</cite>
</p>
<p>
<a id="bib-morgan2023biocparallel"></a><a href="#cite-morgan2023biocparallel">[15]</a><cite>
M. Morgan, J. Wang, V. Obenchain, et al. <em>BiocParallel: Bioconductor
facilities for parallel evaluation</em>. R package version 1.34.1. 2023.
DOI:
<a href="https://doi.org/10.18129/B9.bioc.BiocParallel" class="external-link">10.18129/B9.bioc.BiocParallel</a>.
URL:
<a href="https://bioconductor.org/packages/BiocParallel" class="external-link">https://bioconductor.org/packages/BiocParallel</a>.</cite>
</p>
<p>
<a id="bib-ole2023biocstyle"></a><a href="#cite-ole2023biocstyle">[16]</a><cite>
A. Oleś. <em>BiocStyle: Standard styles for vignettes and other
Bioconductor documents</em>. R package version 2.28.0. 2023. DOI:
<a href="https://doi.org/10.18129/B9.bioc.BiocStyle" class="external-link">10.18129/B9.bioc.BiocStyle</a>.
URL:
<a href="https://bioconductor.org/packages/BiocStyle" class="external-link">https://bioconductor.org/packages/BiocStyle</a>.</cite>
</p>
<p>
<a id="bib-AnnotationDbi"></a><a href="#cite-AnnotationDbi">[17]</a><cite>
H. Pagès, M. Carlson, S. Falcon, et al. <em>AnnotationDbi: Annotation
Database Interface</em>. 2017. DOI:
<a href="https://doi.org/10.18129/B9.bioc.AnnotationDbi" class="external-link">10.18129/B9.bioc.AnnotationDbi</a>.</cite>
</p>
<p>
<a id="bib-S4Vectors"></a><a href="#cite-S4Vectors">[18]</a><cite> H.
Pagès, M. Lawrence, and P. Aboyoun. <em>S4Vectors: S4 implementation of
vector-like and list-like objects</em>. 2017. DOI:
<a href="https://doi.org/10.18129/B9.bioc.S4Vectors" class="external-link">10.18129/B9.bioc.S4Vectors</a>.</cite>
</p>
<p>
<a id="bib-2023language"></a><a href="#cite-2023language">[19]</a><cite>
R Core Team. <em>R: A Language and Environment for Statistical
Computing</em>. R Foundation for Statistical Computing. Vienna, Austria,
2023. URL:
<a href="https://www.R-project.org/" class="external-link">https://www.R-project.org/</a>.</cite>
</p>
<p>
<a id="bib-storey2023qvalue"></a><a href="#cite-storey2023qvalue">[20]</a><cite>
J. D. Storey, A. J. Bass, A. Dabney, et al. <em>qvalue: Q-value
estimation for false discovery rate control</em>. R package version
2.32.0. 2023. DOI:
<a href="https://doi.org/10.18129/B9.bioc.qvalue" class="external-link">10.18129/B9.bioc.qvalue</a>.
URL:
<a href="https://bioconductor.org/packages/qvalue" class="external-link">https://bioconductor.org/packages/qvalue</a>.</cite>
</p>
<p>
<a id="bib-wickham2016ggplot2"></a><a href="#cite-wickham2016ggplot2">[21]</a><cite>
H. Wickham. <em>ggplot2: Elegant Graphics for Data Analysis</em>.
Springer-Verlag New York, 2016. ISBN: 978-3-319-24277-4. URL:
<a href="https://ggplot2.tidyverse.org" class="external-link">https://ggplot2.tidyverse.org</a>.</cite>
</p>
<p>
<a id="bib-wickham2011testthat"></a><a href="#cite-wickham2011testthat">[22]</a><cite>
H. Wickham. “testthat: Get Started with Testing”. In: <em>The R
Journal</em> 3 (2011), pp. 5–10. URL:
<a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf" class="external-link">https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf</a>.</cite>
</p>
<p>
<a id="bib-wickham2021sessioninfo"></a><a href="#cite-wickham2021sessioninfo">[23]</a><cite>
H. Wickham, W. Chang, R. Flight, et al. <em>sessioninfo: R Session
Information</em>. <a href="https://github.com/r-lib/sessioninfo#readme" class="external-link uri">https://github.com/r-lib/sessioninfo#readme</a>, <a href="https://r-lib.github.io/sessioninfo/" class="external-link uri">https://r-lib.github.io/sessioninfo/</a>. 2021.</cite>
</p>
<p>
<a id="bib-xie2014knitr"></a><a href="#cite-xie2014knitr">[24]</a><cite>
Y. Xie. “knitr: A Comprehensive Tool for Reproducible Research in R”.
In: <em>Implementing Reproducible Computational Research</em>. Ed. by V.
Stodden, F. Leisch and R. D. Peng. ISBN 978-1466561595. Chapman and
Hall/CRC, 2014.</cite>
</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Leonardo Collado-Torres, Andrew E. Jaffe, Jeffrey T. Leek.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
